{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a54318ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import listdir\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "304b386e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['09_summarized_workflow.md']\n"
     ]
    }
   ],
   "source": [
    "files = listdir(\"./develop/\")\n",
    "\n",
    "print(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "1be423ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---\n",
      "title: \"Summary of DGE workflow\"\n",
      "summary: You can use this lesson as a template for your own bulk RNAseq analysis\n",
      "date: 2023-01-17\n",
      "hide:\n",
      "  - navigation\n",
      "\n",
      "knit: (function(inputFile, encoding) { \n",
      "      rmarkdown::render(inputFile,\n",
      "                        encoding=encoding,\n",
      "                        output_format='all',\n",
      "                        output_dir='./develop/')})\n",
      "output:\n",
      "  github_document: \n",
      "     preserve_yaml: TRUE\n",
      "     html_preview: FALSE\n",
      "---\n",
      "\n",
      "Summary of DGE workflow\n",
      "================\n",
      "2023-01-17\n",
      "\n",
      "# Summarised workflow\n",
      "\n",
      "We have detailed the various steps in a differential expression analysis\n",
      "workflow, providing theory with example code. To provide a more succinct\n",
      "reference for the code needed to run a DGE analysis, we have summarized\n",
      "the steps in an analysis below:\n",
      "\n",
      "## Libraries\n",
      "\n",
      "``` r\n",
      "library(tidyverse)\n",
      "library(DESeq2)\n",
      "library(ggrepel)\n",
      "library(pheatmap)\n",
      "library(annotables)\n",
      "library(clusterProfiler)\n",
      "library(DOSE)\n",
      "library(pathview)\n",
      "library(org.Hs.eg.db)\n",
      "library(tximport)\n",
      "library(RColorBrewer)\n",
      "```\n",
      "\n",
      "## Obtaining gene-level counts from your preprocessing and create DESeq object\n",
      "\n",
      "### If you have a traditional raw count matrix\n",
      "\n",
      "Load data and metadata\n",
      "\n",
      "``` r\n",
      "data <- read_table(\"../Data/Mov10_full_counts.txt\") \n",
      "\n",
      "meta <- read_table(\"../Data/Mov10_full_meta.txt\")\n",
      "```\n",
      "\n",
      "Check that the row names of the metadata equal the column names of the\n",
      "**raw counts** data\n",
      "\n",
      "``` r\n",
      "all(colnames(data)[-1] == meta$sample)\n",
      "```\n",
      "\n",
      "Create DESeq2Dataset object\n",
      "\n",
      "``` r\n",
      "dds <- DESeqDataSetFromMatrix(countData = data %>% column_to_rownames(\"GeneSymbol\"), \n",
      "                              colData = meta %>% column_to_rownames(\"sample\"), \n",
      "                              design = ~ condition)\n",
      "```\n",
      "\n",
      "### If you have pseudocounts\n",
      "\n",
      "Load samplesheet with all our metadata from our pipeline\n",
      "\n",
      "``` r\n",
      "# Load data, metadata and tx2gene and create a txi object\n",
      "meta <- read_csv(\"../Data/samplesheet.csv\")\n",
      "```\n",
      "\n",
      "Create a list of salmon results\n",
      "\n",
      "``` r\n",
      "dir <- \"/work/sequencing_data/Preprocessing_backup/results_salmon/salmon\"\n",
      "tx2gene <- read_table(file.path(dir,\"salmon_tx2gene.tsv\"), col_names = c(\"transcript_ID\",\"gene_ID\",\"gene_symbol\"))\n",
      "\n",
      "# Get all salmon results files\n",
      "files <- file.path(dir, meta$sample, \"quant.sf\")\n",
      "names(files) <- meta$sample\n",
      "```\n",
      "\n",
      "Create txi object\n",
      "\n",
      "``` r\n",
      "txi <- tximport(files, type=\"salmon\", tx2gene=tx2gene, countsFromAbundance = \"lengthScaledTPM\", ignoreTxVersion = TRUE)\n",
      "```\n",
      "\n",
      "Create dds object\n",
      "\n",
      "``` r\n",
      "dds <- DESeqDataSetFromTximport(txi,\n",
      "                                   colData = meta %>% column_to_rownames(\"sample\"), \n",
      "                              design = ~ condition)\n",
      "```\n",
      "\n",
      "## Exploratory data analysis\n",
      "\n",
      "Prefiltering low count genes + PCA & hierarchical clustering -\n",
      "identifying outliers and sources of variation in the data:\n",
      "\n",
      "### Prefiltering low count genes\n",
      "\n",
      "``` r\n",
      "keep <- rowSums(counts(dds)) > 0\n",
      "dds <- dds[keep,]\n",
      "```\n",
      "\n",
      "### Rlog transformation\n",
      "\n",
      "``` r\n",
      "# Transform counts for data visualization\n",
      "rld <- rlog(dds, \n",
      "            blind=TRUE)\n",
      "```\n",
      "\n",
      "### PCA\n",
      "\n",
      "Plot PCA\n",
      "\n",
      "``` r\n",
      "plotPCA(rld, \n",
      "        intgroup=\"condition\")\n",
      "```\n",
      "\n",
      "### Heatmaps\n",
      "\n",
      "Extract the rlog matrix from the object\n",
      "\n",
      "``` r\n",
      "rld_mat <- assay(rld)\n",
      "rld_cor <- cor(rld_mat) # Pearson correlation betweeen samples\n",
      "rld_dist <- as.matrix(dist(t(assay(rld)))) #distances are computed by rows, so we need to transponse the matrix\n",
      "```\n",
      "\n",
      "Plot heatmap of correlations\n",
      "\n",
      "``` r\n",
      "pheatmap(rld_cor, \n",
      "         annotation = meta %>% column_to_rownames(\"sample\") %>% select(\"condition\"))\n",
      "```\n",
      "\n",
      "Plot heatmap of distances with a new color range\n",
      "\n",
      "``` r\n",
      "heat.colors <- brewer.pal(6, \"Blues\") # Colors from the RColorBrewer package (only 6)\n",
      "heat.colors <- colorRampPalette(heat.colors)(100) # Interpolate 100 colors\n",
      "\n",
      "pheatmap(rld_dist, \n",
      "         annotation = meta %>% column_to_rownames(\"sample\") %>% select(\"condition\"),\n",
      "         color = heat.colors)\n",
      "```\n",
      "\n",
      "## Run DESeq2:\n",
      "\n",
      "**Optional step** - Re-create DESeq2 dataset if the design formula has\n",
      "changed after QC analysis in include other sources of variation using\n",
      "\n",
      "``` r\n",
      "# dds <- DESeqDataSetFromMatrix(data, colData = meta, design = ~ covariate + condition)\n",
      "```\n",
      "\n",
      "Run DEseq2\n",
      "\n",
      "``` r\n",
      "# Run DESeq2 differential expression analysis\n",
      "dds <- DESeq(dds)\n",
      "```\n",
      "\n",
      "**Optional step** - Output normalized counts to save as a file to access\n",
      "outside RStudio using\n",
      "\n",
      "``` r\n",
      "normalized_counts <- counts(dds, normalized=TRUE)\n",
      "```\n",
      "\n",
      "## Check the fit of the dispersion estimates\n",
      "\n",
      "Plot dispersion estimates\n",
      "\n",
      "``` r\n",
      "plotDispEsts(dds)\n",
      "```\n",
      "\n",
      "## Create contrasts to perform Wald testing or the shrunken log2 fold changes between specific conditions\n",
      "\n",
      "Formal LFC calculation\n",
      "\n",
      "``` r\n",
      "# Specify contrast for comparison of interest\n",
      "contrast <- c(\"condition\", \"MOV10_overexpression\", \"control\")\n",
      "\n",
      "# Output results of Wald test for contrast\n",
      "res <- results(dds, \n",
      "               contrast = contrast, \n",
      "               alpha = 0.05)\n",
      "```\n",
      "\n",
      "Shrinkage\n",
      "\n",
      "``` r\n",
      "# Get name of the contrast you would like to use\n",
      "resultsNames(dds)\n",
      "\n",
      "# Shrink the log2 fold changes to be more accurate\n",
      "res <- lfcShrink(dds, \n",
      "                 coef = \"condition_MOV10_overexpression_vs_control\", \n",
      "                 type = \"apeglm\")\n",
      "```\n",
      "\n",
      "## Output significant results:\n",
      "\n",
      "``` r\n",
      "# Set thresholds\n",
      "padj.cutoff <- 0.05\n",
      "\n",
      "# Turn the results object into a tibble for use with tidyverse functions\n",
      "res_tbl <- res %>%\n",
      "  data.frame() %>%\n",
      "  rownames_to_column(var=\"gene\") %>% \n",
      "  as_tibble()\n",
      "\n",
      "# Subset the significant results\n",
      "sig_res <- filter(res_tbl, \n",
      "                  padj < padj.cutoff)\n",
      "```\n",
      "\n",
      "## Visualize results: volcano plots, heatmaps, normalized counts plots of top genes, etc.\n",
      "\n",
      "Function to get gene_IDs based on gene names. The function will take as\n",
      "input a vector of gene names of interest, the tx2gene dataframe and the\n",
      "dds object that you analyzed.\n",
      "\n",
      "``` r\n",
      "lookup <- function(gene_name, tx2gene, dds){\n",
      "  hits <- tx2gene %>% select(gene_symbol, gene_ID) %>% distinct() %>% \n",
      "    filter(gene_symbol %in% gene_name & gene_ID %in% rownames(dds))\n",
      "  return(hits)\n",
      "}\n",
      "\n",
      "lookup(gene_name = \"MOV10\", tx2gene = tx2gene, dds = dds)\n",
      "```\n",
      "\n",
      "Plot expression for single gene\n",
      "\n",
      "``` r\n",
      "plotCounts(dds, gene=\"ENSG00000155363\", intgroup=\"condition\")\n",
      "```\n",
      "\n",
      "### MAplot\n",
      "\n",
      "``` r\n",
      "plotMA(res)\n",
      "```\n",
      "\n",
      "### Volcano plot with labels (top N genes)\n",
      "\n",
      "``` r\n",
      "## Obtain logical vector where TRUE values denote padj values < 0.05 and fold change > 1.5 in either direction\n",
      "res_tbl <- res_tbl %>% \n",
      "mutate(threshold = padj < 0.05 & abs(log2FoldChange) >= 0.58)\n",
      "```\n",
      "\n",
      "``` r\n",
      "## Create an empty column to indicate which genes to label\n",
      "res_tbl <- res_tbl %>% mutate(genelabels = \"\")\n",
      "\n",
      "## Sort by padj values \n",
      "res_tbl <- res_tbl %>% arrange(padj)\n",
      "\n",
      "## Populate the genelabels column with contents of the gene symbols column for the first 10 rows, i.e. the top 10 most significantly expressed genes\n",
      "res_tbl$genelabels[1:10] <- as.character(res_tbl$gene[1:10])\n",
      "\n",
      "head(res_tbl)\n",
      "```\n",
      "\n",
      "``` r\n",
      "ggplot(res_tbl, aes(x = log2FoldChange, y = -log10(padj))) +\n",
      "  geom_point(aes(colour = threshold)) +\n",
      "  geom_text_repel(aes(label = genelabels)) +\n",
      "  ggtitle(\"Mov10 overexpression\") +\n",
      "  xlab(\"log2 fold change\") + \n",
      "  ylab(\"-log10 adjusted p-value\") +\n",
      "  theme(legend.position = \"none\",\n",
      "        plot.title = element_text(size = rel(1.5), hjust = 0.5),\n",
      "        axis.title = element_text(size = rel(1.25))) \n",
      "```\n",
      "\n",
      "### Heatmap of differentially expressed genes\n",
      "\n",
      "``` r\n",
      "# filter significant results from normalized counts\n",
      "norm_sig <- normalized_counts %>% as_tibble(rownames = \"gene\") %>%\n",
      "  dplyr::filter(gene %in% sig_res$gene) %>% column_to_rownames(var=\"gene\")\n",
      "```\n",
      "\n",
      "``` r\n",
      "pheatmap(norm_sig, \n",
      "         cluster_rows = T, #cluster by expression pattern\n",
      "         scale = \"row\", # scale by gene so expression pattern is visible\n",
      "         treeheight_row = 0, # dont show the row dendogram\n",
      "         show_rownames = F, # remove rownames so it is more clear\n",
      "         annotation = meta %>% column_to_rownames(var = \"sample\") %>% dplyr::select(\"condition\")\n",
      "         )\n",
      "```\n",
      "\n",
      "## Perform analysis to extract functional significance of results: GO or KEGG enrichment, GSEA, etc.\n",
      "\n",
      "### Annotate with `annotables`\n",
      "\n",
      "``` r\n",
      "ids <- grch37 %>% dplyr::filter(ensgene %in% res_tbl$gene) \n",
      "res_ids <- inner_join(res_tbl, ids, by=c(\"gene\"=\"ensgene\"))\n",
      "```\n",
      "\n",
      "### Perform enrichment analysis of GO terms (can be done as well with KEGG pathways)\n",
      "\n",
      "``` r\n",
      "# Create background dataset for hypergeometric testing using all genes tested for significance in the results\n",
      "all_genes <- dplyr::filter(res_ids, !is.na(gene)) %>% \n",
      "  pull(gene) %>% \n",
      "  as.character()\n",
      "\n",
      "# Extract significant results\n",
      "sig <- dplyr::filter(res_ids, padj < 0.05 & !is.na(gene))\n",
      "\n",
      "sig_genes <- sig %>% \n",
      "  pull(gene) %>% \n",
      "  as.character()\n",
      "```\n",
      "\n",
      "``` r\n",
      "# Perform enrichment analysis\n",
      "ego <- enrichGO(gene = sig_genes, \n",
      "                universe = all_genes,\n",
      "                keyType = \"ENSEMBL\",\n",
      "                OrgDb = org.Hs.eg.db, \n",
      "                ont = \"BP\", \n",
      "                pAdjustMethod = \"BH\", \n",
      "                qvalueCutoff = 0.05, \n",
      "                readable = TRUE)\n",
      "ego <- enrichplot::pairwise_termsim(ego)\n",
      "```\n",
      "\n",
      "### Visualize result\n",
      "\n",
      "``` r\n",
      "dotplot(ego, showCategory=50)\n",
      "```\n",
      "\n",
      "``` r\n",
      "emapplot(ego, showCategory = 50)\n",
      "```\n",
      "\n",
      "### Cnetplot\n",
      "\n",
      "``` r\n",
      "## To color genes by log2 fold changes, we need to extract the log2 fold changes from our results table creating a named vector\n",
      "sig_foldchanges <- sig$log2FoldChange\n",
      "\n",
      "names(sig_foldchanges) <- sig$gene\n",
      "```\n",
      "\n",
      "``` r\n",
      "## Cnetplot details the genes associated with one or more terms - by default gives the top 5 significant terms (by padj)\n",
      "cnetplot(ego, \n",
      "         categorySize=\"pvalue\", \n",
      "         showCategory = 5, \n",
      "         foldChange=sig_foldchanges, \n",
      "         vertex.label.font=6)\n",
      "```\n",
      "\n",
      "### Perform GSEA analysis of KEGG pathways (can be done as well with GO terms)\n",
      "\n",
      "``` r\n",
      "# Extract entrez IDs. IDs should not be duplicated or NA\n",
      "res_entrez <- dplyr::filter(res_ids, entrez != \"NA\" & entrez != \"NULL\" & duplicated(entrez)==F)\n",
      "\n",
      "## Extract the foldchanges\n",
      "foldchanges <- res_entrez$log2FoldChange\n",
      "\n",
      "## Name each fold change with the corresponding Entrez ID\n",
      "names(foldchanges) <- res_entrez$entrez\n",
      "\n",
      "## Sort fold changes in decreasing order\n",
      "foldchanges <- sort(foldchanges, decreasing = TRUE)\n",
      "```\n",
      "\n",
      "``` r\n",
      "# Run GSEA of KEGG\n",
      "gseaKEGG <- gseKEGG(geneList = foldchanges, # ordered named vector of fold changes (Entrez IDs are the associated names)\n",
      "              organism = \"hsa\", # supported organisms listed below\n",
      "              pvalueCutoff = 0.05, # padj cutoff value\n",
      "              verbose = FALSE)\n",
      "\n",
      "gseaKEGG_results <- gseaKEGG@result\n",
      "head(gseaKEGG_results)\n",
      "```\n",
      "\n",
      "``` r\n",
      "## Plot the GSEA plot for a single enriched pathway:\n",
      "gseaplot(gseaKEGG, geneSetID = gseaKEGG_results$ID[1])\n",
      "```\n",
      "\n",
      "``` r\n",
      "## Output images for a single significant KEGG pathway\n",
      "pathview(gene.data = foldchanges,\n",
      "              pathway.id = gseaKEGG_results$ID[1],\n",
      "              species = \"hsa\",\n",
      "              limit = list(gene = 2, # value gives the max/min limit for foldchanges\n",
      "              cpd = 1))\n",
      "```\n",
      "\n",
      "``` r\n",
      "knitr::include_graphics(paste0(\"./\",gseaKEGG_results$ID[1],\".png\"))\n",
      "```\n",
      "\n",
      "## Make sure to output the versions of all tools used in the DE analysis:\n",
      "\n",
      "``` r\n",
      "sessionInfo()\n",
      "```\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in files:\n",
    "    if i.endswith(\".md\"):\n",
    "        with open(\"./develop/\"+ i) as f:\n",
    "            text = f.read()\n",
    "            text = re.sub(\"\\nknit((.|\\n)*)[0-9]{4}-[0-9]{2}-[0-9]{2}\", \"---\" ,text)\n",
    "        with open(\"./develop/\"+ i, \"w\") as f:\n",
    "            f.write(text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "1eac1a36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---\n",
      "title: \"Summary of DGE workflow\"\n",
      "summary: You can use this lesson as a template for your own bulk RNAseq analysis\n",
      "date: 2023-01-17\n",
      "hide:\n",
      "  - navigation\n",
      "---\n",
      "\n",
      "# Summarised workflow\n",
      "\n",
      "We have detailed the various steps in a differential expression analysis\n",
      "workflow, providing theory with example code. To provide a more succinct\n",
      "reference for the code needed to run a DGE analysis, we have summarized\n",
      "the steps in an analysis below:\n",
      "\n",
      "## Libraries\n",
      "\n",
      "``` r\n",
      "library(tidyverse)\n",
      "library(DESeq2)\n",
      "library(ggrepel)\n",
      "library(pheatmap)\n",
      "library(annotables)\n",
      "library(clusterProfiler)\n",
      "library(DOSE)\n",
      "library(pathview)\n",
      "library(org.Hs.eg.db)\n",
      "library(tximport)\n",
      "library(RColorBrewer)\n",
      "```\n",
      "\n",
      "## Obtaining gene-level counts from your preprocessing and create DESeq object\n",
      "\n",
      "### If you have a traditional raw count matrix\n",
      "\n",
      "Load data and metadata\n",
      "\n",
      "``` r\n",
      "data <- read_table(\"../Data/Mov10_full_counts.txt\") \n",
      "\n",
      "meta <- read_table(\"../Data/Mov10_full_meta.txt\")\n",
      "```\n",
      "\n",
      "Check that the row names of the metadata equal the column names of the\n",
      "**raw counts** data\n",
      "\n",
      "``` r\n",
      "all(colnames(data)[-1] == meta$sample)\n",
      "```\n",
      "\n",
      "Create DESeq2Dataset object\n",
      "\n",
      "``` r\n",
      "dds <- DESeqDataSetFromMatrix(countData = data %>% column_to_rownames(\"GeneSymbol\"), \n",
      "                              colData = meta %>% column_to_rownames(\"sample\"), \n",
      "                              design = ~ condition)\n",
      "```\n",
      "\n",
      "### If you have pseudocounts\n",
      "\n",
      "Load samplesheet with all our metadata from our pipeline\n",
      "\n",
      "``` r\n",
      "# Load data, metadata and tx2gene and create a txi object\n",
      "meta <- read_csv(\"../Data/samplesheet.csv\")\n",
      "```\n",
      "\n",
      "Create a list of salmon results\n",
      "\n",
      "``` r\n",
      "dir <- \"/work/sequencing_data/Preprocessing_backup/results_salmon/salmon\"\n",
      "tx2gene <- read_table(file.path(dir,\"salmon_tx2gene.tsv\"), col_names = c(\"transcript_ID\",\"gene_ID\",\"gene_symbol\"))\n",
      "\n",
      "# Get all salmon results files\n",
      "files <- file.path(dir, meta$sample, \"quant.sf\")\n",
      "names(files) <- meta$sample\n",
      "```\n",
      "\n",
      "Create txi object\n",
      "\n",
      "``` r\n",
      "txi <- tximport(files, type=\"salmon\", tx2gene=tx2gene, countsFromAbundance = \"lengthScaledTPM\", ignoreTxVersion = TRUE)\n",
      "```\n",
      "\n",
      "Create dds object\n",
      "\n",
      "``` r\n",
      "dds <- DESeqDataSetFromTximport(txi,\n",
      "                                   colData = meta %>% column_to_rownames(\"sample\"), \n",
      "                              design = ~ condition)\n",
      "```\n",
      "\n",
      "## Exploratory data analysis\n",
      "\n",
      "Prefiltering low count genes + PCA & hierarchical clustering -\n",
      "identifying outliers and sources of variation in the data:\n",
      "\n",
      "### Prefiltering low count genes\n",
      "\n",
      "``` r\n",
      "keep <- rowSums(counts(dds)) > 0\n",
      "dds <- dds[keep,]\n",
      "```\n",
      "\n",
      "### Rlog transformation\n",
      "\n",
      "``` r\n",
      "# Transform counts for data visualization\n",
      "rld <- rlog(dds, \n",
      "            blind=TRUE)\n",
      "```\n",
      "\n",
      "### PCA\n",
      "\n",
      "Plot PCA\n",
      "\n",
      "``` r\n",
      "plotPCA(rld, \n",
      "        intgroup=\"condition\")\n",
      "```\n",
      "\n",
      "### Heatmaps\n",
      "\n",
      "Extract the rlog matrix from the object\n",
      "\n",
      "``` r\n",
      "rld_mat <- assay(rld)\n",
      "rld_cor <- cor(rld_mat) # Pearson correlation betweeen samples\n",
      "rld_dist <- as.matrix(dist(t(assay(rld)))) #distances are computed by rows, so we need to transponse the matrix\n",
      "```\n",
      "\n",
      "Plot heatmap of correlations\n",
      "\n",
      "``` r\n",
      "pheatmap(rld_cor, \n",
      "         annotation = meta %>% column_to_rownames(\"sample\") %>% select(\"condition\"))\n",
      "```\n",
      "\n",
      "Plot heatmap of distances with a new color range\n",
      "\n",
      "``` r\n",
      "heat.colors <- brewer.pal(6, \"Blues\") # Colors from the RColorBrewer package (only 6)\n",
      "heat.colors <- colorRampPalette(heat.colors)(100) # Interpolate 100 colors\n",
      "\n",
      "pheatmap(rld_dist, \n",
      "         annotation = meta %>% column_to_rownames(\"sample\") %>% select(\"condition\"),\n",
      "         color = heat.colors)\n",
      "```\n",
      "\n",
      "## Run DESeq2:\n",
      "\n",
      "**Optional step** - Re-create DESeq2 dataset if the design formula has\n",
      "changed after QC analysis in include other sources of variation using\n",
      "\n",
      "``` r\n",
      "# dds <- DESeqDataSetFromMatrix(data, colData = meta, design = ~ covariate + condition)\n",
      "```\n",
      "\n",
      "Run DEseq2\n",
      "\n",
      "``` r\n",
      "# Run DESeq2 differential expression analysis\n",
      "dds <- DESeq(dds)\n",
      "```\n",
      "\n",
      "**Optional step** - Output normalized counts to save as a file to access\n",
      "outside RStudio using\n",
      "\n",
      "``` r\n",
      "normalized_counts <- counts(dds, normalized=TRUE)\n",
      "```\n",
      "\n",
      "## Check the fit of the dispersion estimates\n",
      "\n",
      "Plot dispersion estimates\n",
      "\n",
      "``` r\n",
      "plotDispEsts(dds)\n",
      "```\n",
      "\n",
      "## Create contrasts to perform Wald testing or the shrunken log2 fold changes between specific conditions\n",
      "\n",
      "Formal LFC calculation\n",
      "\n",
      "``` r\n",
      "# Specify contrast for comparison of interest\n",
      "contrast <- c(\"condition\", \"MOV10_overexpression\", \"control\")\n",
      "\n",
      "# Output results of Wald test for contrast\n",
      "res <- results(dds, \n",
      "               contrast = contrast, \n",
      "               alpha = 0.05)\n",
      "```\n",
      "\n",
      "Shrinkage\n",
      "\n",
      "``` r\n",
      "# Get name of the contrast you would like to use\n",
      "resultsNames(dds)\n",
      "\n",
      "# Shrink the log2 fold changes to be more accurate\n",
      "res <- lfcShrink(dds, \n",
      "                 coef = \"condition_MOV10_overexpression_vs_control\", \n",
      "                 type = \"apeglm\")\n",
      "```\n",
      "\n",
      "## Output significant results:\n",
      "\n",
      "``` r\n",
      "# Set thresholds\n",
      "padj.cutoff <- 0.05\n",
      "\n",
      "# Turn the results object into a tibble for use with tidyverse functions\n",
      "res_tbl <- res %>%\n",
      "  data.frame() %>%\n",
      "  rownames_to_column(var=\"gene\") %>% \n",
      "  as_tibble()\n",
      "\n",
      "# Subset the significant results\n",
      "sig_res <- filter(res_tbl, \n",
      "                  padj < padj.cutoff)\n",
      "```\n",
      "\n",
      "## Visualize results: volcano plots, heatmaps, normalized counts plots of top genes, etc.\n",
      "\n",
      "Function to get gene_IDs based on gene names. The function will take as\n",
      "input a vector of gene names of interest, the tx2gene dataframe and the\n",
      "dds object that you analyzed.\n",
      "\n",
      "``` r\n",
      "lookup <- function(gene_name, tx2gene, dds){\n",
      "  hits <- tx2gene %>% select(gene_symbol, gene_ID) %>% distinct() %>% \n",
      "    filter(gene_symbol %in% gene_name & gene_ID %in% rownames(dds))\n",
      "  return(hits)\n",
      "}\n",
      "\n",
      "lookup(gene_name = \"MOV10\", tx2gene = tx2gene, dds = dds)\n",
      "```\n",
      "\n",
      "Plot expression for single gene\n",
      "\n",
      "``` r\n",
      "plotCounts(dds, gene=\"ENSG00000155363\", intgroup=\"condition\")\n",
      "```\n",
      "\n",
      "### MAplot\n",
      "\n",
      "``` r\n",
      "plotMA(res)\n",
      "```\n",
      "\n",
      "### Volcano plot with labels (top N genes)\n",
      "\n",
      "``` r\n",
      "## Obtain logical vector where TRUE values denote padj values < 0.05 and fold change > 1.5 in either direction\n",
      "res_tbl <- res_tbl %>% \n",
      "mutate(threshold = padj < 0.05 & abs(log2FoldChange) >= 0.58)\n",
      "```\n",
      "\n",
      "``` r\n",
      "## Create an empty column to indicate which genes to label\n",
      "res_tbl <- res_tbl %>% mutate(genelabels = \"\")\n",
      "\n",
      "## Sort by padj values \n",
      "res_tbl <- res_tbl %>% arrange(padj)\n",
      "\n",
      "## Populate the genelabels column with contents of the gene symbols column for the first 10 rows, i.e. the top 10 most significantly expressed genes\n",
      "res_tbl$genelabels[1:10] <- as.character(res_tbl$gene[1:10])\n",
      "\n",
      "head(res_tbl)\n",
      "```\n",
      "\n",
      "``` r\n",
      "ggplot(res_tbl, aes(x = log2FoldChange, y = -log10(padj))) +\n",
      "  geom_point(aes(colour = threshold)) +\n",
      "  geom_text_repel(aes(label = genelabels)) +\n",
      "  ggtitle(\"Mov10 overexpression\") +\n",
      "  xlab(\"log2 fold change\") + \n",
      "  ylab(\"-log10 adjusted p-value\") +\n",
      "  theme(legend.position = \"none\",\n",
      "        plot.title = element_text(size = rel(1.5), hjust = 0.5),\n",
      "        axis.title = element_text(size = rel(1.25))) \n",
      "```\n",
      "\n",
      "### Heatmap of differentially expressed genes\n",
      "\n",
      "``` r\n",
      "# filter significant results from normalized counts\n",
      "norm_sig <- normalized_counts %>% as_tibble(rownames = \"gene\") %>%\n",
      "  dplyr::filter(gene %in% sig_res$gene) %>% column_to_rownames(var=\"gene\")\n",
      "```\n",
      "\n",
      "``` r\n",
      "pheatmap(norm_sig, \n",
      "         cluster_rows = T, #cluster by expression pattern\n",
      "         scale = \"row\", # scale by gene so expression pattern is visible\n",
      "         treeheight_row = 0, # dont show the row dendogram\n",
      "         show_rownames = F, # remove rownames so it is more clear\n",
      "         annotation = meta %>% column_to_rownames(var = \"sample\") %>% dplyr::select(\"condition\")\n",
      "         )\n",
      "```\n",
      "\n",
      "## Perform analysis to extract functional significance of results: GO or KEGG enrichment, GSEA, etc.\n",
      "\n",
      "### Annotate with `annotables`\n",
      "\n",
      "``` r\n",
      "ids <- grch37 %>% dplyr::filter(ensgene %in% res_tbl$gene) \n",
      "res_ids <- inner_join(res_tbl, ids, by=c(\"gene\"=\"ensgene\"))\n",
      "```\n",
      "\n",
      "### Perform enrichment analysis of GO terms (can be done as well with KEGG pathways)\n",
      "\n",
      "``` r\n",
      "# Create background dataset for hypergeometric testing using all genes tested for significance in the results\n",
      "all_genes <- dplyr::filter(res_ids, !is.na(gene)) %>% \n",
      "  pull(gene) %>% \n",
      "  as.character()\n",
      "\n",
      "# Extract significant results\n",
      "sig <- dplyr::filter(res_ids, padj < 0.05 & !is.na(gene))\n",
      "\n",
      "sig_genes <- sig %>% \n",
      "  pull(gene) %>% \n",
      "  as.character()\n",
      "```\n",
      "\n",
      "``` r\n",
      "# Perform enrichment analysis\n",
      "ego <- enrichGO(gene = sig_genes, \n",
      "                universe = all_genes,\n",
      "                keyType = \"ENSEMBL\",\n",
      "                OrgDb = org.Hs.eg.db, \n",
      "                ont = \"BP\", \n",
      "                pAdjustMethod = \"BH\", \n",
      "                qvalueCutoff = 0.05, \n",
      "                readable = TRUE)\n",
      "ego <- enrichplot::pairwise_termsim(ego)\n",
      "```\n",
      "\n",
      "### Visualize result\n",
      "\n",
      "``` r\n",
      "dotplot(ego, showCategory=50)\n",
      "```\n",
      "\n",
      "``` r\n",
      "emapplot(ego, showCategory = 50)\n",
      "```\n",
      "\n",
      "### Cnetplot\n",
      "\n",
      "``` r\n",
      "## To color genes by log2 fold changes, we need to extract the log2 fold changes from our results table creating a named vector\n",
      "sig_foldchanges <- sig$log2FoldChange\n",
      "\n",
      "names(sig_foldchanges) <- sig$gene\n",
      "```\n",
      "\n",
      "``` r\n",
      "## Cnetplot details the genes associated with one or more terms - by default gives the top 5 significant terms (by padj)\n",
      "cnetplot(ego, \n",
      "         categorySize=\"pvalue\", \n",
      "         showCategory = 5, \n",
      "         foldChange=sig_foldchanges, \n",
      "         vertex.label.font=6)\n",
      "```\n",
      "\n",
      "### Perform GSEA analysis of KEGG pathways (can be done as well with GO terms)\n",
      "\n",
      "``` r\n",
      "# Extract entrez IDs. IDs should not be duplicated or NA\n",
      "res_entrez <- dplyr::filter(res_ids, entrez != \"NA\" & entrez != \"NULL\" & duplicated(entrez)==F)\n",
      "\n",
      "## Extract the foldchanges\n",
      "foldchanges <- res_entrez$log2FoldChange\n",
      "\n",
      "## Name each fold change with the corresponding Entrez ID\n",
      "names(foldchanges) <- res_entrez$entrez\n",
      "\n",
      "## Sort fold changes in decreasing order\n",
      "foldchanges <- sort(foldchanges, decreasing = TRUE)\n",
      "```\n",
      "\n",
      "``` r\n",
      "# Run GSEA of KEGG\n",
      "gseaKEGG <- gseKEGG(geneList = foldchanges, # ordered named vector of fold changes (Entrez IDs are the associated names)\n",
      "              organism = \"hsa\", # supported organisms listed below\n",
      "              pvalueCutoff = 0.05, # padj cutoff value\n",
      "              verbose = FALSE)\n",
      "\n",
      "gseaKEGG_results <- gseaKEGG@result\n",
      "head(gseaKEGG_results)\n",
      "```\n",
      "\n",
      "``` r\n",
      "## Plot the GSEA plot for a single enriched pathway:\n",
      "gseaplot(gseaKEGG, geneSetID = gseaKEGG_results$ID[1])\n",
      "```\n",
      "\n",
      "``` r\n",
      "## Output images for a single significant KEGG pathway\n",
      "pathview(gene.data = foldchanges,\n",
      "              pathway.id = gseaKEGG_results$ID[1],\n",
      "              species = \"hsa\",\n",
      "              limit = list(gene = 2, # value gives the max/min limit for foldchanges\n",
      "              cpd = 1))\n",
      "```\n",
      "\n",
      "``` r\n",
      "knitr::include_graphics(paste0(\"./\",gseaKEGG_results$ID[1],\".png\"))\n",
      "```\n",
      "\n",
      "## Make sure to output the versions of all tools used in the DE analysis:\n",
      "\n",
      "``` r\n",
      "sessionInfo()\n",
      "```\n",
      "\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a04c21b1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
