<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
    {
      "@context": "https://schema.org",
      "@id": "https://hds-sandbox.github.io/bulk_RNAseq_course/",
      "@type": "LearningResource",
      "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
      "description": "This workshop includes a tutorial on how to approach RNAseq data, starting from your sequencing reads (fastq files). Thus, the workshop only briefly touches upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, mainly for the purpose of outlining considerations in the downstream bioinformatic analysis. This workshop is based on the materials developed by members of the teaching team at the Harvard Chan Bioinformatics Core (HBC), a collection of modified tutorials from the DESeq2, R language vignettes and the nf-core rnaseq pipeline.",
      "keywords": [
        "bioinformatics, RNAseq, bulk, NGS"
      ],
      "learningResourceType": [
        "e-learning"
      ],
      "license": [
        {
          "@id": "https://creativecommons.org/licenses/by-sa/4.0/legalcode",
          "@type": "CreativeWork"
        }
      ],
      "mentions": [
        {
          "@id": "10.5281/zenodo.8046218",
          "@type": "Thing"
        },
        {
          "@id": "https://zenodo.org/record/7956067",
          "@type": "Thing"
        }
      ],
      "name": "Introduction to bulk RNAseq analysis workshop",
      "educationalLevel": "Beginner",
      "audience": "Researchers, PhDs, Postdocs, MSc"
    }
    </script><title>nf-core pipelines - Bulk RNAseq data analysis</title><link rel=stylesheet href=assets/stylesheets/main.6543a935.min.css><link rel=stylesheet href=assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/neoteroi-mkdocs.css><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script id=__analytics>function __md_analytics(){function n(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],n("js",new Date),n("config",""),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){this.value&&n("event","search",{search_term:this.value})}),document$.subscribe(function(){var a=document.forms.feedback;if(void 0!==a)for(var e of a.querySelectorAll("[type=submit]"))e.addEventListener("click",function(e){e.preventDefault();var t=document.location.pathname,e=this.getAttribute("data-md-value");n("event","feedback",{page:t,data:e}),a.firstElementChild.disabled=!0;e=a.querySelector(".md-feedback__note [data-md-value='"+e+"']");e&&(e.hidden=!1)}),a.hidden=!1}),location$.subscribe(function(e){n("config","",{page_path:e.pathname})})});var e=document.createElement("script");e.async=!0,e.src="https://www.googletagmanager.com/gtag/js?id=",document.getElementById("__analytics").insertAdjacentElement("afterEnd",e)}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#automating-your-workflow-nextflow-and-nf-core-pipelines class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bulk RNAseq data analysis </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> nf-core pipelines </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=02_experimental_planning.html class=md-tabs__link> Experimental planning </a> </li> <li class=md-tabs__item> <a href=03_data_explanation.html class=md-tabs__link> Data Explanation </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=04a_preprocessing.html class=md-tabs__link> Read Preprocessing </a> </li> <li class=md-tabs__item> <a href=05a_data_analysis_setup.html class=md-tabs__link> Analysing RNAseq data </a> </li> <li class=md-tabs__item> <a href=summary.html class=md-tabs__link> Workshop Summary </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> Bulk RNAseq data analysis </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> <span class=md-ellipsis> Course Introduction </span> </a> </li> <li class=md-nav__item> <a href=02_experimental_planning.html class=md-nav__link> <span class=md-ellipsis> Experimental planning </span> </a> </li> <li class=md-nav__item> <a href=03_data_explanation.html class=md-nav__link> <span class=md-ellipsis> Data Explanation </span> </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4 checked> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex> <span class=md-ellipsis> Read Preprocessing </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=true> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Read Preprocessing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=04a_preprocessing.html class=md-nav__link> <span class=md-ellipsis> Preprocessing steps </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Nextflow and nf-core pipelines </span> <span class="md-nav__icon md-icon"></span> </label> <a href=04b_pipelines.html class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Nextflow and nf-core pipelines </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#workflows-and-pipelines class=md-nav__link> <span class=md-ellipsis> Workflows and pipelines </span> </a> </li> <li class=md-nav__item> <a href=#nextflow-pipelines class=md-nav__link> <span class=md-ellipsis> Nextflow pipelines </span> </a> <nav class=md-nav aria-label="Nextflow pipelines"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#features class=md-nav__link> <span class=md-ellipsis> Features </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#the-nf-core-project class=md-nav__link> <span class=md-ellipsis> The nf-core project </span> </a> </li> <li class=md-nav__item> <a href=#nf-corernaseq-usage-for-version-3112 class=md-nav__link> <span class=md-ellipsis> nf-core/rnaseq: Usage for version 3.11.2 </span> </a> <nav class=md-nav aria-label="nf-core/rnaseq: Usage for version 3.11.2"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#running-the-pipeline class=md-nav__link> <span class=md-ellipsis> Running the pipeline </span> </a> </li> <li class=md-nav__item> <a href=#core-nextflow-arguments class=md-nav__link> <span class=md-ellipsis> Core Nextflow arguments </span> </a> <nav class=md-nav aria-label="Core Nextflow arguments"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#-work-dir class=md-nav__link> <span class=md-ellipsis> -work-dir </span> </a> </li> <li class=md-nav__item> <a href=#-profile class=md-nav__link> <span class=md-ellipsis> -profile </span> </a> </li> <li class=md-nav__item> <a href=#-name class=md-nav__link> <span class=md-ellipsis> -name </span> </a> </li> <li class=md-nav__item> <a href=#-resume class=md-nav__link> <span class=md-ellipsis> -resume </span> </a> </li> <li class=md-nav__item> <a href=#-r class=md-nav__link> <span class=md-ellipsis> -r </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#rnaseq-pipeline-arguments class=md-nav__link> <span class=md-ellipsis> RNAseq pipeline arguments </span> </a> <nav class=md-nav aria-label="RNAseq pipeline arguments"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#samplesheet-input class=md-nav__link> <span class=md-ellipsis> Samplesheet input </span> </a> </li> <li class=md-nav__item> <a href=#results-folder class=md-nav__link> <span class=md-ellipsis> Results folder </span> </a> </li> <li class=md-nav__item> <a href=#alignment-options class=md-nav__link> <span class=md-ellipsis> Alignment options </span> </a> </li> <li class=md-nav__item> <a href=#reference-genome-files class=md-nav__link> <span class=md-ellipsis> Reference genome files </span> </a> </li> <li class=md-nav__item> <a href=#process-skipping-options class=md-nav__link> <span class=md-ellipsis> Process skipping options </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#understanding-the-results-folder class=md-nav__link> <span class=md-ellipsis> Understanding the results folder </span> </a> <nav class=md-nav aria-label="Understanding the results folder"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-pipeline_info class=md-nav__link> <span class=md-ellipsis> 1. pipeline_info </span> </a> </li> <li class=md-nav__item> <a href=#2-genome class=md-nav__link> <span class=md-ellipsis> 2. genome </span> </a> </li> <li class=md-nav__item> <a href=#3-multiqc class=md-nav__link> <span class=md-ellipsis> 3. multiqc </span> </a> </li> <li class=md-nav__item> <a href=#4-fastqc class=md-nav__link> <span class=md-ellipsis> 4. fastqc </span> </a> </li> <li class=md-nav__item> <a href=#5-trimgalore class=md-nav__link> <span class=md-ellipsis> 5. trimgalore </span> </a> </li> <li class=md-nav__item> <a href=#6-aligner class=md-nav__link> <span class=md-ellipsis> 6. aligner </span> </a> </li> <li class=md-nav__item> <a href=#7-pseudoaligner class=md-nav__link> <span class=md-ellipsis> 7. pseudoaligner </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=04c_preprocessing_setup.html class=md-nav__link> <span class=md-ellipsis> UCloud setup </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex=0> <span class=md-ellipsis> Analysing RNAseq data </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Analysing RNAseq data </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=05a_data_analysis_setup.html class=md-nav__link> <span class=md-ellipsis> Ucloud setup </span> </a> </li> <li class=md-nav__item> <a href=05b_count_matrix.html class=md-nav__link> <span class=md-ellipsis> Count matrix </span> </a> </li> <li class=md-nav__item> <a href=05c_count_normalization.html class=md-nav__link> <span class=md-ellipsis> Normalization </span> </a> </li> <li class=md-nav__item> <a href=06_exploratory_analysis.html class=md-nav__link> <span class=md-ellipsis> Exploratory Analysis </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_5> <label class=md-nav__link for=__nav_5_5 id=__nav_5_5_label tabindex=0> <span class=md-ellipsis> Differential Expression </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5_5> <span class="md-nav__icon md-icon"></span> Differential Expression </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=07_extra_contrast_design.html class=md-nav__link> <span class=md-ellipsis> Design formula </span> </a> </li> <li class=md-nav__item> <a href=07a_DEA.html class=md-nav__link> <span class=md-ellipsis> DESeq2 </span> </a> </li> <li class=md-nav__item> <a href=07b_hypothesis_testing.html class=md-nav__link> <span class=md-ellipsis> Hypothesis testing </span> </a> </li> <li class=md-nav__item> <a href=07c_DEA_visualization.html class=md-nav__link> <span class=md-ellipsis> Visualization </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_6> <label class=md-nav__link for=__nav_5_6 id=__nav_5_6_label tabindex=0> <span class=md-ellipsis> Functional Analysis </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_6_label aria-expanded=false> <label class=md-nav__title for=__nav_5_6> <span class="md-nav__icon md-icon"></span> Functional Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=08a_FA_genomic_annotation.html class=md-nav__link> <span class=md-ellipsis> Gene annotation </span> </a> </li> <li class=md-nav__item> <a href=08b_FA_overrepresentation.html class=md-nav__link> <span class=md-ellipsis> Over representation analysis </span> </a> </li> <li class=md-nav__item> <a href=08c_FA_GSEA.html class=md-nav__link> <span class=md-ellipsis> Class scoring </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=09_summarized_workflow.html class=md-nav__link> <span class=md-ellipsis> Summarized workflow </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=summary.html class=md-nav__link> <span class=md-ellipsis> Workshop Summary </span> </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#workflows-and-pipelines class=md-nav__link> <span class=md-ellipsis> Workflows and pipelines </span> </a> </li> <li class=md-nav__item> <a href=#nextflow-pipelines class=md-nav__link> <span class=md-ellipsis> Nextflow pipelines </span> </a> <nav class=md-nav aria-label="Nextflow pipelines"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#features class=md-nav__link> <span class=md-ellipsis> Features </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#the-nf-core-project class=md-nav__link> <span class=md-ellipsis> The nf-core project </span> </a> </li> <li class=md-nav__item> <a href=#nf-corernaseq-usage-for-version-3112 class=md-nav__link> <span class=md-ellipsis> nf-core/rnaseq: Usage for version 3.11.2 </span> </a> <nav class=md-nav aria-label="nf-core/rnaseq: Usage for version 3.11.2"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#running-the-pipeline class=md-nav__link> <span class=md-ellipsis> Running the pipeline </span> </a> </li> <li class=md-nav__item> <a href=#core-nextflow-arguments class=md-nav__link> <span class=md-ellipsis> Core Nextflow arguments </span> </a> <nav class=md-nav aria-label="Core Nextflow arguments"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#-work-dir class=md-nav__link> <span class=md-ellipsis> -work-dir </span> </a> </li> <li class=md-nav__item> <a href=#-profile class=md-nav__link> <span class=md-ellipsis> -profile </span> </a> </li> <li class=md-nav__item> <a href=#-name class=md-nav__link> <span class=md-ellipsis> -name </span> </a> </li> <li class=md-nav__item> <a href=#-resume class=md-nav__link> <span class=md-ellipsis> -resume </span> </a> </li> <li class=md-nav__item> <a href=#-r class=md-nav__link> <span class=md-ellipsis> -r </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#rnaseq-pipeline-arguments class=md-nav__link> <span class=md-ellipsis> RNAseq pipeline arguments </span> </a> <nav class=md-nav aria-label="RNAseq pipeline arguments"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#samplesheet-input class=md-nav__link> <span class=md-ellipsis> Samplesheet input </span> </a> </li> <li class=md-nav__item> <a href=#results-folder class=md-nav__link> <span class=md-ellipsis> Results folder </span> </a> </li> <li class=md-nav__item> <a href=#alignment-options class=md-nav__link> <span class=md-ellipsis> Alignment options </span> </a> </li> <li class=md-nav__item> <a href=#reference-genome-files class=md-nav__link> <span class=md-ellipsis> Reference genome files </span> </a> </li> <li class=md-nav__item> <a href=#process-skipping-options class=md-nav__link> <span class=md-ellipsis> Process skipping options </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#understanding-the-results-folder class=md-nav__link> <span class=md-ellipsis> Understanding the results folder </span> </a> <nav class=md-nav aria-label="Understanding the results folder"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-pipeline_info class=md-nav__link> <span class=md-ellipsis> 1. pipeline_info </span> </a> </li> <li class=md-nav__item> <a href=#2-genome class=md-nav__link> <span class=md-ellipsis> 2. genome </span> </a> </li> <li class=md-nav__item> <a href=#3-multiqc class=md-nav__link> <span class=md-ellipsis> 3. multiqc </span> </a> </li> <li class=md-nav__item> <a href=#4-fastqc class=md-nav__link> <span class=md-ellipsis> 4. fastqc </span> </a> </li> <li class=md-nav__item> <a href=#5-trimgalore class=md-nav__link> <span class=md-ellipsis> 5. trimgalore </span> </a> </li> <li class=md-nav__item> <a href=#6-aligner class=md-nav__link> <span class=md-ellipsis> 6. aligner </span> </a> </li> <li class=md-nav__item> <a href=#7-pseudoaligner class=md-nav__link> <span class=md-ellipsis> 7. pseudoaligner </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <div><h1 id=automating-your-workflow-nextflow-and-nf-core-pipelines>Automating your workflow: Nextflow and nf-core pipelines<a class=headerlink href=#automating-your-workflow-nextflow-and-nf-core-pipelines title="Permanent link">¶</a></h1> <p><strong>Last updated:</strong> <em><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">May 21, 2024</span></em></p> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>🕰 <strong>Time Estimation:</strong> X minutes </p> <p>💬 <strong>Learning Objectives:</strong> </p> <ol> <li>Understand what is a pipeline/workflow and the Nextflow language.</li> <li>Learn about existing automated workflows from the bioinformatics community.</li> <li>Learn how to use the nf-core pipeline for bulk RNAseq analysis.</li> </ol> </div> <h2 id=workflows-and-pipelines>Workflows and pipelines<a class=headerlink href=#workflows-and-pipelines title="Permanent link">¶</a></h2> <p>A pipeline consists of a chain of processing elements (processes, threads, coroutines, functions, etc.), arranged so that the output of each element is the input of the next; the name is by analogy to a physical pipeline. Narrowly speaking, a pipeline is linear and one-directional, though sometimes the term is applied to more general flows. For example, a primarily one-directional pipeline may have some communication in the other direction, known as a return channel or backchannel, or a pipeline may be fully bi-directional. Flows with one-directional tree and directed acyclic graph topologies behave similarly to (linear) pipelines – the lack of cycles makes them simple – and thus may be loosely referred to as "pipelines".</p> <p>In our case, a "preprocessing" pipeline consists on concatenating all the processes explained in the previous lesson so that we have one continuous workflow from raw sequencing reads to a count matrix. For example, the RNASeq pipeline developed by the nf-core community (<a href=#the-nf-core-project>see below</a>). </p> <p align=center> <img src=./img/04b_pipelines/nf-core-rnaseq_metro_map_grey.png> <figcaption>nf-core/rnaseq pipeline metro schematic.</figcaption> </p> <p>As you can see in the image above, each process, such as read trimming, QC, alignment, etc, are connected to each other. A workflow created in this way is ideal to reproduce analysis and makes the task of analysing RNAseq data much much easier. </p> <h2 id=nextflow-pipelines>Nextflow pipelines<a class=headerlink href=#nextflow-pipelines title="Permanent link">¶</a></h2> <p align=center> <img src=./img/04b_pipelines/nextflow2014_no-bg.png%20> </p> <p><a href=https://www.nextflow.io/ >Nextflow</a> is a bioinformatics workflow manager that enables the development of portable and reproducible workflows. It supports deploying workflows on a variety of execution platforms including local, HPC schedulers, AWS Batch, Google Cloud Life Sciences, and Kubernetes. Additionally, it provides support for manage your workflow dependencies through built-in support for Conda, Spack, Docker, Podman, Singularity, Modules, and more.</p> <p>With the rise of big data, techniques to analyse and run experiments on large datasets are increasingly necessary. Parallelization and distributed computing are the best ways to tackle this problem, but the tools commonly available to the bioinformatics community often lack good support for these techniques, or provide a model that fits badly with the specific requirements in the bioinformatics domain and, most of the time, require the knowledge of complex tools or low-level APIs.</p> <p>Nextflow framework is based on the dataflow programming model, which greatly simplifies writing parallel and distributed pipelines without adding unnecessary complexity and letting you concentrate on the flow of data, i.e. the functional logic of the application/algorithm.</p> <p>It doesn't aim to be another pipeline scripting language yet, but it is built around the idea that the Linux platform is the lingua franca of data science, since it provides many simple command line and scripting tools, which by themselves are powerful, but when chained together facilitate complex data manipulations.</p> <p>In practice, this means that a Nextflow script is defined by composing many different processes. Each process can execute a given bioinformatics tool or scripting language, to which is added the ability to coordinate and synchronize the processes execution by simply specifying their inputs and outputs.</p> <h3 id=features>Features<a class=headerlink href=#features title="Permanent link">¶</a></h3> <ul> <li><strong>Fast prototyping</strong>: Nextflow allows you to write a computational pipeline by making it simpler to put together many different tasks. You may reuse your existing scripts and tools and you don't need to learn a new language or API to start using it.</li> <li><strong>Reproducibility</strong>: Nextflow supports Docker and Singularity containers technology. This, along with the integration of the GitHub code sharing platform, allows you to write self-contained pipelines, manage versions and to rapidly reproduce any former configuration.</li> <li><strong>Portable</strong>: Nextflow provides an abstraction layer between your pipeline's logic and the execution layer, so that it can be executed on multiple platforms without it changing. It provides out of the box executors for GridEngine, SLURM, LSF, PBS, Moab and HTCondor batch schedulers and for Kubernetes, Amazon AWS, Google Cloud and Microsoft Azure platforms.</li> <li><strong>Unified parallelism</strong>: Nextflow is based on the dataflow programming model which greatly simplifies writing complex distributed pipelines. Parallelisation is implicitly defined by the processes input and output declarations. The resulting applications are inherently parallel and can scale-up or scale-out, transparently, without having to adapt to a specific platform architecture.</li> <li><strong>Continuous checkpoints</strong>: All the intermediate results produced during the pipeline execution are automatically tracked. This allows you to resume its execution, from the last successfully executed step, no matter what the reason was for it stopping.</li> </ul> <h2 id=the-nf-core-project>The nf-core project<a class=headerlink href=#the-nf-core-project title="Permanent link">¶</a></h2> <p align=center> <img src=./img/04b_pipelines/nf-core-logo.svg> </p> <p>The <a href=https://nf-co.re/ >nf-core project</a> is a community effort to collect a curated set of analysis pipelines built using <a href=https://www.nextflow.io/ >Nextflow</a>, an incredibly powerful and flexible workflow language. This means that all the tools and steps used in your RNAseq workflow can be automated and easily reproduced by other researchers if necessary. In addition, if you use any of the nf-core pipelines, you will be sure that all the necessary tools are available to you in any computer platform (Cloud computing, HPC or your personal computer).</p> <p align=center> <img src=./img/04b_pipelines/nf-core-rnaseq_metro_map_grey.png> <figcaption>nf-core/rnaseq pipeline metro schematic.</figcaption> </p> <p>The <a href=https://nf-co.re/rnaseq>RNAseq pipeline</a> enables using many different tools, such as STAR, RSEM, HISAT2 or Salmon, and allows quantification of gene/isoform counts and provides extensive quality control checks at each step of the workflow. We encourage your to take a look at the pipeline and its documentation if you need to preprocess your RNAseq reads from stratch, as well as checkout this introductory video.</p> <p><div class=video-container><iframe src=https://youtube.com/embed/qMuUt8oVhHw style=position:relative;width:100%;height:22.172vw frameborder=0 allowfullscreen alt=type:video></iframe></div></p> <h2 id=nf-corernaseq-usage-for-version-3112>nf-core/rnaseq: Usage for version 3.11.2<a class=headerlink href=#nf-corernaseq-usage-for-version-3112 title="Permanent link">¶</a></h2> <p>In this section, we will see some of the most important arguments that the pipeline uses to run a RNAseq preprocessing workflow. There are many more options for advanced users and we really encourage you to check them out at the <a href=https://nf-co.re/rnaseq/3.11.2/usage>official webpage</a>. Below you will find the arguments we will use for our own data.</p> <h3 id=running-the-pipeline>Running the pipeline<a class=headerlink href=#running-the-pipeline title="Permanent link">¶</a></h3> <p>The typical command for running the pipeline is as follows:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=go>nextflow run nf-core/rnaseq --input samplesheet.csv --outdir &lt;OUTDIR&gt; --genome GRCh37 -profile docker</span>
</code></pre></div> <p>This will launch the pipeline with the <code>docker</code> configuration profile. See below for more information about profiles.</p> <p>Note that the pipeline will create the following files in your working directory:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=go>work            # Directory containing the nextflow working files</span>
<a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a><span class=go>results         # Finished results (configurable, see below)</span>
<a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a><span class=go>.nextflow_log   # Log file from Nextflow</span>
<a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a><span class=gp># </span>Other<span class=w> </span>nextflow<span class=w> </span>hidden<span class=w> </span>files,<span class=w> </span>eg.<span class=w> </span><span class=nb>history</span><span class=w> </span>of<span class=w> </span>pipeline<span class=w> </span>runs<span class=w> </span>and<span class=w> </span>old<span class=w> </span>logs.
</code></pre></div> <div class="admonition info"> <p class=admonition-title>Info</p> <ul> <li>Options with a <em>single</em> hyphen are part of Nextflo, i.e. <code>-profile</code>.</li> <li>Pipeline specific parameters use a <em>double-hyphen</em>, i.e. <code>--input</code>.</li> </ul> </div> <h3 id=core-nextflow-arguments>Core Nextflow arguments<a class=headerlink href=#core-nextflow-arguments title="Permanent link">¶</a></h3> <h4 id=-work-dir><code>-work-dir</code><a class=headerlink href=#-work-dir title="Permanent link">¶</a></h4> <p>Use this parameter to choose a path where the <code>work</code> folder, which containing all the intermediary files from the pipeline, should be saved.</p> <h4 id=-profile><code>-profile</code><a class=headerlink href=#-profile title="Permanent link">¶</a></h4> <p>Use this parameter to choose a configuration profile. Profiles can give configuration presets for different compute environments.</p> <p>Several generic profiles are bundled with the pipeline which instruct the pipeline to use software packaged using different methods (Docker, Singularity, Podman, Shifter, Charliecloud, Conda) - see below. When using Biocontainers, most of these software packaging methods pull Docker containers from quay.io e.g <a href=https://quay.io/repository/biocontainers/fastqc>FastQC</a> except for Singularity which directly downloads Singularity images via https hosted by the <a href=https://depot.galaxyproject.org/singularity/ >Galaxy project</a> and Conda which downloads and installs software locally from <a href=https://bioconda.github.io/ >Bioconda</a>.</p> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>It is highly recommended to use Docker or Singularity containers for full pipeline reproducibility, however when this is not possible, Conda is also supported.</p> </div> <p>The pipeline also dynamically loads configurations from <a href=https://github.com/nf-core/configs>https://github.com/nf-core/configs</a> when it runs, making multiple config profiles for various institutional clusters available at run time. For more information and to see if your system is available in these configs please see the <a href=https://github.com/nf-core/configs#documentation>nf-core/configs documentation</a>.</p> <p>Note that multiple profiles can be loaded, for example: <code>-profile test,docker</code> - the order of arguments is important! They are loaded in sequence, so later profiles can overwrite earlier profiles.</p> <p>If <code>-profile</code> is not specified, the pipeline will run locally and expect all software to be installed and available on the <code>PATH</code>. This is <em>not</em> recommended.</p> <ul> <li><code>docker</code><ul> <li>A generic configuration profile to be used with <a href=https://docker.com/ >Docker</a></li> </ul> </li> <li><code>singularity</code><ul> <li>A generic configuration profile to be used with <a href=https://sylabs.io/docs/ >Singularity</a></li> </ul> </li> <li><code>podman</code><ul> <li>A generic configuration profile to be used with <a href=https://podman.io/ >Podman</a></li> </ul> </li> <li><code>shifter</code><ul> <li>A generic configuration profile to be used with <a href=https://nersc.gitlab.io/development/shifter/how-to-use/ >Shifter</a></li> </ul> </li> <li><code>charliecloud</code><ul> <li>A generic configuration profile to be used with <a href=https://hpc.github.io/charliecloud/ >Charliecloud</a></li> </ul> </li> <li><code>conda</code><ul> <li>A generic configuration profile to be used with <a href=https://conda.io/docs/ >Conda</a>. Please only use Conda as a last resort i.e. when it's not possible to run the pipeline with Docker, Singularity, Podman, Shifter or Charliecloud.</li> </ul> </li> <li><code>test</code><ul> <li>A profile with a complete configuration for automated testing</li> <li>Includes links to test data so needs no other parameters</li> </ul> </li> </ul> <h4 id=-name><code>-name</code><a class=headerlink href=#-name title="Permanent link">¶</a></h4> <p>Use this parameter to give a unique name to the run. This name cannot be used again for another run in the same folder. This is very useful to track different runs since otherwise Nextflow will assign a random unique name to the run.</p> <h4 id=-resume><code>-resume</code><a class=headerlink href=#-resume title="Permanent link">¶</a></h4> <p>Specify this when restarting a pipeline. Nextflow will used cached results from any pipeline steps where the inputs are the same, continuing from where it got to previously.</p> <p>You can also supply a run name to resume a specific run: <code>-resume [run-name]</code>. Use the <code>nextflow log</code> command to show previous run names.</p> <h4 id=-r><code>-r</code><a class=headerlink href=#-r title="Permanent link">¶</a></h4> <p>It is a good idea to specify a pipeline version when running the pipeline on your data. This ensures that a specific version of the pipeline code and software are used when you run your pipeline. If you keep using the same tag, you'll be running the same version of the pipeline, even if there have been changes to the code since.</p> <p>First, go to the <a href=https://github.com/nf-core/rnaseq/releases>nf-core/rnaseq releases page</a> and find the latest version number (numeric only), e.g., <code>1.3.1</code>. Then specify this when running the pipeline with <code>-r</code> (one hyphen), e.g., <code>-r 1.3.1</code>.</p> <p>This version number will be logged in reports when you run the pipeline, so that you'll know what you used when you look back in the future.</p> <h3 id=rnaseq-pipeline-arguments>RNAseq pipeline arguments<a class=headerlink href=#rnaseq-pipeline-arguments title="Permanent link">¶</a></h3> <h4 id=samplesheet-input>Samplesheet input<a class=headerlink href=#samplesheet-input title="Permanent link">¶</a></h4> <p>You will need to create a samplesheet with information about the samples you would like to analyse before running the pipeline. Use this parameter to specify its location. It has to be a comma-separated file with 4 columns, and a header row as shown in the examples below.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=go>--input '[path to samplesheet file]'</span>
</code></pre></div> <p><strong>Multiple runs of the same sample</strong></p> <p>The <code>sample</code> identifiers have to be the same when you have re-sequenced the same sample more than once e.g. to increase sequencing depth. The pipeline will concatenate the raw reads before performing any downstream analysis. Below is an example for the same sample sequenced across 3 lanes:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=go>sample,fastq_1,fastq_2,strandedness</span>
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=go>CONTROL_REP1,AEG588A1_S1_L002_R1_001.fastq.gz,AEG588A1_S1_L002_R2_001.fastq.gz,unstranded</span>
<a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a><span class=go>CONTROL_REP1,AEG588A1_S1_L003_R1_001.fastq.gz,AEG588A1_S1_L003_R2_001.fastq.gz,unstranded</span>
<a id=__codelineno-3-4 name=__codelineno-3-4 href=#__codelineno-3-4></a><span class=go>CONTROL_REP1,AEG588A1_S1_L004_R1_001.fastq.gz,AEG588A1_S1_L004_R2_001.fastq.gz,unstranded</span>
</code></pre></div> <p><strong>Full samplesheet</strong></p> <p>The pipeline will auto-detect whether a sample is single- or paired-end using the information provided in the samplesheet. The samplesheet can have as many columns as you desire, however, there is a strict requirement for the first 4 columns to match those defined in the table below.</p> <p>A final samplesheet file consisting of both single- and paired-end data may look something like the one below. This is for 6 samples, where <code>TREATMENT_REP3</code> has been sequenced twice.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=go>sample,fastq_1,fastq_2,strandedness</span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=go>CONTROL_REP1,AEG588A1_S1_L002_R1_001.fastq.gz,AEG588A1_S1_L002_R2_001.fastq.gz,forward</span>
<a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a><span class=go>CONTROL_REP2,AEG588A2_S2_L002_R1_001.fastq.gz,AEG588A2_S2_L002_R2_001.fastq.gz,forward</span>
<a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a><span class=go>CONTROL_REP3,AEG588A3_S3_L002_R1_001.fastq.gz,AEG588A3_S3_L002_R2_001.fastq.gz,forward</span>
<a id=__codelineno-4-5 name=__codelineno-4-5 href=#__codelineno-4-5></a><span class=go>TREATMENT_REP1,AEG588A4_S4_L003_R1_001.fastq.gz,,reverse</span>
<a id=__codelineno-4-6 name=__codelineno-4-6 href=#__codelineno-4-6></a><span class=go>TREATMENT_REP2,AEG588A5_S5_L003_R1_001.fastq.gz,,reverse</span>
<a id=__codelineno-4-7 name=__codelineno-4-7 href=#__codelineno-4-7></a><span class=go>TREATMENT_REP3,AEG588A6_S6_L003_R1_001.fastq.gz,,reverse</span>
<a id=__codelineno-4-8 name=__codelineno-4-8 href=#__codelineno-4-8></a><span class=go>TREATMENT_REP3,AEG588A6_S6_L004_R1_001.fastq.gz,,reverse</span>
</code></pre></div> <table> <thead> <tr> <th>Column</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>sample</code></td> <td>Custom sample name. This entry will be identical for multiple sequencing libraries/runs from the same sample. Spaces in sample names are automatically converted to underscores (<code>_</code>).</td> </tr> <tr> <td><code>fastq_1</code></td> <td>Full path to FastQ file for Illumina short reads 1. File has to be gzipped and have the extension ".fastq.gz" or ".fq.gz".</td> </tr> <tr> <td><code>fastq_2</code></td> <td>Full path to FastQ file for Illumina short reads 2. File has to be gzipped and have the extension ".fastq.gz" or ".fq.gz".</td> </tr> <tr> <td><code>strandedness</code></td> <td>Sample strand-specificity. Must be one of <code>unstranded</code>, <code>forward</code> or <code>reverse</code>.</td> </tr> </tbody> </table> <div class="admonition info"> <p class=admonition-title>Info</p> <p>The <code>group</code> and <code>replicate</code> columns were replaced with a single <code>sample</code> column as of v3.1 of the pipeline. The <code>sample</code> column is essentially a concatenation of the <code>group</code> and <code>replicate</code> columns, however it now also offers more flexibility in instances where replicate information is not required e.g. when sequencing clinical samples. If all values of <code>sample</code> have the same number of underscores, fields defined by these underscore-separated names may be used in the PCA plots produced by the pipeline, to regain the ability to represent different groupings.</p> </div> <h4 id=results-folder>Results folder<a class=headerlink href=#results-folder title="Permanent link">¶</a></h4> <p>The output directory where the results of the pipeline will be saved.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=go>--outdir '[path to output]'</span>
</code></pre></div> <h4 id=alignment-options>Alignment options<a class=headerlink href=#alignment-options title="Permanent link">¶</a></h4> <p>By default, the pipeline uses <a href=https://github.com/alexdobin/STAR>STAR</a> (i.e. <code>--aligner star_salmon</code>) to map the raw FastQ reads to the reference genome, project the alignments onto the transcriptome and to perform the downstream BAM-level quantification with <a href=https://salmon.readthedocs.io/en/latest/salmon.html>Salmon</a>. STAR is fast but requires a lot of memory to run, typically around 38GB for the Human GRCh37 reference genome. Since the <a href=https://github.com/deweylab/RSEM>RSEM</a> (i.e. <code>--aligner star_rsem</code>) workflow in the pipeline also uses STAR you should use the <a href=https://ccb.jhu.edu/software/hisat2/index.shtml>HISAT2</a> aligner (i.e. <code>--aligner hisat2</code>) if you have memory limitations.</p> <p>You also have the option to pseudo-align and quantify your data with <a href=https://salmon.readthedocs.io/en/latest/salmon.html>Salmon</a> by providing the <code>--pseudo_aligner salmon</code> parameter. Salmon will then be run in addition to the standard alignment workflow defined by <code>--aligner</code>, mainly because it allows you to obtain QC metrics with respect to the genomic alignments. However, you can provide the <code>--skip_alignment</code> parameter if you would like to run Salmon in isolation.</p> <p>Two additional parameters <code>--extra_star_align_args</code> and <code>--extra_salmon_quant_args</code> were added in v3.10 of the pipeline that allow you to append any custom parameters to the STAR align and Salmon quant commands, respectively. Note, the <code>--seqBias</code> and <code>--gcBias</code> are not provided to Salmon quant by default so you can provide these via <code>--extra_salmon_quant_args '--seqBias --gcBias'</code> if required.</p> <h4 id=reference-genome-files>Reference genome files<a class=headerlink href=#reference-genome-files title="Permanent link">¶</a></h4> <p>The minimum reference genome requirements are a FASTA and GTF file, all other files required to run the pipeline can be generated from these files. However, it is more storage and compute friendly if you are able to re-use reference genome files as efficiently as possible. It is recommended to use the <code>--save_reference</code> parameter if you are using the pipeline to build new indices (e.g. those unavailable on <a href=https://nf-co.re/usage/reference_genomes>AWS iGenomes</a>) so that you can save them somewhere locally. The index building step can be quite a time-consuming process and it permits their reuse for future runs of the pipeline to save disk space. You can then either provide the appropriate reference genome files on the command-line via the appropriate parameters (e.g. <code>--star_index '/path/to/STAR/index/'</code>) or via a custom config file.</p> <ul> <li>If <code>--genome</code> is provided then the FASTA and GTF files (and existing indices) will be automatically obtained from AWS-iGenomes unless these have already been downloaded locally in the path specified by <code>--igenomes_base</code>.</li> <li>If <code>--gff</code> is provided as input then this will be converted to a GTF file, or the latter will be used if both are provided.</li> <li>If <code>--gene_bed</code> is not provided then it will be generated from the GTF file.</li> <li>If <code>--additional_fasta</code> is provided then the features in this file (e.g. ERCC spike-ins) will be automatically concatenated onto both the reference FASTA file as well as the GTF annotation before building the appropriate indices.</li> </ul> <p>When using <code>--aligner star_rsem</code>, both the STAR and RSEM indices should be present in the path specified by <code>--rsem_index</code> (see <a href=https://github.com/nf-core/rnaseq/issues/568>#568</a>)</p> <div class="admonition note"> <p class=admonition-title>Note</p> <p>Compressed reference files are also supported by the pipeline i.e. standard files with the <code>.gz</code> extension and indices folders with the <code>tar.gz</code> extension.</p> </div> <h4 id=process-skipping-options>Process skipping options<a class=headerlink href=#process-skipping-options title="Permanent link">¶</a></h4> <p>There are several options to skip various steps within the workflow.</p> <ul> <li><code>--skip_bigwig</code>: Skip bigWig file creation</li> <li><code>--skip_stringtie</code>: Skip StringTie.</li> <li><code>--skip_fastqc</code>: Skip FastQC.</li> <li><code>--skip_preseq</code>: Skip Preseq.</li> <li><code>--skip_dupradar</code>: Skip dupRadar.</li> <li><code>--skip_qualimap</code>: Skip Qualimap.</li> <li><code>--skip_rseqc</code>: Skip RSeQC.</li> <li><code>--skip_biotype_qc</code>: Skip additional featureCounts process for biotype QC.</li> <li><code>--skip_deseq2_qc</code>: Skip DESeq2 PCA and heatmap plotting.</li> <li><code>--skip_multiqc</code>: Skip MultiQC.</li> <li><code>--skip_qc</code>: Skip all QC steps except for MultiQC.</li> </ul> <h3 id=understanding-the-results-folder>Understanding the results folder<a class=headerlink href=#understanding-the-results-folder title="Permanent link">¶</a></h3> <p>The pipeline will save everything inside the <code>--outdir</code> folder. Inside it you will find different results. In this section we will go through the most relevant results for this workshop. If you are interested in the full documentation, visit the <a href=https://nf-co.re/rnaseq/3.6/output>nf-core rnaseq output docs</a>.</p> <h4 id=1-pipeline_info>1. <code>pipeline_info</code><a class=headerlink href=#1-pipeline_info title="Permanent link">¶</a></h4> <p>First, we will check the <code>pipeline_info</code> folder. <a href=https://www.nextflow.io/docs/latest/tracing.html>Nextflow</a> provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.</p> <p><code>pipeline_info/</code></p> <ul> <li>Reports generated by Nextflow: <code>execution_report.html</code>, <code>execution_timeline.html</code>, <code>execution_trace.txt</code> and <code>pipeline_dag.dot</code>/<code>pipeline_dag.svg</code>.</li> <li>Reports generated by the pipeline: <code>pipeline_report.html</code>, <code>pipeline_report.txt</code> and <code>software_versions.yml</code>. The <code>pipeline_report*</code> files will only be present if the <code>--email</code> / <code>--email_on_fail</code> parameter's are used when running the pipeline.</li> <li>Reformatted samplesheet files used as input to the pipeline: <code>samplesheet.valid.csv</code>.</li> </ul> <h4 id=2-genome>2. <code>genome</code><a class=headerlink href=#2-genome title="Permanent link">¶</a></h4> <p>A number of genome-specific files are generated by the pipeline because they are required for the downstream processing of the results. If the <code>--save_reference</code> parameter is provided then these will be saved in the <code>genome/</code> directory. It is recommended to use the <code>--save_reference</code> parameter if you are using the pipeline to build new indices so that you can save them somewhere locally. The index building step can be quite a time-consuming process and it permits their reuse for future runs of the pipeline to save disk space.</p> <div class="admonition info"> <p class=admonition-title>Info</p> <p>If you have not selected <code>--save_reference</code>, you will instead get a README.txt file containing information about the reference that was used for the run.</p> </div> <p><code>genome/</code></p> <ul> <li><code>*.fa</code>, <code>*.gtf</code>, <code>*.gff</code>, <code>*.bed</code>, <code>.tsv</code>: If the <code>--save_reference</code> parameter is provided then all of the genome reference files will be placed in this directory</li> </ul> <p><code>genome/index/</code></p> <ul> <li><code>star/</code>: Directory containing STAR indices.</li> <li><code>hisat2/</code>: Directory containing HISAT2 indices.</li> <li><code>rsem/</code>: Directory containing STAR and RSEM indices.</li> <li><code>salmon/</code>: Directory containing Salmon indices.</li> </ul> <h4 id=3-multiqc>3. <code>multiqc</code><a class=headerlink href=#3-multiqc title="Permanent link">¶</a></h4> <p>Results generated by MultiQC collate pipeline QC from supported tools i.e. FastQC, Cutadapt, SortMeRNA, STAR, RSEM, HISAT2, Salmon, SAMtools, Picard, RSeQC, Qualimap, Preseq and featureCounts. Additionally, various custom content has been added to the report to assess the output of dupRadar, DESeq2 and featureCounts biotypes, and to highlight samples failing a mimimum mapping threshold or those that failed to match the strand-specificity provided in the input samplesheet. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability.</p> <p><code>multiqc/</code></p> <ul> <li><code>multiqc_report.html</code>: a standalone HTML file that can be viewed in your web browser.</li> <li><code>multiqc_data/</code>: directory containing parsed statistics from the different tools used in the pipeline.</li> <li><code>multiqc_plots/</code>: directory containing individual plots from the different tools used in the pipeline.</li> </ul> <h4 id=4-fastqc>4. <code>fastqc</code><a class=headerlink href=#4-fastqc title="Permanent link">¶</a></h4> <p>The FastQC plots in this directory are generated relative to the raw, input reads. They may contain adapter sequence and regions of low quality.</p> <p><code>fastqc/</code></p> <ul> <li><code>*_fastqc.html</code>: FastQC report containing quality metrics.</li> <li><code>*_fastqc.zip</code>: Zip archive containing the FastQC report, tab-delimited data file and plot images.</li> </ul> <h4 id=5-trimgalore>5. <code>trimgalore</code><a class=headerlink href=#5-trimgalore title="Permanent link">¶</a></h4> <p>In this folder you will find your trimmed and filtered fastq files from TrimGalore, including its fastqc results!</p> <p><code>trimgalore/</code></p> <ul> <li><code>*.fq.gz</code>: If <code>--save_trimmed</code> is specified, FastQ files <strong>after</strong> adapter trimming will be placed in this directory.</li> <li><code>*_trimming_report.txt</code>: Log file generated by Trim Galore!.</li> </ul> <p><code>trimgalore/fastqc/</code></p> <ul> <li><code>*_fastqc.html</code>: FastQC report containing quality metrics for read 1 (<em>and read2 if paired-end</em>) <strong>after</strong> adapter trimming.</li> <li><code>*_fastqc.zip</code>: Zip archive containing the FastQC report, tab-delimited data file and plot images.</li> </ul> <h4 id=6-aligner>6. <code>aligner</code><a class=headerlink href=#6-aligner title="Permanent link">¶</a></h4> <p>Depending on the aligner that you use, this folder may change its contents and name. Generally, here you can find the aligned reads in <code>.bam</code> format, samtool stats, duplicated stats and different QC tools related to mapped reads.</p> <p>Also, depending on the <code>--aligner</code> option you can either find the quantification results from either salmon (pseudoquantification) or rsem (traditional quantification, i.e., a count matrix)</p> <p><strong>From star_salmon</strong></p> <p><code>star_salmon/</code></p> <ul> <li><code>*.Aligned.out.bam</code>: If <code>--save_align_intermeds</code> is specified the original BAM file containing read alignments to the reference genome will be placed in this directory.</li> </ul> <p><code>star_salmon/log/</code></p> <ul> <li><code>*.SJ.out.tab</code>: File containing filtered splice junctions detected after mapping the reads.</li> <li><code>*.Log.final.out</code>: STAR alignment report containing the mapping results summary.</li> <li><code>*.Log.out</code> and <code>*.Log.progress.out</code>: STAR log files containing detailed information about the run. Typically only useful for debugging purposes.</li> </ul> <p><code>star_salmon/salmon/</code></p> <ul> <li><code>salmon.merged.gene_counts.tsv</code>: Matrix of gene-level raw counts across all samples.</li> <li><code>salmon.merged.gene_tpm.tsv</code>: Matrix of gene-level TPM values across all samples.</li> <li><code>salmon.merged.gene_counts.rds</code>: RDS object that can be loaded in R that contains a <a href=https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html>SummarizedExperiment</a> container with the TPM (<code>abundance</code>), estimated counts (<code>counts</code>) and transcript length (<code>length</code>) in the assays slot for genes.</li> <li><code>salmon.merged.gene_counts_scaled.tsv</code>: Matrix of gene-level scaled counts across all samples.</li> <li><code>salmon.merged.gene_counts_scaled.rds</code>: RDS object that can be loaded in R that contains a <a href=https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html>SummarizedExperiment</a> container with the TPM (<code>abundance</code>), estimated counts (<code>counts</code>) and transcript length (<code>length</code>) in the assays slot for genes.</li> <li><code>salmon.merged.gene_counts_length_scaled.tsv</code>: Matrix of gene-level length-scaled counts across all samples.</li> <li><code>salmon.merged.gene_counts_length_scaled.rds</code>: RDS object that can be loaded in R that contains a <a href=https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html>SummarizedExperiment</a> container with the TPM (<code>abundance</code>), estimated counts (<code>counts</code>) and transcript length (<code>length</code>) in the assays slot for genes.</li> <li><code>salmon.merged.transcript_counts.tsv</code>: Matrix of isoform-level raw counts across all samples.</li> <li><code>salmon.merged.transcript_tpm.tsv</code>: Matrix of isoform-level TPM values across all samples.</li> <li><code>salmon.merged.transcript_counts.rds</code>: RDS object that can be loaded in R that contains a <a href=https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html>SummarizedExperiment</a> container with the TPM (<code>abundance</code>), estimated counts (<code>counts</code>) and transcript length (<code>length</code>) in the assays slot for transcripts.</li> <li><code>salmon_tx2gene.tsv</code>: Tab-delimited file containing gene to transcripts ids mappings.</li> </ul> <p><code>star_salmon/salmon/&lt;SAMPLE&gt;/</code></p> <ul> <li><code>logs/</code>: Contains the file <code>salmon_quant.log</code> giving a record of Salmon's quantification.</li> <li><code>quant.genes.sf</code>: Salmon <em>gene</em>-level quantification of the sample, including feature length, effective length, TPM, and number of reads.</li> <li><code>quant.sf</code>: Salmon <em>transcript</em>-level quantification of the sample, including feature length, effective length, TPM, and number of reads.</li> </ul> <p><strong>From star_rsem</strong></p> <p><code>star_rsem/</code></p> <ul> <li><code>rsem.merged.gene_counts.tsv</code>: Matrix of gene-level raw counts across all samples.</li> <li><code>rsem.merged.gene_tpm.tsv</code>: Matrix of gene-level TPM values across all samples.</li> <li><code>rsem.merged.transcript_counts.tsv</code>: Matrix of isoform-level raw counts across all samples.</li> <li><code>rsem.merged.transcript_tpm.tsv</code>: Matrix of isoform-level TPM values across all samples.</li> <li><code>*.genes.results</code>: RSEM gene-level quantification results for each sample.</li> <li><code>*.isoforms.results</code>: RSEM isoform-level quantification results for each sample.</li> </ul> <p><code>star_rsem/&lt;SAMPLE&gt;.stat/</code></p> <ul> <li><code>*.cnt</code>, <code>*.model</code>, <code>*.theta</code>: RSEM counts and statistics for each sample.</li> </ul> <p><code>star_rsem/log/</code></p> <ul> <li><code>*.log</code>: STAR alignment report containing the mapping results summary.</li> </ul> <h4 id=7-pseudoaligner>7. <code>pseudoaligner</code><a class=headerlink href=#7-pseudoaligner title="Permanent link">¶</a></h4> <p>This is the same output as the <code>star_salmon/salmon</code> folder.</p> <hr> <p><em>Parts of this lesson have been taken from <a href=https://en.wikipedia.org/wiki/Pipeline_%28software%29>Wikipedia</a>, the <a href=https://www.nextflow.io/ >Nextflow webpage</a> and the <a href=https://nf-co.re/ >nf-core project webpage</a>.</em></p></div> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">May 21, 2024</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3h-2Z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">May 21, 2024</span> </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=04a_preprocessing.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Preprocessing steps"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Preprocessing steps </div> </div> </a> <a href=04c_preprocessing_setup.html class="md-footer__link md-footer__link--next" aria-label="Next: UCloud setup"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> UCloud setup </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> CC-BY-SA 4.0 license </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=assets/javascripts/bundle.081f42fc.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>