---
title: "Log Fold Shrinkage and DEA visualizations"
author: "You!"
date: '`r Sys.Date()`'
knit: (function(inputFile, encoding) { 
      rmarkdown::render(inputFile,
                        encoding=encoding,
                        output_format='all',
                        output_dir='./')})
output:
  # To create PDF report, uncomment below
  #pdf_document:
  #  toc: yes  
  html_document:
    theme: yeti # nice theme for the webpage
    toc: yes # table of contents
    toc_float: yes # table of contents "floats" in the document
    df_print: paged # data frames are interactive
    dev: png # what format do you want for the figures?
---

```{r knitr, include = FALSE}
DOCNAME = knitr::current_input()
DOCNAME = gsub(DOCNAME, pattern = ".Rmd", replacement = "", fixed = T)
knitr::opts_chunk$set(autodep        = TRUE,
                      cache          = FALSE,
                      echo           = TRUE,
                      error          = FALSE,
                      fig.align      = "center",
                      fig.path       = paste0("./img/", DOCNAME, "/"), #images will be put in this folder, under the notebook name
                      message        = FALSE,
                      warning        = FALSE,
                      eval           = TRUE)
```

```{r setup, include = FALSE, echo = FALSE}
# DO NOT RUN IF YOU HAVE ALREADY RUN PREVIOUS NOTEBOOKS
# This chunk is ONLY necessary if you want to knit this document into a pdf!!
library(tidyverse)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(DEGreport)
library(ggrepel)
library(tximport)

setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
meta <- read_table("../Data/Mov10_full_meta.txt")
dir <- "/work/sequencing_data/Preprocessing_backup/results_salmon/salmon"
tx2gene <- read_table(file.path(dir,"salmon_tx2gene.tsv"), col_names = c("transcript_ID","gene_ID","gene_symbol"))
files <- file.path(dir, meta$samplename, "quant.sf")
names(files) <- meta$samplename
txi <- tximport(files, type="salmon", tx2gene=tx2gene, countsFromAbundance = "lengthScaledTPM", ignoreTxVersion	= TRUE)
dds <- DESeqDataSetFromTximport(txi,
                                   colData = meta %>% column_to_rownames("samplename"), 
                              design = ~ sampletype)
dds <- DESeq(dds)

contrast_oe <- c("sampletype", "MOV10_overexpression","control")
res_tableOE <- results(dds, contrast=contrast_oe, alpha = 0.05)
res_tableOE_tb <- res_tableOE %>%
  as_tibble(rownames = "gene") %>%
  relocate(gene, .before = baseMean)

sigOE <- res_tableOE_tb %>%
  dplyr::filter(padj < 0.05)
```

Approximate time: 75 minutes

## Learning Objectives

-   Explain log fold change shrinkage
-   Setup results data for application of visualization techniques
-   Describe different data visualization useful for exploring results from a DGE analysis
-   Create a volcano plot and MA plot to evaluate relationship among DGE statistics
-   Create a heatmap to illustrate expression changes of differentially expressed genes

## More accurate LFC estimates

In the previous lessons, we learned about how to generate a table with Differentially Expressed Genes

```{r, eval = FALSE}
## DO NOT RUN
res_tableOE <- results(dds, contrast=contrast_oe, alpha = 0.05)

head(res_tableOE)
```

The problem with these fold change estimates is that they are not entirely accurate as they do not account for the large dispersion we observe with low read counts. To address this, the **log2 fold changes need to be adjusted**.

To generate the shrunken log2 fold change estimates, you have to run an additional step on your results object (that we will create below) with the function `lfcShrink()`.

```{r}
## Save the unshrunken results to compare
res_tableOE_unshrunken <- res_tableOE

# Apply fold change shrinkage
res_tableOE <- lfcShrink(dds, coef="sampletype_MOV10_overexpression_vs_control", type="apeglm")
```

> For more information on shrinkage, the DESeq2 vignette has an [Extended section on shrinkage estimators](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#extended-section-on-shrinkage-estimators) that is quite useful.

> #### `contrast` vs `coef`
>
> When using the shrinkage method, rather than using the `contrast` argument you will be required to specify `coef`. Using contrast forms an expanded model matrix, treating all factor levels equally, and averages over all distances between all pairs of factor levels to estimate the prior. Using coef, means looking only at that column of the model matrix (so usually that would be one level against the reference level) and estimates the prior for that coefficient from the distribution of those MLE of coefficients. When using coef, the shrinkage depends on which level is chosen as reference.

> #### How do I know what to value to provide to the `coef` argument?
>
> The value you provide here needs to match identically to what is stored in the column header of the coefficients table. To see what values you have to work with you can use `resultsNames(dds)`.

## Visualizing the results

### MA plot

A plot that can be useful to exploring our results is the MA plot. The MA plot shows the **mean of the normalized counts versus the log2 fold changes for all genes tested**. The genes that are significantly DE are colored to be easily identified (adjusted p-value \< 0.01 by default).

**Let's start with the unshrunken results:**

```{r MAplot_unshrunken}
# MA plot using unshrunken fold changes
plotMA(res_tableOE_unshrunken, ylim=c(-2,2))
```

**And now the shrunken results:**

```{r MAplot_shrunken}
# MA plot using shrunken fold changes
plotMA(res_tableOE, ylim=c(-2,2))
```

### Question

**Why are there genes with high mean and big log2 fold changes, but are not statistically significant?**

## Advanced visualizations

We will be working with three different data objects we have already created in earlier lessons:

-   Metadata for our samples (a dataframe): `meta`
-   Normalized expression data for every gene in each of our samples (a matrix): `normalized_counts`
-   Tibble versions of the DESeq2 results we generated in the last lesson: `res_tableOE_tb` and `res_tableKD_tb`

First, we already have a metadata tibble.

```{r}
meta %>% head()
```

Next, let's bring in the `normalized_counts` object with our gene names.

```{r}
# DESeq2 creates a matrix when you use the counts() function
## First convert normalized_counts to a data frame and transfer the row names to a new column called "gene"
normalized_counts <- counts(dds, normalized=T) %>% 
  data.frame() %>%
  rownames_to_column(var="gene") 
```

### Plotting significant DE genes

One way to visualize results would be to simply plot the expression data for a handful of genes. We could do that by picking out specific genes of interest or selecting a range of genes.

**Using DESeq2 `plotCounts()` to plot expression of a single gene**

To pick out a specific gene of interest to plot, for example MOV10, we can use the `plotCounts()` from DESeq2. `plotCounts()` requires that the gene specified matches the original input to DESeq2.

```{r countPlot}
# Plot expression for single gene
plotCounts(dds, gene="MOV10", intgroup="sampletype") 
```

**Using ggplot2 to plot expression of a single gene**

If you wish to change the appearance of this plot, we can save the output of `plotCounts()` to a variable specifying the `returnData=TRUE` argument, then use `ggplot()`:

```{r}
# Save plotcounts to a data frame object
d <- plotCounts(dds, gene="MOV10", intgroup="sampletype", returnData=TRUE)

# What is the data output of plotCounts()?
d %>% head()
```

```{r custom_countPlot}
# Plot the MOV10 normalized counts, using the samplenames (rownames(d) as labels)
ggplot(d, aes(x = sampletype, y = count, color = sampletype)) + 
geom_point(position=position_jitter(w = 0.1,h = 0)) +
geom_text_repel(aes(label = rownames(d))) + 
theme_bw() +
ggtitle("MOV10") +
theme(plot.title = element_text(hjust = 0.5))
```

### Heatmap

In addition to plotting subsets, we could also extract the normalized values of *all* the significant genes and plot a heatmap of their expression using `pheatmap()`.

```{r}
### Extract normalized expression for significant genes from the OE and control samples (columns 4 to 9)
### also get gene name (column 1)
norm_OEsig <- normalized_counts[,c(1,4:9)] %>% 
  dplyr::filter(gene %in% sigOE$gene)  
```

Now let's draw the heatmap using `pheatmap`:

```{r sigOE_heatmap}
### Run pheatmap using the metadata data frame for the annotation
pheatmap(log(norm_OEsig[2:7]+1), 
         cluster_rows = T, 
         show_rownames = F,
         annotation = meta %>% column_to_rownames(var = "samplename"), 
         border_color = NA, 
         fontsize = 10, 
         #scale = "row", 
         fontsize_row = 10, 
         height = 20)
```

> *NOTE:* There are several additional arguments we have included in the function for aesthetics. One important one is `scale="row"`, in which Z-scores are plotted, rather than the actual normalized count value.
>
> Z-scores are computed on a gene-by-gene basis by subtracting the mean and then dividing by the standard deviation. The Z-scores are computed **after the clustering**, so that it only affects the graphical aesthetics and the color visualization is improved.

### Volcano plot

To generate a volcano plot, we first need to have a column in our results data indicating whether or not the gene is considered differentially expressed based on p-adjusted values and we will include a log2fold change here.

```{r}
## Obtain logical vector where TRUE values denote padj values < 0.05 and fold change > 1.5 in either direction

res_tableOE_tb <- res_tableOE_tb %>% 
mutate(threshold_OE = padj < 0.05 & abs(log2FoldChange) >= 0.58)
```

Now we can start plotting. The `geom_point` object is most applicable, as this is essentially a scatter plot:

```{r volcano_plot}
## Volcano plot
ggplot(res_tableOE_tb) +
geom_point(aes(x = log2FoldChange, y = -log10(padj), colour = threshold_OE)) +
ggtitle("Mov10 overexpression") +
xlab("log2 fold change") + 
ylab("-log10 adjusted p-value") +
#scale_y_continuous(limits = c(0,50)) +
theme(legend.position = "none",
plot.title = element_text(size = rel(1.5), hjust = 0.5),
axis.title = element_text(size = rel(1.25)))  
```

**Checking the top DE genes**

This is a great way to get an overall picture of what is going on, but what if we also wanted to know where the top 10 genes (lowest padj) in our DE list are located on this plot? We could label those dots with the gene name on the Volcano plot using `geom_text_repel()`.

First, we need to order the res_tableOE tibble by `padj`, and add an additional column to it, to include on those gene names we want to use to label the plot.

```{r}
## Create an empty column to indicate which genes to label
res_tableOE_tb <- res_tableOE_tb %>% mutate(genelabels = "")

## Sort by padj values 
res_tableOE_tb <- res_tableOE_tb %>% arrange(padj)

## Populate the genelabels column with contents of the gene symbols column for the first 10 rows, i.e. the top 10 most significantly expressed genes
res_tableOE_tb$genelabels[1:10] <- as.character(res_tableOE_tb$gene[1:10])

head(res_tableOE_tb)
```

Next, we plot it as before with an additional layer for `geom_text_repel()` wherein we can specify the column of gene labels we just created.

```{r volcano_plot_label}
ggplot(res_tableOE_tb, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(colour = threshold_OE)) +
  geom_text_repel(aes(label = genelabels)) +
  ggtitle("Mov10 overexpression") +
  xlab("log2 fold change") + 
  ylab("-log10 adjusted p-value") +
  theme(legend.position = "none",
        plot.title = element_text(size = rel(1.5), hjust = 0.5),
        axis.title = element_text(size = rel(1.25))) 
```

------------------------------------------------------------------------

## Extra: Reports for DEA

***NOTE:** If using the DESeq2 tool for differential expression analysis, the package 'DEGreport' can use the DESeq2 results output to make the top20 genes and the volcano plots generated above by writing a few lines of simple code. While you can customize the plots above, you may be interested in using the easier code. Below are examples of the code to create these plots:*

```{r degPlot, fig.height=10}
DEGreport::degPlot(dds = dds, res = res_tableOE_unshrunken, n = 20, xs = "sampletype", group = "sampletype") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))# dds object is output from DESeq2
```

```{r degPlotWide}
DEGreport::degPlotWide(counts = dds, genes = row.names(res_tableOE)[1:5], group = "sampletype")
```

```{r degVolcano}
DEGreport::degVolcano(
    data.frame(res_tableOE[,c("log2FoldChange","padj")]), # table - 2 columns
    plot_text = data.frame(res_tableOE_tb[1:10,c("log2FoldChange","padj","gene")])) # table to add names
```

------------------------------------------------------------------------
