---
title: "bulk RNAseq basics"
author: "Jose Alejandro Romero Herrera"
date: '`r Sys.Date()`'
knit: (function(inputFile, encoding) { 
      rmarkdown::render(inputFile,
                        encoding=encoding,
                        output_format='all',
                        output_dir='./reports/')})
output:
  # To create PDF report, uncomment below
  #pdf_document:
  #  toc: yes
  # html_document:
  #   number_sections: yes
  #   theme: yeti
  #   toc: yes
  #   toc_float: yes
  #   df_print: paged
  #   dev: png
  md_document
---

```{r knitr, include = FALSE}
DOCNAME = knitr::current_input()
knitr::opts_chunk$set(autodep        = TRUE,
                      cache          = FALSE,
                      cache.path     = paste0("cache/", DOCNAME, "/"),
                      cache.comments = TRUE,
                      echo           = TRUE,
                      error          = FALSE,
                      fig.align      = "center",
                      fig.path       = paste0("./figures/", DOCNAME, "/"),
                      fig.width      = 10,
                      fig.height     = 8,
                      message        = FALSE,
                      warning        = FALSE)
```

# Differential expression analysis
In order to test for differential expression, we operate on raw counts and use discrete distributions
After checking that our samples are well correlated within conditions, we can proceed to test for Differantially Expressed Genes using the *results()* function.

```{r}
res <- results(dds)
head(data.frame(res))
```

Information about the results columns can be retrieved using the following snippet:
```{r}
mcols(res)$description
```

In this example, we only have two conditions: "untreated" and "treated", so there is no need to specify which comparison we would like to make.
If your experiment contains more than two conditions, there are different ways of specifying which comparison you would like to make.

First, the "contrast" argument will take a vector that specifies the metadata and the order of the comparison. In this case, you can swap the order of the condition levels.
```{r}
res <- results(dds, contrast=c("condition","untreated","treated"))
head(data.frame(res))
```

Second, "name" argument will use the results of the *resultsNames()* function. This function creates possible combinations of conditions, but always using the reference condition for the comparison.
```{r}
resultsNames(dds)
res <- results(dds, name="condition_treated_vs_untreated")
head(data.frame(res))
```

The notebook *contrast_design.Rmd* contains information about how to understand and define contrasts using model matrices, which might be a more intuitive way of comparing samples or conditions.

## Extracting results
The output of the *res()* function is the one you want to use/save to identify differentially expressed genes using log2 Fold Change or adjusted p-value thresholds.

There is a chance that the adjusted p-value is NA. If you will work on this results, it might be helpful to change all NA to an adjusted p-value of 1.
```{r}
res$padj[is.na(res$padj)] <- 1
```

We can have a quick look of the adjusted p-value results by using the following custom function:
```{r}
results_summary <- function(x, alpha = 0.05, LFC = 1) {
  ngenes <- nrow(x)
  signif <- sum(x$padj < alpha, na.rm = T)
  up <- sum(x$padj < alpha & x$log2FoldChange > LFC, na.rm = T)
  down <- sum(x$padj < alpha & x$log2FoldChange < -LFC, na.rm = T)
  results <- c(paste0("Number of genes: ", ngenes),
    paste0("Number of genes with adjusted p-value < ",alpha,": ", signif, " (", round((signif/ngenes)*100,digits = 2),"%)"),
    "  Of those:",
    paste0("    with LFC < ", -LFC, ": ", down, " (", round((down/ngenes)*100,digits = 2),"%)"),
    paste0("    with LFC > ", LFC, ": ", up, " (", round((up/ngenes)*100,digits = 2),"%)"))
  writeLines(paste(results, collapse = "\n"))
  return(paste(results, collapse = "\n"))
}
```


```{r}
res_summary <- results_summary(res, alpha = 0.05, LFC = 1)
```

### Save significant results
Extracting results based on adjusted p-value and LFC. this way we will get a filteres table with genes with LFC > 1 or LFC < -1 with an adjusted p-value of 0.05
```{r}
LFC <- 1
adj_pvalue <- 0.05

sig_res <- res[res$padj < adj_pvalue & abs(res$log2FoldChange) > LFC,]
#write.table(x = sig_res, "./significant_results.tsv", quote = F, col.names = T, row.names = T, sep = "\t")
```

# DEA visualizations 
## Count matrix heatmaps
We can visualize different heatmaps depending on what we are interested on. We can see the top most expressed genes, the top most variable genes and the DE genes.

### Top most variable genes
```{r most_var_genes_heatmap}
ntop <- 20
select <- order(rowSds(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:ntop]
df <- as.data.frame(colData(dds)[,c("condition","type")])
pheatmap(assay(vsd)[select,], cluster_rows=T, show_rownames=FALSE, scale = "row",
         cluster_cols=FALSE, annotation_col=df)
```

### Top expressed genes
```{r top_exp_genes_heatmap}
ntop <- 20
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:ntop]
df <- as.data.frame(colData(dds)[,c("condition","type")])
pheatmap(assay(vsd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)
```

### Differentially expressed genes
```{r DEA_heatmap}
select <- rownames(sig_res)
df <- as.data.frame(colData(dds)[,c("condition","type")])
pheatmap(assay(vsd)[rownames(vsd) %in% select,], cluster_rows=TRUE, show_rownames=FALSE, scale = "row",
         cluster_cols=FALSE, annotation_col=df)
```

## Fold change plots
For fold change plots, it is useful to shrunk log2 fold changes, which removes the noise associated with log2 fold changes from low count genes without requiring arbitrary filtering thresholds.
```{r}
resLFC <- lfcShrink(dds, coef="condition_treated_vs_untreated", type="apeglm")
resLFC$padj[is.na(resLFC$padj)] <- 1
```

### MA plot
```{r MA_plot}
alpha <- 0.05

ggplot(data.frame(resLFC)) + theme_bw() + 
  labs(title = "MA plot", subtitle = res_summary,
       color = paste0("Adjusted p-value < ",alpha)) +
  geom_point(aes(x = log10(baseMean), y = log2FoldChange, 
                 color = factor((padj < alpha), levels = c("TRUE","FALSE"))), size = 1)
```

### Volcano plot
```{r volcano_plot}
alpha = 0.05
LFC = 1
ggplot(data.frame(resLFC)) + theme_bw() + 
  labs(color ="Significant genes", title = "Volcano plot", subtitle = res_summary) + 
  geom_hline(yintercept = -log10(alpha), linetype = "dashed") + 
  geom_vline(xintercept = c(-LFC,LFC), linetype = "dashed") + 
  geom_point(aes(x = log2FoldChange, y = -log10(padj),
                 color = factor((padj < alpha & abs(log2FoldChange) > LFC), levels = c("TRUE","FALSE"))), size = 1)
```

### Glimma plots
We can visualize our results interactively using the `glimmaMA()` and `glimmaVolcano()` functions.

```{r}
#glimmaMA(dds, groups = dds$condition)
#glimmaVolcano(dds, groups = dds$condition)
```

Save your interactive plots using the `htmlwidgets::saveWidget()` function.
```{r}
#htmlwidgets::saveWidget(glimmaMA(dds, groups = dds$condition), "ma-plot.html")
#htmlwidgets::saveWidget(glimmaVolcano(dds, groups = dds$condition), "volcano-plot.html")
```

# Session info
Finally, we create a `session_info()` table that will allow anyone to check what versions of R and packages are we using for reproducibility purposes.

```{r session-info}
devtools::session_info()
```