---
summary: Index page, intro to course
listing:
  fields: [image, title, description]
  contents: cards/*.qmd
  type: grid
  image-height: 200px
  grid-columns: 5
css: ../index.css
date-modified: last-modified
date-format: long
format:
  html:
    page-layout: full
    toc: false
---

<!--
# Put above to hide navigation (left), toc (right) or footer (bottom)
hide: navigation, toc or footer

# You should hide the navigation if there are no subsections
# You should hide the Table of Contents if there are no important titles
-->

::: {.hero-banner}
# Welcome to the bulk RNA-seq analysis workshop
:::

<hr>

This workshop material includes a tutorial on how to approach RNAseq data, starting from your sequencing reads (fastq files). Thus, the workshop only briefly touches upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, mainly for the purpose of outlining considerations in the downstream bioinformatic analysis. This workshop is based on the materials developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/), a collection of modified tutorials from the [DESeq2](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html), R language vignettes and the [nf-core rnaseq pipeline](https://nf-co.re/rnaseq).

<hr>

:::{.callout-note title="Course Overview"}
- üìñ **Syllabus:**  

1.  Course introduction  
2.  Experimental planning   
3.  Data explanation    
4.  Read reprocessing and preprocessing pipelines   
5.  Analysing RNAseq data   
    1.  RNAseq counts   
    2.  Exploratory analysis    
    3.  Differential Expression Analysis    
    4.  Functional analysis 
6.  Summarized workflow  

- ‚è∞ **Total Time Estimation:** 8 hours
- üìÅ **Supporting Materials:**  Workshop slides with theory on bulk RNAseq can be found in this [zenodo repository](https://zenodo.org/record/8046218).
- üë®‚Äçüíª **Target Audience:** Ph.D., MSc, etc.
- üë©‚Äçüéì **Level:** Beginner.
- üîí **License:** [Creative Commons Attribution-ShareAlike 4.0 International (CC BY-SA 4.0) license.](https://creativecommons.org/licenses/by-sa/4.0/legalcode)  
- üí∞ **Funding:** This project was funded by the Novo Nordisk Fonden (NNF20OC0063268).
:::

:::{.callout-warning title="Course Requirements"}
  
  - Knowledge of R, Rstudio and Rmarkdown. It is recommended that you have at least followed our workshop [R basics](https://github.com/Center-for-Health-Data-Science/FromExceltoR_2022)
  - Basic knowledge of RNAseq technology
  - Basic knowledge of data science and statistics such as PCA, clustering and statistical testing
:::

This workshop material includes a tutorial on how to approach RNAseq data, starting from your sequencing reads (fastq files). Thus, the workshop only briefly touches upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, mainly for the purpose of outlining considerations in the downstream bioinformatic analysis. This workshop is based on the materials developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/), a collection of modified tutorials from the [DESeq2](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html), R language vignettes and the [nf-core rnaseq pipeline](https://nf-co.re/rnaseq).

The aim of this repository is to run a comprehensive but introductory workshop on bulk-RNAseq bioinformatic analyses. Each of the modules of this workshop is accompanied by a powerpoint slideshow explaining the steps and the theory behind a typical bioinformatics analysis (ideally with a teacher). Many of the slides are annotated with extra information and/or point to original sources for extra reading material.

::: {.callout-tip title="Course Goals"}
By the end of this workshop, you should be able to analyse your own bulk RNAseq data:

  -   Preprocess your reads into a count matrix.
  -   Normalize your data.
  -   Explore your samples with PCAs and heatmaps.
  -   Perform Differential Expression Analysis.
  -   Annotate your results.
:::

### Acknowledgements

We recognize the substantial contribution of Jos√© Alejandro Romero Herrera, a former team member, in developing the course material. Other members that have contributed to the development of this course:

| Member                        | Role                                | Institution                        | PI                                 |
|-------------------------------|-------------------------------------|------------------------------------|------------------------------------|
| Jennifer Bartell              | Project Manager, Data Scientist | Center for Health Data Science, KU |Anders Krogh                        |
| Diana Andrejeva          | Data Scientist                      | Center for Health Data Science, KU |Anders Krogh                        |
| Samuele Soraggi               | Data Scientist                      | Bioinformatics Research Centre, AU |Mikkel Schierup                     |

We would also like to extend our gratitude to:

- [Center for Health Data Science](https://heads.ku.dk/), University of Copenhagen.
- [Hugo Tavares](https://bioinfotraining.bio.cam.ac.uk/about), Bioinformatics Training Facility, University of Cambridge.
- [Silvia Raineri](https://renew.ku.dk/research/reseach-groups/serup-group/), Center for Stem Cell Medicine (reNew), University of Copenhagen.
- [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/), check out their [github repo](https://github.com/hbctraining/DGE_workshop_salmon_online)
- [nf-core community](https://nf-co.re/)

### Course Instructors 

<!-- AUTHORS  Format: [author name](link to author page){ .md-button } 
<h4>Authors</h4>

[cards cols="6"(./develop/cards/cards.yaml)]
-->