---
title: "Bulk RNAseq preprocessing"
author: "Jose Alejandro Romero Herrera"
date: '`r Sys.Date()`'
knit: (function(inputFile, encoding) { 
      rmarkdown::render(inputFile,
                        encoding=encoding,
                        output_format='all',
                        output_dir='./reports/')})
output:
  # To create PDF report, uncomment below
  #pdf_document:
  #  toc: yes
  # html_document:
  #   number_sections: yes
  #   theme: yeti
  #   toc: yes
  #   toc_float: yes
  #   df_print: paged
  #   dev: png
  md_document
---

```{r knitr, include = FALSE}
DOCNAME = knitr::current_input()
knitr::opts_chunk$set(autodep        = TRUE,
                      cache          = FALSE,
                      cache.path     = paste0("cache/", DOCNAME, "/"),
                      cache.comments = TRUE,
                      echo           = TRUE,
                      error          = FALSE,
                      fig.align      = "center",
                      fig.path       = paste0("./figures/", DOCNAME, "/"),
                      fig.width      = 10,
                      fig.height     = 8,
                      message        = FALSE,
                      warning        = FALSE)
```

# Introduction

This is a basic bulk RNAseq analysis using the package **DESeq2** and **gprofiler2** for Differential Expression Analysis and Functional Annotation. The code is a modified version of the [DESeq2](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) and [gprofiler2](https://cran.r-project.org/web/packages/gprofiler2/vignettes/gprofiler2.html) vignettes. The data used for this example below is the *pasilla* dataset: an RNAseq experiment that studied the effect of RNAi knockdown of Pasilla, the Drosophila melanogaster ortholog of mammalian NOVA1 and NOVA2, on the transcriptome.

# Load libraries
```{r, results='hide'}
library(tidyverse, quietly = T)
library(DESeq2, quietly = T)
library(RColorBrewer, quietly = T)
library(pheatmap, quietly = T)
library(gprofiler2, quietly = T)
library(Glimma, quietly = T)
```

# Load data
In order to proceed with the analysis we need two dataframes:
- Count matrix dataframe with gene names as rows (and rownames) and samples as columns (and column names). 
- Sample metadata dataframe with samples as rows (and rownames) and metadata variables as columns (and column names).

Additionally, you can add information about your features (genes), with genes as rows and gene metadata on columns.

Ideally, the sample metadata will contain information about the conditions, treatments, replicates, etc.

The pasilla dataset contains a count matrix and the metadata, which is a good example to use as a template. The count matrix looks like this:
```{r}
cts <- data.frame(read.csv(file = "../../data/example_data/pasilla_cts.tsv", sep="\t", row.names="flybase_id"))
head(cts)
```

While the sample metadata looks like this:
```{r}
coldata <- read.csv(file = "../../data/example_data/pasilla_metadata.tsv", row.names=1, sep = "\t")
head(coldata)
```

Making the metadata factors will establish the order of the metadata variables. If you never tell the DESeq2 functions which level you want to compare against (e.g. which level represents the control group), the comparisons will be based on the alphabetical order of the levels. This is important since the first level of the metadata will be the reference sample/condition for the DE analysis.
```{r}
coldata$condition <- factor(coldata$condition)
paste("The order of the conditions is:", paste(unique(coldata$condition), collapse=", "))
coldata$type <- factor(coldata$type)
paste("The order of the types is:", paste(unique(coldata$type), collapse=", "))
```

## Metadata factor levels
If you want to change the order of the factor levels you can use one of these two.
NOTE: this has to be done before running the `DESeq()` function.
```{r}
#coldata$condition <- factor(coldata$condition, levels = c("untreated","treated"))
coldata$condition <- relevel(coldata$condition, ref = "untreated") #Specifies "untreated" as the reference level
```

## Create DESeq object
We create the DESeq object from the count matrix and the metadata. We specify that the analysis will be based on the design column.
```{r}
dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata, rowData = rownames(cts),
                              design = ~ condition)
dds
```
## Pre-filtering
We can reduce the computational resources of the analysis by removing genes that are very hardly expressed. Additionally, you can collapse technical replicates using the function `collapseReplicates()`.

```{r}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
```

## Normalization
Finally the `DESeq()` function will normalize your read counts and estimate size factors per genes. This function is essential for the rest of the analysis.

```{r}
dds <- DESeq(dds)
```

# Session info
```{r session-info}
devtools::session_info()
```