<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
{
  "@context": "https://schema.org",
  "@id": "https://hds-sandbox.github.io/bulk_RNAseq_course/",
  "@type": "Course",
  "coursePrerequisites": [
    "Knowledge of R, Rstudio and Rmarkdown",
    "Basic knowledge of RNAseq technology",
    "Basic knowledge of data science and statistics such as PCA, clustering and statistical testing"
  ],
  "dct:conformsTo": "https://bioschemas.org/profiles/Course/1.0-RELEASE",
  "description": "This workshop includes a tutorial on how to approach RNAseq data, starting from your sequencing reads (fastq files). Thus, the workshop only briefly touches upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, mainly for the purpose of outlining considerations in the downstream bioinformatic analysis. This workshop is based on the materials developed by members of the teaching team at the Harvard Chan Bioinformatics Core (HBC), a collection of modified tutorials from the DESeq2, R language vignettes and the nf-core rnaseq pipeline.",
  "isBasedOn": [
    "https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html",
    "https://nf-co.re/rnaseq",
    "https://github.com/hbctraining/Intro-to-rnaseq-hpc-salmon-flipped"
  ],
  "keywords": "bioinformatics, rnaseq, bulk, R",
  "license": {
    "@id": "https://creativecommons.org/licenses/by/4.0/",
    "@type": "CreativeWork"
  },
  "mentions": [
    {
      "@id": "10.5281/zenodo.7660225",
      "@type": "Thing"
    },
    {
      "@id": "10.5281/zenodo.7565997",
      "@type": "Thing"
    }
  ],
  "name": "Introduction to bulk RNAseq analysis workshop"
}
</script><title>Experimental design considerations - Bulk RNAseq data analysis</title><link rel=stylesheet href=assets/stylesheets/main.975780f9.min.css><link rel=stylesheet href=assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#experimental-design-considerations class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bulk RNAseq data analysis </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Experimental design considerations </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=02_experimental_planning.html class="md-tabs__link md-tabs__link--active"> Experimental planning </a> </li> <li class=md-tabs__item> <a href=03_data_explanation.html class=md-tabs__link> Data Explanation </a> </li> <li class=md-tabs__item> <a href=04a_preprocessing.html class=md-tabs__link> Read Preprocessing </a> </li> <li class=md-tabs__item> <a href=05a_data_analysis_setup.html class=md-tabs__link> Analysing RNAseq data </a> </li> <li class=md-tabs__item> <a href=09_summarized_workflow.html class=md-tabs__link> Summarized workflow </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation hidden> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> Bulk RNAseq data analysis </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> Course Introduction </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Experimental planning <span class="md-nav__icon md-icon"></span> </label> <a href=02_experimental_planning.html class="md-nav__link md-nav__link--active"> Experimental planning </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#controls class=md-nav__link> Controls </a> </li> <li class=md-nav__item> <a href=#replicates class=md-nav__link> Replicates </a> </li> <li class=md-nav__item> <a href=#confounding-variables class=md-nav__link> Confounding variables </a> </li> <li class=md-nav__item> <a href=#batch-effects class=md-nav__link> Batch effects </a> <nav class=md-nav aria-label="Batch effects"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#how-to-know-whether-you-have-batches class=md-nav__link> How to know whether you have batches? </a> </li> <li class=md-nav__item> <a href=#best-practices-regarding-batches class=md-nav__link> Best practices regarding batches: </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=03_data_explanation.html class=md-nav__link> Data Explanation </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_4 type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4> Read Preprocessing <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Read Preprocessing" data-md-level=1> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Read Preprocessing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=04a_preprocessing.html class=md-nav__link> Preprocessing steps </a> </li> <li class=md-nav__item> <a href=04b_pipelines.html class=md-nav__link> Nextflow and nf-core pipelines </a> </li> <li class=md-nav__item> <a href=04c_preprocessing_setup.html class=md-nav__link> UCloud setup </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5 type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5> Analysing RNAseq data <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Analysing RNAseq data" data-md-level=1> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Analysing RNAseq data </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=05a_data_analysis_setup.html class=md-nav__link> Ucloud setup </a> </li> <li class=md-nav__item> <a href=05b_count_matrix.html class=md-nav__link> Count matrix </a> </li> <li class=md-nav__item> <a href=05c_count_normalization.html class=md-nav__link> Normalization </a> </li> <li class=md-nav__item> <a href=06_exploratory_analysis.html class=md-nav__link> Exploratory Analysis </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5_5 type=checkbox id=__nav_5_5> <label class=md-nav__link for=__nav_5_5> Differential Expression <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Differential Expression" data-md-level=2> <label class=md-nav__title for=__nav_5_5> <span class="md-nav__icon md-icon"></span> Differential Expression </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=07_extra_contrast_design.html class=md-nav__link> Design formula </a> </li> <li class=md-nav__item> <a href=07a_DEA.html class=md-nav__link> DESeq2 </a> </li> <li class=md-nav__item> <a href=07b_hypothesis_testing.html class=md-nav__link> Hypothesis testing </a> </li> <li class=md-nav__item> <a href=07c_DEA_visualization.html class=md-nav__link> Visualization </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5_6 type=checkbox id=__nav_5_6> <label class=md-nav__link for=__nav_5_6> Functional Analysis <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Functional Analysis" data-md-level=2> <label class=md-nav__title for=__nav_5_6> <span class="md-nav__icon md-icon"></span> Functional Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=08a_FA_genomic_annotation.html class=md-nav__link> Gene annotation </a> </li> <li class=md-nav__item> <a href=08b_FA_overrepresentation.html class=md-nav__link> Over representation analysis </a> </li> <li class=md-nav__item> <a href=08c_FA_GSEA.html class=md-nav__link> Class scoring </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=09_summarized_workflow.html class=md-nav__link> Summarized workflow </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#controls class=md-nav__link> Controls </a> </li> <li class=md-nav__item> <a href=#replicates class=md-nav__link> Replicates </a> </li> <li class=md-nav__item> <a href=#confounding-variables class=md-nav__link> Confounding variables </a> </li> <li class=md-nav__item> <a href=#batch-effects class=md-nav__link> Batch effects </a> <nav class=md-nav aria-label="Batch effects"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#how-to-know-whether-you-have-batches class=md-nav__link> How to know whether you have batches? </a> </li> <li class=md-nav__item> <a href=#best-practices-regarding-batches class=md-nav__link> Best practices regarding batches: </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course/edit/master/docs/02_experimental_planning.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1 id=experimental-design-considerations>Experimental design considerations<a class=headerlink href=#experimental-design-considerations title="Permanent link">&para;</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>&#128368; <strong>Time Estimation:</strong> 30 minutes </p> <p>&#128172; <strong>Learning Objectives:</strong> </p> <ol> <li>Describe the importance of replicates for RNA-seq differential expression experiments.</li> <li>Explain the relationship between the number of biological replicates, sequencing depth and the differentially expressed genes identified.</li> <li>Demonstrate how to design an RNA-seq experiment that avoids confounding and batch effects.</li> </ol> </div> <p>Understanding the steps in the experimental process of RNA extraction and preparation of RNA-Seq libraries is helpful for designing an RNA-Seq experiment, but there are special considerations that should be highlighted that can greatly affect the quality of a differential expression analysis.</p> <p>These important considerations include:</p> <ol> <li>Proper experiment <strong>controls</strong></li> <li>Number and type of <strong>replicates</strong></li> <li>Issues related to <strong>confounding</strong></li> <li>Addressing <strong>batch effects</strong></li> </ol> <p>We will go over each of these considerations in detail, discussing best practice and optimal design.</p> <h2 id=controls>Controls<a class=headerlink href=#controls title="Permanent link">&para;</a></h2> <p><strong>Experimental controls</strong> must be used in order to minimize the effect of variables which are not the interest of the study. Thus, it allows the experiment to minimize the changes in all other variables except the one being tested, and help us ensure that there have been no deviations in the environment of the experiment that could end up influencing the outcome of the experiment, besides the variable they are investigating.</p> <p>There are different types of controls, but we will mainly see <strong>positive</strong> and <strong>negative</strong> controls:</p> <ul> <li><strong>Negative</strong>: The negative control is a variable or group of samples where no response is expected.</li> <li><strong>Positive</strong>: A positive control is a variable or group of samples that receives a treatment with a known positive result.</li> </ul> <p>It is very important that you give serious thought about proper controls of your experiment so you can control as many sources of variation as possible. This will greatly strengthen the results of your experiment.</p> <h2 id=replicates>Replicates<a class=headerlink href=#replicates title="Permanent link">&para;</a></h2> <p>Experimental replicates can be performed as <strong>technical replicates</strong> or <strong>biological replicates</strong>.</p> <p align=center> <img src=./img/02_experimental_planning/replicates.png width=500> </p> <p><em>Image credit: <a href=https://dx.doi.org/10.15252%2Fembj.201592958>Klaus B., EMBO J (2015) <strong>34</strong>: 2727-2730</a></em></p> <ul> <li> <p><strong>Technical replicates:</strong> use the same biological sample to repeat the technical or experimental steps in order to accurately measure technical variation and remove it during analysis.</p> </li> <li> <p><strong>Biological replicates</strong> use different biological samples of the same condition to measure the biological variation between samples.</p> </li> </ul> <p>In the days of microarrays, technical replicates were considered a necessity; however, with the current RNA-Seq technologies, technical variation is much lower than biological variation and <strong>technical replicates are unnecessary</strong>.</p> <p>In contrast, <strong>biological replicates are absolutely essential</strong> for differential expression analysis. For mice or rats, it might be easy to determine what constitutes a different biological sample, but it's a bit <strong>more difficult to determine for cell lines</strong>. <a href=http://paasp.net/accurate-design-of-in-vitro-experiments-why-does-it-matter/ >This article</a> gives some great recommendations for cell line replicates.</p> <p>For differential expression analysis, the more biological replicates, the better the estimates of biological variation and the more precise our estimates of the mean expression levels. This leads to more accurate modeling of our data and identification of more differentially expressed genes.</p> <p align=center> <img src=./img/02_experimental_planning/de_replicates_img2.png width=500> </p> <p><em>Image credit: <a href=https://doi.org/10.1093/bioinformatics/btt688>Liu, Y., et al., Bioinformatics (2014) <strong>30</strong>(3): 301--304</a></em></p> <p>As the figure above illustrates, <strong>biological replicates are of greater importance than sequencing depth</strong>, which is the total number of reads sequenced per sample. The figure shows the relationship between sequencing depth and number of replicates on the number of differentially expressed genes identified [<a href=https://academic.oup.com/bioinformatics/article/30/3/301/228651/RNA-seq-differential-expression-studies-more>1</a>]. Note that an <strong>increase in the number of replicates tends to return more DE genes than increasing the sequencing depth</strong>. Therefore, generally more replicates are better than higher sequencing depth, with the caveat that higher depth is required for detection of lowly expressed DE genes and for performing isoform-level differential expression.</p> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p><strong>Sample pooling:</strong> Try to avoid pooling of individuals/experiments, if possible; however, if absolutely necessary, then each pooled set of samples would count as a <strong>single replicate</strong>. To ensure similar amounts of variation between replicates, you would want to pool the <strong>same number of individuals</strong> for each pooled set of samples.</p> <p><em>For example, if you need at least 3 individuals to get enough material for your <code>control</code> replicate and at least 5 individuals to get enough material for your <code>treatment</code> replicate, you would pool 5 individuals for the <code>control</code> and 5 individuals for the <code>treatment</code> conditions. You would also make sure that the individuals that are pooled in both conditions are similar in sex, age, etc.</em></p> </div> <p>Replicates are almost always preferred to greater sequencing depth for bulk RNA-Seq. However, <strong>guidelines depend on the experiment performed and the desired analysis</strong>. Below we list some general guidelines for replicates and sequencing depth to help with experimental planning:</p> <ul> <li> <p><strong>General gene-level differential expression:</strong></p> <ul> <li>ENCODE guidelines suggest 30 million SE reads per sample (stranded).</li> <li>15 million reads per sample is often sufficient, if there are a good number of replicates (>3).</li> <li>Spend money on more biological replicates, if possible.</li> <li>Generally recommended to have read length >= 50 bp</li> </ul> </li> <li> <p><strong>Gene-level differential expression with detection of lowly-expressed genes:</strong></p> <ul> <li>Similarly benefits from replicates more than sequencing depth.</li> <li>Sequence deeper with at least 30-60 million reads depending on level of expression (start with 30 million with a good number of replicates).</li> <li>Generally recommended to have read length >= 50 bp</li> </ul> </li> <li> <p><strong>Isoform-level differential expression:</strong></p> <ul> <li>Of known isoforms, suggested to have a depth of at least 30 million reads per sample and paired-end reads.</li> <li>Of novel isoforms should have more depth (> 60 million reads per sample).</li> <li>Choose biological replicates over paired/deeper sequencing.</li> <li>Generally recommended to have read length >= 50 bp, but longer is better as the reads will be more likely to cross exon junctions</li> <li>Perform careful QC of RNA quality. Be careful to use high quality preparation methods and restrict analysis to high quality RIN # samples.</li> </ul> </li> <li> <p><strong>Other types of RNA analyses (intron retention, small RNA-Seq, etc.):</strong></p> <ul> <li>Different recommendations depending on the analysis.</li> <li>Almost always more biological replicates are better!</li> </ul> </li> </ul> <div class="admonition info"> <p class=admonition-title>What is coverage?</p> </div> <p>The factor used to estimate the depth of sequencing for genomes is "coverage" - how many times do the number of nucleotides sequenced "cover" the genome. This metric is not exact for genomes (whole genome sequencing), but it is good enough and is used extensively. However, the metric <strong>does not work for transcriptomes</strong> because even though you may know what % of the genome has transcriptional activity, the expression of the genes is highly variable.</p> <h2 id=confounding-variables>Confounding variables<a class=headerlink href=#confounding-variables title="Permanent link">&para;</a></h2> <p>A confounded RNA-Seq experiment is one where you <strong>cannot distinguish the separate effects of two different sources of variation</strong> in the data.</p> <p>For example, we know that sex has large effects on gene expression, and if all of our <em>control</em> mice were female and all of the <em>treatment</em> mice were male, then our treatment effect would be confounded by sex. <strong>We could not differentiate the effect of treatment from the effect of sex.</strong></p> <p align=center> <img src=./img/02_experimental_planning/confounded_design.png width=500> </p> <p><strong>To AVOID confounding:</strong></p> <ul> <li> <p>Ensure animals in each condition are all the <strong>same sex, age, litter, and batch</strong>, if possible.</p> </li> <li> <p>If not possible, then ensure to split the animals equally between conditions</p> </li> </ul> <p align=center> <img src=./img/02_experimental_planning/non_confounded_design.png width=400> </p> <h2 id=batch-effects>Batch effects<a class=headerlink href=#batch-effects title="Permanent link">&para;</a></h2> <p>A batch effect appears when variance is introduced into your data as a consequence of technical issues such as sample collection, storage, experimental protocol, etc. Batch effects are problematic for RNA-Seq analyses, since you may see significant differences in expression due solely to the batch effect.</p> <p align=center> <img src=./img/02_experimental_planning/batch_effect_pca.png width=600> </p> <p><em>Image credit: <a href=https://www.biorxiv.org/content/early/2015/08/25/025528>Hicks SC, et al., bioRxiv (2015)</a></em></p> <p>To explore the issues generated by poor batch study design, they are highlighted nicely in <a href=https://f1000research.com/articles/4-121/v1>this paper</a>.</p> <h3 id=how-to-know-whether-you-have-batches>How to know whether you have batches?<a class=headerlink href=#how-to-know-whether-you-have-batches title="Permanent link">&para;</a></h3> <ul> <li> <p>Were all RNA isolations performed on the same day?</p> </li> <li> <p>Were all library preparations performed on the same day?</p> </li> <li> <p>Did the same person perform the RNA isolation/library preparation for all samples?</p> </li> <li> <p>Did you use the same reagents for all samples?</p> </li> <li> <p>Did you perform the RNA isolation/library preparation in the same location?</p> </li> </ul> <p>If <em>any</em> of the answers is <strong>'No'</strong>, then you have batches.</p> <h3 id=best-practices-regarding-batches>Best practices regarding batches:<a class=headerlink href=#best-practices-regarding-batches title="Permanent link">&para;</a></h3> <ul> <li> <p>Design the experiment in a way to <strong>avoid batches</strong>, if possible.</p> </li> <li> <p>If unable to avoid batches:</p> <ul> <li><strong>Do NOT confound</strong> your experiment by batch:</li> </ul> </li> </ul> <p align=center> <img src=./img/02_experimental_planning/confounded_batch.png width=300> </p> <p><em>Image credit: <a href=https://www.biorxiv.org/content/early/2015/08/25/025528>Hicks SC, et al., bioRxiv (2015)</a></em></p> <ul> <li><strong>DO</strong> split replicates of the different sample groups across batches. The more replicates the better (definitely more than 2).</li> </ul> <p align=center> <img src=./img/02_experimental_planning/batch_effect.png width=300> </p> <p><em>Image credit: <a href=https://www.biorxiv.org/content/early/2015/08/25/025528>Hicks SC, et al., bioRxiv (2015)</a></em></p> <ul> <li> <p><strong>DO</strong> make a balanced batch design. For example if you can only prepare a subset of samples in the lab on a given day, do not do 90% of samples on day 1 and the remaining 10% on day 2, aim for balance, 50% each day.</p> </li> <li> <p><strong>DO</strong> include batch information in your <strong>experimental metadata</strong>. During the analysis, we can regress out the variation due to batch if not confounded so it doesn't affect our results if we have that information.</p> </li> </ul> <p align=center> <img src=./img/02_experimental_planning/metadata_batch.png width=300> </p> <div class="admonition warning"> <p class=admonition-title>Warning on sample preparations</p> </div> <p>The sample preparation of cell line "biological" replicates "should be performed as independently as possible" (as batches), "meaning that cell culture media should be prepared freshly for each experiment, different frozen cell stocks and growth factor batches, etc. should be used [<a href=http://paasp.net/accurate-design-of-in-vitro-experiments-why-does-it-matter/ >2</a>]." However, preparation across all conditions should be performed at the same time.</p> <hr> <p><em>This lesson was originally developed by members of the teaching team (Mary Piper, Meeta Mistry, Radhika Khetani) at the <a href=http://bioinformatics.sph.harvard.edu/ >Harvard Chan Bioinformatics Core (HBC)</a>.</em></p> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=index.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Course Introduction" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Course Introduction </div> </div> </a> <a href=03_data_explanation.html class="md-footer__link md-footer__link--next" aria-label="Next: Data Explanation" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Data Explanation </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.footer", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=assets/javascripts/bundle.5a2dcb6a.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>