<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="This course is an introduction for how to approach bulk RNAseq data, starting from the sequencing reads.  It will provide an overview of the fundamentals of RNAseq analysis, including read preprocessing, data normalization, data exploration with PCAs and heatmaps, performing differential expression analysis and annotation of the differentially expressed genes.  Participants will also learn how to evaluate confounding and batch effects in the data.  The course will further touch upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, especially about how they influence downstream bioinformatic analysis.   "><meta name=author content="Jose Alejandro Romero Herrera"><link rel=prev href=05a_data_analysis_setup.html><link rel=next href=05c_count_normalization.html><link rel=icon href=img/favicon.png><meta name=generator content="mkdocs-1.4.2, mkdocs-material-9.0.8"><title>RNAseq count matrices - Bulk RNAseq data analysis</title><link rel=stylesheet href=assets/stylesheets/main.0d440cfe.min.css><link rel=stylesheet href=assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#differential-gene-expression-dge-analysis-overview class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bulk RNAseq data analysis </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> RNAseq count matrices </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=02_experimental_planning.html class=md-tabs__link> Experimental planning </a> </li> <li class=md-tabs__item> <a href=03_data_explanation.html class=md-tabs__link> Data Explanation </a> </li> <li class=md-tabs__item> <a href=04a_preprocessing.html class=md-tabs__link> Read Preprocessing </a> </li> <li class=md-tabs__item> <a href=05a_data_analysis_setup.html class="md-tabs__link md-tabs__link--active"> Analysing RNAseq data </a> </li> <li class=md-tabs__item> <a href=09_summarized_workflow.html class=md-tabs__link> Summarized workflow </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> Bulk RNAseq data analysis </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> Course Introduction </a> </li> <li class=md-nav__item> <a href=02_experimental_planning.html class=md-nav__link> Experimental planning </a> </li> <li class=md-nav__item> <a href=03_data_explanation.html class=md-nav__link> Data Explanation </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> Read Preprocessing <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Read Preprocessing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=04a_preprocessing.html class=md-nav__link> Preprocessing steps </a> </li> <li class=md-nav__item> <a href=04b_pipelines.html class=md-nav__link> Nextflow and nf-core pipelines </a> </li> <li class=md-nav__item> <a href=04c_preprocessing_setup.html class=md-nav__link> UCloud setup </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5 checked> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex=0> Analysing RNAseq data <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=true> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Analysing RNAseq data </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=05a_data_analysis_setup.html class=md-nav__link> Ucloud setup </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Count matrix <span class="md-nav__icon md-icon"></span> </label> <a href=05b_count_matrix.html class="md-nav__link md-nav__link--active"> Count matrix </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#setting-up class=md-nav__link> Setting up </a> <nav class=md-nav aria-label="Setting up"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#loading-libraries class=md-nav__link> Loading libraries </a> </li> <li class=md-nav__item> <a href=#loading-data class=md-nav__link> Loading data </a> </li> <li class=md-nav__item> <a href=#viewing-data class=md-nav__link> Viewing data </a> </li> <li class=md-nav__item> <a href=#using-the-abundance-estimates-from-salmon-as-input-to-deseq2 class=md-nav__link> Using the abundance estimates from Salmon as input to DESeq2 </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#differential-gene-expression-analysis-overview class=md-nav__link> Differential gene expression analysis overview </a> <nav class=md-nav aria-label="Differential gene expression analysis overview"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#rna-seq-count-distribution class=md-nav__link> RNA-seq count distribution </a> </li> <li class=md-nav__item> <a href=#modeling-count-data class=md-nav__link> Modeling count data </a> </li> <li class=md-nav__item> <a href=#improving-mean-estimates-ie-reducing-variance-with-biological-replicates class=md-nav__link> Improving mean estimates (i.e. reducing variance) with biological replicates </a> </li> <li class=md-nav__item> <a href=#differential-expression-analysis-workflow class=md-nav__link> Differential expression analysis workflow </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=05c_count_normalization.html class=md-nav__link> Normalization </a> </li> <li class=md-nav__item> <a href=06_exploratory_analysis.html class=md-nav__link> Exploratory Analysis </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_5> <label class=md-nav__link for=__nav_5_5 id=__nav_5_5_label tabindex=0> Differential Expression <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5_5> <span class="md-nav__icon md-icon"></span> Differential Expression </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=07_extra_contrast_design.html class=md-nav__link> Design formula </a> </li> <li class=md-nav__item> <a href=07a_DEA.html class=md-nav__link> DESeq2 </a> </li> <li class=md-nav__item> <a href=07b_hypothesis_testing.html class=md-nav__link> Hypothesis testing </a> </li> <li class=md-nav__item> <a href=07c_DEA_visualization.html class=md-nav__link> Visualization </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_6> <label class=md-nav__link for=__nav_5_6 id=__nav_5_6_label tabindex=0> Functional Analysis <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_6_label aria-expanded=false> <label class=md-nav__title for=__nav_5_6> <span class="md-nav__icon md-icon"></span> Functional Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=08a_FA_genomic_annotation.html class=md-nav__link> Gene annotation </a> </li> <li class=md-nav__item> <a href=08b_FA_overrepresentation.html class=md-nav__link> Over representation analysis </a> </li> <li class=md-nav__item> <a href=08c_FA_GSEA.html class=md-nav__link> Class scoring </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=09_summarized_workflow.html class=md-nav__link> Summarized workflow </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#setting-up class=md-nav__link> Setting up </a> <nav class=md-nav aria-label="Setting up"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#loading-libraries class=md-nav__link> Loading libraries </a> </li> <li class=md-nav__item> <a href=#loading-data class=md-nav__link> Loading data </a> </li> <li class=md-nav__item> <a href=#viewing-data class=md-nav__link> Viewing data </a> </li> <li class=md-nav__item> <a href=#using-the-abundance-estimates-from-salmon-as-input-to-deseq2 class=md-nav__link> Using the abundance estimates from Salmon as input to DESeq2 </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#differential-gene-expression-analysis-overview class=md-nav__link> Differential gene expression analysis overview </a> <nav class=md-nav aria-label="Differential gene expression analysis overview"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#rna-seq-count-distribution class=md-nav__link> RNA-seq count distribution </a> </li> <li class=md-nav__item> <a href=#modeling-count-data class=md-nav__link> Modeling count data </a> </li> <li class=md-nav__item> <a href=#improving-mean-estimates-ie-reducing-variance-with-biological-replicates class=md-nav__link> Improving mean estimates (i.e. reducing variance) with biological replicates </a> </li> <li class=md-nav__item> <a href=#differential-expression-analysis-workflow class=md-nav__link> Differential expression analysis workflow </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=differential-gene-expression-dge-analysis-overview>Differential gene expression (DGE) analysis overview<a class=headerlink href=#differential-gene-expression-dge-analysis-overview title="Permanent link">&para;</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>&#128368; <strong>Time Estimation:</strong> 20 minutes </p> <p>&#128172; <strong>Learning Objectives:</strong> </p> <ol> <li>Describe the RNA-seq and the differential gene expression analysiscworkflow</li> <li>Explain why negative binomial distribution is used to model RNA-seq count data</li> </ol> </div> <p>The goal of RNA-seq is often to perform differential expression testing to determine which genes are expressed at different levels between conditions. These genes can offer biological insight into the processes affected by the condition(s) of interest.</p> <p>To determine the expression levels of genes, our RNA-seq workflow followed the steps detailed in the image below. All steps were performed on the command line (Linux/Unix) through the generation of the read counts per gene. The differential expression analysis and any downstream functional analysis are generally performed in R using R packages specifically designed for the complex statistical analyses required to determine whether genes are differentially expressed.</p> <p><img src=./img/05b_count_matrix/rnaseq_full_workflow.png style="display: block; margin: auto;"></p> <p>In the next few lessons, we will walk you through an <strong>end-to-end gene-level RNA-seq differential expression workflow</strong> using various R packages. We will start with the count matrix, do some exploratory data analysis for quality assessment and explore the relationship between samples. Next, we will perform differential expression analysis, and visually explore the results prior to performing downstream functional analysis.</p> <h2 id=setting-up>Setting up<a class=headerlink href=#setting-up title="Permanent link">&para;</a></h2> <p>Before we get into the details of the analysis, let’s get started by opening up RStudio and setting up a new project for this analysis.</p> <ol> <li>Go to the <code>File</code> menu and select <code>New Project</code>.</li> <li>In the <code>New Project</code> window, choose <code>Existing Directory</code>. Then, choose <code>introduction_bulkRNAseq_analysis</code> as your project working directory.</li> <li>The new project should automatically open in RStudio.</li> </ol> <p>To check whether or not you are in the correct working directory, use <code>getwd()</code>. The path <code>/work/introduction_bulkRNAseq_analysis</code> should be returned to you in the console. When finished your working directory should now look similar to this:</p> <p><img src=./img/05b_count_matrix/settingup.png style="display: block; margin: auto;"></p> <ul> <li> <p>Inside the folder <code>Notebooks</code> you will find the scripts (in <code>Rmd</code> format) that we will follow during the sessions.</p> </li> <li> <p>In the folder <code>Results</code> you will save the results of your scripts, analysis and tests.</p> </li> </ul> <p>To avoid copying the original dataset for each student (very inefficient) the dataset is contained inside the shared folder <code>/work/introduction_bulkRNAseq_analysis/Data/</code>. Do not attempt to modify this folder, as it might mess up the files for the rest of your colleagues.</p> <p>Now you can open the first practical session: <code>06a_count_matrix.Rmd</code></p> <h3 id=loading-libraries>Loading libraries<a class=headerlink href=#loading-libraries title="Permanent link">&para;</a></h3> <p>For this analysis we will be using several R packages, some which have been installed from CRAN and others from Bioconductor. To use these packages (and the functions contained within them), we need to <strong>load the libraries.</strong> Add the following to your script and don’t forget to comment liberally!</p> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=c1>## Setup</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a><span class=c1>### Bioconductor and CRAN libraries used</span>
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a><span class=nf>library</span><span class=p>(</span><span class=n>tidyverse</span><span class=p>)</span>
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=nf>library</span><span class=p>(</span><span class=n>RColorBrewer</span><span class=p>)</span>
<a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a><span class=nf>library</span><span class=p>(</span><span class=n>DESeq2</span><span class=p>)</span>
<a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a><span class=nf>library</span><span class=p>(</span><span class=n>pheatmap</span><span class=p>)</span>
<a id=__codelineno-0-7 name=__codelineno-0-7 href=#__codelineno-0-7></a><span class=nf>library</span><span class=p>(</span><span class=n>DEGreport</span><span class=p>)</span>
<a id=__codelineno-0-8 name=__codelineno-0-8 href=#__codelineno-0-8></a><span class=nf>library</span><span class=p>(</span><span class=n>ggrepel</span><span class=p>)</span>
</code></pre></div> <h3 id=loading-data>Loading data<a class=headerlink href=#loading-data title="Permanent link">&para;</a></h3> <p>To load the data into our current environment, we will be using the <code>read.table</code> function. By default the function expects tab-delimited files, which is what we have.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=c1>## Load in data</span>
<a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a><span class=n>data</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read_table</span><span class=p>(</span><span class=s>&quot;/work/introduction_bulkRNAseq_analysis/Data/Mov10_full_counts.txt&quot;</span><span class=p>)</span><span class=w> </span>
<a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a>
<a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a><span class=n>meta</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read_table</span><span class=p>(</span><span class=s>&quot;/work/introduction_bulkRNAseq_analysis/Data/Mov10_full_meta.txt&quot;</span><span class=p>)</span>
</code></pre></div> <p>Use <code>class()</code> to inspect our data and make sure we are working with data frames:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=c1>### Check classes of the data we just brought in</span>
<a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a><span class=nf>class</span><span class=p>(</span><span class=n>meta</span><span class=p>)</span>
<a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a><span class=nf>class</span><span class=p>(</span><span class=n>data</span><span class=p>)</span>
</code></pre></div> <h3 id=viewing-data>Viewing data<a class=headerlink href=#viewing-data title="Permanent link">&para;</a></h3> <p>Make sure your datasets contain the expected samples / information before proceeding to perfom any type of analysis.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=nf>View</span><span class=p>(</span><span class=n>meta</span><span class=p>)</span>
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=nf>View</span><span class=p>(</span><span class=n>data</span><span class=p>)</span>
</code></pre></div> <blockquote> <h3 id=using-the-abundance-estimates-from-salmon-as-input-to-deseq2>Using the abundance estimates from Salmon as input to DESeq2<a class=headerlink href=#using-the-abundance-estimates-from-salmon-as-input-to-deseq2 title="Permanent link">&para;</a></h3> <p>The counts used in these lessons were generated using the standard approach for RNA-seq analysis, where samples were aligned to the genome using a splice-aware aligner followed by counting. If you are using lightweight algorithms such as Salmon, Sailfish or Kallisto to generate abundance estimates, you can also use DESeq2 to perform gene-level differential expression analysis. These transcript abundance estimates, often referred to as ‘pseudocounts’, can be converted for use with DESeq2 but the setup is slightly more involved. If you are interested in knowing more about using Salmon pseudocounts for DESeq2, there are <a href=https://hbctraining.github.io/DGE_workshop_salmon/lessons/01_DGE_setup_and_overview.html>materials linked here</a>.</p> </blockquote> <h2 id=differential-gene-expression-analysis-overview>Differential gene expression analysis overview<a class=headerlink href=#differential-gene-expression-analysis-overview title="Permanent link">&para;</a></h2> <p>So, what does this count data actually represent? The count data used for differential expression analysis represents the number of sequence reads that originated from a particular gene. The higher the number of counts, the more reads associated with that gene, and the assumption that there was a higher level of expression of that gene in the sample.</p> <p><img src=./img/05b_count_matrix/deseq_counts_overview.png style="display: block; margin: auto;"></p> <p>With differential expression analysis, we are looking for genes that change in expression between two or more groups (defined in the metadata) - case vs. control - correlation of expression with some variable or clinical outcome</p> <p><strong>Why does it not work to identify differentially expressed gene by ranking the genes by how different they are between the two groups (based on fold change values)?</strong></p> <p><img src=./img/05b_count_matrix/foldchange_heatmap.png style="display: block; margin: auto;"></p> <p>Genes that vary in expression level between groups of samples may do so solely as a consequence of the biological variable(s) of interest. However, this difference is often also related to extraneous effects, in fact, sometimes these effects exclusively account for the observed variation. The goal of differential expression analysis to determine the relative role of these effects, hence separating the “interesting” variance from the “uninteresting” variance.</p> <p><img src=./img/05b_count_matrix/de_variation.png style="display: block; margin: auto;"></p> <p>Although the mean expression levels between sample groups may appear to be quite different, it is possible that the difference is not actually significant. This is illustrated for ‘GeneA’ expression between ‘untreated’ and ‘treated’ groups in the figure below. The mean expression level of geneA for the ‘treated’ group is twice as large as for the ‘untreated’ group, but the variation between replicates indicates that this may not be a significant difference. <strong>We need to take into account the variation in the data (and where it might be coming from) when determining whether genes are differentially expressed.</strong></p> <p><img src=./img/05b_count_matrix/de_norm_counts_var.png style="display: block; margin: auto;"></p> <p>Differential expression analysis is used to determine, for each gene, whether the differences in expression (counts) <strong>between groups</strong> is significant given the amount of variation observed <strong>within groups</strong> (replicates). To test for significance, we need an appropriate statistical model that accurately performs normalization (to account for differences in sequencing depth, etc.) and variance modeling (to account for few numbers of replicates and large dynamic expression range).</p> <h3 id=rna-seq-count-distribution>RNA-seq count distribution<a class=headerlink href=#rna-seq-count-distribution title="Permanent link">&para;</a></h3> <p>To determine the appropriate statistical model, we need information about the distribution of counts. To get an idea about how RNA-seq counts are distributed, let’s plot the counts of all the samples:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=n>pdata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>data</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=w>  </span><span class=nf>gather</span><span class=p>(</span><span class=n>key</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>Sample</span><span class=p>,</span><span class=w> </span><span class=n>value</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>Count</span><span class=p>,</span><span class=w> </span><span class=o>-</span><span class=n>GeneSymbol</span><span class=p>)</span>
<a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a>
<a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a><span class=n>pdata</span>
<a id=__codelineno-4-5 name=__codelineno-4-5 href=#__codelineno-4-5></a>
<a id=__codelineno-4-6 name=__codelineno-4-6 href=#__codelineno-4-6></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>pdata</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-4-7 name=__codelineno-4-7 href=#__codelineno-4-7></a><span class=w>  </span><span class=nf>geom_histogram</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>Count</span><span class=p>),</span><span class=w> </span><span class=n>stat</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;bin&quot;</span><span class=p>,</span><span class=w> </span><span class=n>bins</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>200</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-4-8 name=__codelineno-4-8 href=#__codelineno-4-8></a><span class=w>  </span><span class=nf>xlab</span><span class=p>(</span><span class=s>&quot;Raw expression counts&quot;</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-4-9 name=__codelineno-4-9 href=#__codelineno-4-9></a><span class=w>  </span><span class=nf>ylab</span><span class=p>(</span><span class=s>&quot;Number of genes&quot;</span><span class=p>)</span>
</code></pre></div> <p><img src=./img/05b_count_matrix/deseq_counts_distribution.png style="display: block; margin: auto;"></p> <p>If we zoom in close to zero, we can see a large number of genes with counts of zero:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>pdata</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a><span class=w>   </span><span class=nf>geom_histogram</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>Count</span><span class=p>),</span><span class=w> </span><span class=n>stat</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;bin&quot;</span><span class=p>,</span><span class=w> </span><span class=n>bins</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>200</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span>
<a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a><span class=w>   </span><span class=nf>xlim</span><span class=p>(</span><span class=m>-5</span><span class=p>,</span><span class=w> </span><span class=m>500</span><span class=p>)</span><span class=w>  </span><span class=o>+</span>
<a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a><span class=w>   </span><span class=nf>xlab</span><span class=p>(</span><span class=s>&quot;Raw expression counts&quot;</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a><span class=w>   </span><span class=nf>ylab</span><span class=p>(</span><span class=s>&quot;Number of genes&quot;</span><span class=p>)</span>
</code></pre></div> <p><img src=./img/05b_count_matrix/deseq_counts_distribution_zoomed.png style="display: block; margin: auto;"></p> <p>These images illustrate some common features of RNA-seq count data, including a <strong>low number of counts associated with a large proportion of genes</strong>, and a long right tail due to the <strong>lack of any upper limit for expression</strong>. Unlike microarray data, which has a dynamic range maximum limited due to when the probes max out, there is no limit of maximum expression for RNA-seq data. Due to the differences in these technologies, the statistical models used to fit the data are different between the two methods.</p> <blockquote> <p><strong>NOTE:</strong> The log intensities of the microarray data approximate a normal distribution. However, due to the different properties of the of RNA-seq count data, such as integer counts instead of continuous measurements and non-normally distributed data, the normal distribution does not accurately model RNA-seq counts [<a href=https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3541212/ >1</a>].</p> </blockquote> <h3 id=modeling-count-data>Modeling count data<a class=headerlink href=#modeling-count-data title="Permanent link">&para;</a></h3> <p>RNAseq count data can be modeled using a <strong>Poisson distribution</strong>. this particular distribution is fitting for data where the <strong>number of cases is very large but the probability of an event occurring is very small</strong>. To give you an example, think of the lottery: many people buy lottery tickets (high number of cases), but only very few win (the probability of the event is small). <a href=https://youtu.be/fxtB8c3u6l8>Check this video from Rafael Irizarry in the EdX class for more details</a>.</p> <p>With RNA-Seq data, <strong>a very large number of RNAs are represented and the probability of pulling out a particular transcript is very small</strong>. Thus, it would be an appropriate situation to use the Poisson distribution. However, a unique property of this distribution is that the mean == variance. Realistically, with RNA-Seq data there is always some biological variation present across the replicates (within a sample class). Genes with larger average expression levels will tend to have larger observed variances across replicates.</p> <p>The model that fits best, given this type of variability observed for replicates, is the <strong>Negative Binomial (NB) model</strong>. Essentially, <strong>the NB model is a good approximation for data where the mean \&lt; variance</strong>, as is the case with RNA-Seq count data.</p> <p><img src=./img/05b_count_matrix/deseq_nb.png style="display: block; margin: auto;"></p> <blockquote> <p><strong>NOTE:</strong></p> <ul> <li><strong>Biological replicates</strong> represent multiple samples (i.e. RNA from different mice) representing the same sample class</li> <li><strong>Technical replicates</strong> represent the same sample (i.e. RNA from the same mouse) but with technical steps replicated</li> <li>Usually biological variance is much greater than technical variance, so we do not need to account for technical variance to identify biological differences in expression</li> <li><strong>Don’t spend money on technical replicates - biological replicates are much more useful</strong></li> </ul> <p><strong>NOTE:</strong> If you are using <strong>cell lines</strong> and are unsure whether or not you have prepared biological or technical replicates, take a look at <a href=https://web.archive.org/web/20170807192514/http://www.labstats.net:80/articles/cell_culture_n.html>this link</a>. This is a useful resource in helping you determine how best to set up your <em>in-vitro</em> experiment.</p> </blockquote> <details> <summary> **How do I know if my data should be modeled using the Poisson distribution or Negative Binomial distribution??** </summary> If it’s count data, it should fit the negative binomial, as discussed previously. However, it can be helpful to plot the *mean versus the variance* of your data. *Remember for the Poisson model, mean = variance, but for NB, mean \< variance.* Run the following code to plot the *mean versus variance* for our data: <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=n>df</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>data</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=w>  </span><span class=nf>rowwise</span><span class=p>()</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a><span class=w>  </span><span class=nf>summarise</span><span class=p>(</span><span class=n>mean_counts</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>mean</span><span class=p>(</span><span class=n>Mov10_kd_2</span><span class=o>:</span><span class=n>Irrel_kd_3</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a><span class=w>            </span><span class=n>variance_counts</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>var</span><span class=p>(</span><span class=n>Mov10_kd_2</span><span class=o>:</span><span class=n>Irrel_kd_3</span><span class=p>))</span>
<a id=__codelineno-6-5 name=__codelineno-6-5 href=#__codelineno-6-5></a>
<a id=__codelineno-6-6 name=__codelineno-6-6 href=#__codelineno-6-6></a>
<a id=__codelineno-6-7 name=__codelineno-6-7 href=#__codelineno-6-7></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>df</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-6-8 name=__codelineno-6-8 href=#__codelineno-6-8></a><span class=w>  </span><span class=nf>geom_point</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>mean_counts</span><span class=p>,</span><span class=w> </span><span class=n>y</span><span class=o>=</span><span class=n>variance_counts</span><span class=p>))</span><span class=w> </span><span class=o>+</span><span class=w> </span>
<a id=__codelineno-6-9 name=__codelineno-6-9 href=#__codelineno-6-9></a><span class=w>  </span><span class=nf>geom_abline</span><span class=p>(</span><span class=n>intercept</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0</span><span class=p>,</span><span class=w> </span><span class=n>slope</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=p>,</span><span class=w> </span><span class=n>color</span><span class=o>=</span><span class=s>&quot;red&quot;</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-6-10 name=__codelineno-6-10 href=#__codelineno-6-10></a><span class=w>  </span><span class=nf>scale_y_log10</span><span class=p>()</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-6-11 name=__codelineno-6-11 href=#__codelineno-6-11></a><span class=w>  </span><span class=nf>scale_x_log10</span><span class=p>()</span>
</code></pre></div> <img src=./img/05b_count_matrix/deseq_mean_vs_variance.png style="display: block; margin: auto;"> Note that in the above figure, the variance across replicates tends to be greater than the mean (red line), especially for genes with large mean expression levels. *This is a good indication that our data do not fit the Poisson distribution and we need to account for this increase in variance using the Negative Binomial model (i.e. Poisson will underestimate variability leading to an increase in false positive DE genes).* </details> <h3 id=improving-mean-estimates-ie-reducing-variance-with-biological-replicates>Improving mean estimates (i.e. reducing variance) with biological replicates<a class=headerlink href=#improving-mean-estimates-ie-reducing-variance-with-biological-replicates title="Permanent link">&para;</a></h3> <p>The variance or scatter tends to reduce as we increase the number of biological replicates (<em>the distribution will approach the Poisson distribution with increasing numbers of replicates</em>), since standard deviations of averages are smaller than standard deviations of individual observations. <strong>The value of additional replicates is that as you add more data (replicates), you get increasingly precise estimates of group means, and ultimately greater confidence in the ability to distinguish differences between sample classes (i.e. more DE genes).</strong></p> <p>The figure below illustrates the relationship between sequencing depth and number of replicates on the number of differentially expressed genes identified (from <a href=https://doi.org/10.1093/bioinformatics/btt688>Liu et al. (2013)</a>:</p> <p><img src=./img/05b_count_matrix/seqDepth_DEA.png style="display: block; margin: auto;"></p> <p>Note that an <strong>increase in the number of replicates tends to return more DE genes than increasing the sequencing depth</strong>. Therefore, generally more replicates are better than higher sequencing depth, with the caveat that higher depth is required for detection of lowly expressed DE genes and for performing isoform-level differential expression. Generally, the minimum sequencing depth recommended is 20-30 million reads per sample, but we have seen good RNA-seq experiments with 10 million reads if there are a good number of replicates.</p> <p><img src=./img/05b_count_matrix/de_replicates_img.png style="display: block; margin: auto;"></p> <h3 id=differential-expression-analysis-workflow>Differential expression analysis workflow<a class=headerlink href=#differential-expression-analysis-workflow title="Permanent link">&para;</a></h3> <p>To model counts appropriately when performing a differential expression analysis, there are a number of software packages that have been developed for differential expression analysis of RNA-seq data. Even as new methods are continuously being developed a few tools are generally recommended as best practice, like <a href=https://bioconductor.org/packages/release/bioc/html/DESeq2.html><strong>DESeq2</strong></a>, <a href=https://bioconductor.org/packages/release/bioc/html/edgeR.html><strong>EdgeR</strong></a> and <a href=https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29><strong>Limma-Voom</strong></a>.</p> <p>Many studies describing comparisons between these methods show that while there is some agreement, there is also much variability between tools. <strong>Additionally, there is no one method that performs optimally under all conditions</strong> <strong>(<a href=https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-91>Soneson and Dleorenzi, 2013</a>, <a href=https://www.nature.com/articles/s41598-020-76881-x>Corchete et al, 2020</a>)</strong>.</p> <p><img src=./img/05b_count_matrix/deg_methods1.png style="display: block; margin: auto;"></p> <p><img src=./img/05b_count_matrix/deg_methods2.png style="display: block; margin: auto;"></p> <p><strong>We will be using <a href=https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8>DESeq2</a> for the DE analysis, and the analysis steps with DESeq2 are shown in the flowchart below in green</strong>. DESeq2 first normalizes the count data to account for differences in library sizes and RNA composition between samples. Then, we will use the normalized counts to make some plots for QC at the gene and sample level. The final step is to use the appropriate functions from the DESeq2 package to perform the differential expression analysis.</p> <p><img src=./img/05b_count_matrix/deseq_workflow_full_2018.png style="display: block; margin: auto;"></p> <p>We will go in-depth into each of these steps in the following lessons, but additional details and helpful suggestions regarding DESeq2 can be found in the <a href=http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html>DESeq2 vignette</a>. As you go through this workflow and questions arise, you can reference the vignette from within RStudio:</p> <div class=highlight><pre><span></span><code>vignette(&quot;DESeq2&quot;)
</code></pre></div> <p>This is very convenient, as it provides a wealth of information at your fingertips! Be sure to use this as you need during the workshop.</p> <hr> <p><em>This lesson was originally developed by members of the teaching team (Mary Piper, Meeta Mistry, Radhika Khetani) at the <a href=http://bioinformatics.sph.harvard.edu/ >Harvard Chan Bioinformatics Core (HBC)</a>.</em></p> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=05a_data_analysis_setup.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Ucloud setup" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Ucloud setup </div> </div> </a> <a href=05c_count_normalization.html class="md-footer__link md-footer__link--next" aria-label="Next: Normalization" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Normalization </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.footer", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.db81ec45.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=assets/javascripts/bundle.a00a7c5e.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>