<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
    {
      "@context": "https://schema.org",
      "@id": "https://hds-sandbox.github.io/bulk_RNAseq_course/",
      "@type": "TrainingMaterial",
      "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
      "description": "This workshop includes a tutorial on how to approach RNAseq data, starting from your sequencing reads (fastq files). Thus, the workshop only briefly touches upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, mainly for the purpose of outlining considerations in the downstream bioinformatic analysis. This workshop is based on the materials developed by members of the teaching team at the Harvard Chan Bioinformatics Core (HBC), a collection of modified tutorials from the DESeq2, R language vignettes and the nf-core rnaseq pipeline.",
      "keywords": [
        "bioinformatics, RNAseq, bulk, NGS"
      ],
      "learningResourceType": [
        "e-Learning"
      ],
      "license": [
        {
          "@id": "https://creativecommons.org/licenses/by/4.0/",
          "@type": "CreativeWork"
        }
      ],
      "mentions": [
        {
          "@id": "10.5281/zenodo.7660225",
          "@type": "Thing"
        },
        {
          "@id": "10.5281/zenodo.7565997",
          "@type": "Thing"
        }
      ],
      "name": "Introduction to bulk RNAseq analysis workshop"
    }
    </script><title>Exploratory analysis of bulk RNAseq - Bulk RNAseq data analysis</title><link rel=stylesheet href=assets/stylesheets/main.975780f9.min.css><link rel=stylesheet href=assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#exploratory-analysis-and-quality-control class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bulk RNAseq data analysis </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Exploratory analysis of bulk RNAseq </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=02_experimental_planning.html class=md-tabs__link> Experimental planning </a> </li> <li class=md-tabs__item> <a href=03_data_explanation.html class=md-tabs__link> Data Explanation </a> </li> <li class=md-tabs__item> <a href=04a_preprocessing.html class=md-tabs__link> Read Preprocessing </a> </li> <li class=md-tabs__item> <a href=05a_data_analysis_setup.html class="md-tabs__link md-tabs__link--active"> Analysing RNAseq data </a> </li> <li class=md-tabs__item> <a href=09_summarized_workflow.html class=md-tabs__link> Summarized workflow </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> Bulk RNAseq data analysis </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> Course Introduction </a> </li> <li class=md-nav__item> <a href=02_experimental_planning.html class=md-nav__link> Experimental planning </a> </li> <li class=md-nav__item> <a href=03_data_explanation.html class=md-nav__link> Data Explanation </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_4 type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4> Read Preprocessing <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Read Preprocessing" data-md-level=1> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Read Preprocessing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=04a_preprocessing.html class=md-nav__link> Preprocessing steps </a> </li> <li class=md-nav__item> <a href=04b_pipelines.html class=md-nav__link> Nextflow and nf-core pipelines </a> </li> <li class=md-nav__item> <a href=04c_preprocessing_setup.html class=md-nav__link> UCloud setup </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5 type=checkbox id=__nav_5 checked> <label class=md-nav__link for=__nav_5> Analysing RNAseq data <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Analysing RNAseq data" data-md-level=1> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Analysing RNAseq data </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=05a_data_analysis_setup.html class=md-nav__link> Ucloud setup </a> </li> <li class=md-nav__item> <a href=05b_count_matrix.html class=md-nav__link> Count matrix </a> </li> <li class=md-nav__item> <a href=05c_count_normalization.html class=md-nav__link> Normalization </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Exploratory Analysis <span class="md-nav__icon md-icon"></span> </label> <a href=06_exploratory_analysis.html class="md-nav__link md-nav__link--active"> Exploratory Analysis </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#sample-level-qc class=md-nav__link> Sample-level QC </a> </li> <li class=md-nav__item> <a href=#principal-component-analysis-pca class=md-nav__link> Principal Component Analysis (PCA) </a> <nav class=md-nav aria-label="Principal Component Analysis (PCA)"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#interpreting-pca-plots class=md-nav__link> Interpreting PCA plots </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#hierarchical-clustering-heatmap class=md-nav__link> Hierarchical Clustering Heatmap </a> </li> <li class=md-nav__item> <a href=#mov10-quality-assessment-and-exploratory-analysis-using-deseq2 class=md-nav__link> Mov10 quality assessment and exploratory analysis using DESeq2 </a> <nav class=md-nav aria-label="Mov10 quality assessment and exploratory analysis using DESeq2"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#transform-normalized-counts-for-the-mov10-dataset class=md-nav__link> Transform normalized counts for the MOV10 dataset </a> </li> <li class=md-nav__item> <a href=#principal-component-analysis-pca-for-the-mov10-dataset class=md-nav__link> Principal component analysis (PCA) for the MOV10 dataset </a> <nav class=md-nav aria-label="Principal component analysis (PCA) for the MOV10 dataset"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#custom-pca-plot class=md-nav__link> Custom PCA plot </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#hierarchical-clustering-for-the-mov10-dataset class=md-nav__link> Hierarchical Clustering for the MOV10 dataset </a> <nav class=md-nav aria-label="Hierarchical Clustering for the MOV10 dataset"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#custom-heatmap class=md-nav__link> Custom heatmap </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5_5 type=checkbox id=__nav_5_5> <label class=md-nav__link for=__nav_5_5> Differential Expression <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Differential Expression" data-md-level=2> <label class=md-nav__title for=__nav_5_5> <span class="md-nav__icon md-icon"></span> Differential Expression </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=07_extra_contrast_design.html class=md-nav__link> Design formula </a> </li> <li class=md-nav__item> <a href=07a_DEA.html class=md-nav__link> DESeq2 </a> </li> <li class=md-nav__item> <a href=07b_hypothesis_testing.html class=md-nav__link> Hypothesis testing </a> </li> <li class=md-nav__item> <a href=07c_DEA_visualization.html class=md-nav__link> Visualization </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5_6 type=checkbox id=__nav_5_6> <label class=md-nav__link for=__nav_5_6> Functional Analysis <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Functional Analysis" data-md-level=2> <label class=md-nav__title for=__nav_5_6> <span class="md-nav__icon md-icon"></span> Functional Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=08a_FA_genomic_annotation.html class=md-nav__link> Gene annotation </a> </li> <li class=md-nav__item> <a href=08b_FA_overrepresentation.html class=md-nav__link> Over representation analysis </a> </li> <li class=md-nav__item> <a href=08c_FA_GSEA.html class=md-nav__link> Class scoring </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=09_summarized_workflow.html class=md-nav__link> Summarized workflow </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#sample-level-qc class=md-nav__link> Sample-level QC </a> </li> <li class=md-nav__item> <a href=#principal-component-analysis-pca class=md-nav__link> Principal Component Analysis (PCA) </a> <nav class=md-nav aria-label="Principal Component Analysis (PCA)"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#interpreting-pca-plots class=md-nav__link> Interpreting PCA plots </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#hierarchical-clustering-heatmap class=md-nav__link> Hierarchical Clustering Heatmap </a> </li> <li class=md-nav__item> <a href=#mov10-quality-assessment-and-exploratory-analysis-using-deseq2 class=md-nav__link> Mov10 quality assessment and exploratory analysis using DESeq2 </a> <nav class=md-nav aria-label="Mov10 quality assessment and exploratory analysis using DESeq2"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#transform-normalized-counts-for-the-mov10-dataset class=md-nav__link> Transform normalized counts for the MOV10 dataset </a> </li> <li class=md-nav__item> <a href=#principal-component-analysis-pca-for-the-mov10-dataset class=md-nav__link> Principal component analysis (PCA) for the MOV10 dataset </a> <nav class=md-nav aria-label="Principal component analysis (PCA) for the MOV10 dataset"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#custom-pca-plot class=md-nav__link> Custom PCA plot </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#hierarchical-clustering-for-the-mov10-dataset class=md-nav__link> Hierarchical Clustering for the MOV10 dataset </a> <nav class=md-nav aria-label="Hierarchical Clustering for the MOV10 dataset"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#custom-heatmap class=md-nav__link> Custom heatmap </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course/edit/master/docs/06_exploratory_analysis.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1 id=exploratory-analysis-and-quality-control>Exploratory Analysis and Quality Control<a class=headerlink href=#exploratory-analysis-and-quality-control title="Permanent link">&para;</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>&#128368; <strong>Time Estimation:</strong> 80 minutes </p> <p>&#128172; <strong>Learning Objectives:</strong> </p> <ol> <li>Recognize the importance of methods for count data transformation</li> <li>Describe the PCA (principal component analysis) technique</li> <li>Interpret different examples of PCA plots</li> <li>Evaluate sample quality using PCA and hierarchical clustering</li> </ol> </div> <p>The next step in the DESeq2 workflow is QC, which includes sample-level and gene-level steps to perform QC checks on the count data to help us ensure that the samples/replicates look good.</p> <p><img src=./img/06_exploratory_analysis/deseq_workflow_qc_2018.png width=821 style="display: block; margin: auto;"></p> <h2 id=sample-level-qc>Sample-level QC<a class=headerlink href=#sample-level-qc title="Permanent link">&para;</a></h2> <p>A useful initial step in an RNA-seq analysis is often to assess overall similarity between samples:</p> <ul> <li>Which samples are similar to each other, which are different?</li> <li>Does this fit to the expectation from the experiment’s design?</li> <li>What are the major sources of variation in the dataset?</li> </ul> <p>To explore the similarity of our samples, we will be performing sample-level QC using Principal Component Analysis (PCA) and hierarchical clustering methods. These methods/tools allow us to check <strong>how similar the replicates are to each other</strong> (clustering) and <strong>to make sure that the experimental condition is the major source of variation</strong> in the data. Sample-level QC can also help identify any samples behaving like outliers; we can further explore any potential outliers to determine whether they need to be removed prior to DE analysis.</p> <p><img src=./img/06_exploratory_analysis/sample_qc.png width=1362 style="display: block; margin: auto;"></p> <p>These unsupervised clustering methods are run using <strong>log2 transformed normalized counts</strong>. The log2 transformation <strong>improves the sample distances for clustering visualization</strong>, i.e., it reduces the impact of large outlier counts. Instead of using a classical log2 transform, we will be using the <strong>regularized log transform</strong> (rlog). This type of transformation helps to avoid any bias from the abundance of low-count genes; Note1 below explains this in more detail.</p> <p><img src=./img/06_exploratory_analysis/rlog_transformation_new.png width=2699 style="display: block; margin: auto;"></p> <p><em>Image adapted from "<a href=https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf>Beginner’s guide to using the DESeq2 package</a>" by Love, Anders and Huber, 2014</em></p> <div class="admonition note"> <p class=admonition-title>Note</p> <p>Many common statistical methods for exploratory analysis of multidimensional data, especially methods for clustering and ordination (e. g., principal-component analysis and the like), work best for (at least approximately) homoskedastic data; this means that the variance of an observable quantity (i.e., here, the expression strength of a gene) does not depend on the mean. In RNA-Seq data, however, variance grows with the mean. For example, if one performs PCA directly on a matrix of normalized read counts, the result typically depends only on the few most strongly expressed genes because they show the largest absolute differences between samples. A simple and often used strategy to avoid this is to take the logarithm of the normalized count values plus a small pseudocount; however, now the genes with low counts tend to dominate the results because, due to the strong Poisson noise inherent to small count values, they show the strongest relative differences between samples.</p> <p><em>As a solution, DESeq2 offers the regularized-logarithm transformation, or rlog for short. For genes with high counts, the rlog transformation differs not much from an ordinary log2 transformation. For genes with lower counts, however, the values are shrunken towards the genes’ averages across all samples. Using an empirical Bayesian prior in the form of a ridge penality, this is done such that the rlog-transformed data are approximately homoskedastic."</em> - From the "Beginner's guide to using the DESeq2 package" by Love, Anders and Huber, 2014 (the <a href=http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html>DESeq2 vignette</a> is the updated version of this doc).</p> </div> <div class="admonition note"> <p class=admonition-title>Note</p> <p>The <a href=http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html>DESeq2 vignette</a> suggests large datasets (100s of samples) to use the variance-stabilizing transformation (vst) instead of rlog for transformation of the counts, since the rlog function might take too long to run and the <code>vst()</code> function is faster with similar properties to rlog.</p> </div> <h2 id=principal-component-analysis-pca>Principal Component Analysis (PCA)<a class=headerlink href=#principal-component-analysis-pca title="Permanent link">&para;</a></h2> <p>Principal Component Analysis (PCA) is a technique used to represent and visualize the variation in a dataset of high dimensionality. <strong>The number of dimensions, d</strong>, in a dataset may be thought of <strong>as the number of variables</strong> it has, e.g., for an RNA-seq dataset with 20.000 different transcripts, d is 20.000. Principally, this means we would need a dimensional space of size <strong>d</strong> to fully represent that dataset. However, as we are only able to view and comprehend things in 1,2 or 3 dimensions, we would like to project this dataset into a lower dimensional space, a process called <strong>dimensionality reduction</strong>. This makes PCA is a very important technique used in the QC and analysis of both bulk and single-cell RNAseq data, specially because many their dimensions (transcripts) do not contain any information.</p> <p>To better understand how it works, <strong>please go through <a href=https://youtu.be/FgakZw6K1QQ%20that%20explains%20PCA>this YouTube video from StatQuest</a></strong>. After you have gone through the video, please proceed with the interpretation section below.</p> <h3 id=interpreting-pca-plots>Interpreting PCA plots<a class=headerlink href=#interpreting-pca-plots title="Permanent link">&para;</a></h3> <p><strong>Essentially, if two samples have similar levels of expression for the genes that contribute significantly to the variation represented by a given PC (Principal Component), they will be plotted close together on the axis that represents that PC.</strong> Therefore, we would expect that biological replicates to have similar scores (because our expectation is that the same genes are changing) and cluster together. This is easiest to understand by visualizing some example PCA plots.</p> <p>We have an example dataset and a few associated PCA plots below to get a feel for how to interpret them. The metadata for the experiment is displayed below. The main condition of interest is <code>treatment</code>.</p> <p><img src=./img/06_exploratory_analysis/example_metadata.png width=802 style="display: block; margin: auto;"></p> <p>When visualizing on PC1 and PC2, we don’t see the samples separate by <code>treatment</code>, so we decide to explore other sources of variation present in the data. We hope that we have included all possible known sources of variation in our metadata table, and we can use these factors to color the PCA plot.</p> <p><img src=./img/06_exploratory_analysis/example_PCA_treatmentPC1.png width=547 style="display: block; margin: auto;"></p> <p>We start with the factor <code>cage</code>, but the <code>cage</code> factor does not seem to explain the variation on PC1 or PC2.</p> <p><img src=./img/06_exploratory_analysis/example_PCA_cage.png width=547 style="display: block; margin: auto;"></p> <p>Then, we color by the <code>sex</code> factor, which appears to separate samples on PC2. This is good information to take note of, as we can use it downstream to account for the variation due to sex in the model and regress it out.</p> <p><img src=./img/06_exploratory_analysis/example_PCA_sex.png width=547 style="display: block; margin: auto;"></p> <p>Next we explore the <code>strain</code> factor and find that it explains the variation on PC1.</p> <p><img src=./img/06_exploratory_analysis/example_PCA_strain.png width=547 style="display: block; margin: auto;"></p> <p>It’s great that we have been able to identify the sources of variation for both PC1 and PC2. By accounting for it in our model, we should be able to detect more genes differentially expressed due to <code>treatment</code>.</p> <p>Worrisome about this plot is that we see two samples that do not cluster with the correct strain. This would indicate a likely <strong>sample swap</strong> and should be investigated to determine whether these samples are indeed the labeled strains. If we found there was a switch, we could swap the samples in the metadata. However, if we think they are labeled correctly or are unsure, we could just remove the samples from the dataset.</p> <p>Still we haven’t found if <code>treatment</code> is a major source of variation after <code>strain</code> and <code>sex</code>. So, we explore PC3 and PC4 to see if <code>treatment</code> is driving the variation represented by either of these PCs.</p> <p><img src=./img/06_exploratory_analysis/example_PCA_treatmentPC3.png width=547 style="display: block; margin: auto;"></p> <p>We find that the samples separate by <code>treatment</code> on PC3, and are optimistic about our DE analysis since our condition of interest, <code>treatment</code>, is separating on PC3 and we can regress out the variation driving PC1 and PC2.</p> <p>Depending on how much variation is explained by the first few principal components, you <strong>may want to explore more (i.e consider more components and plot pairwise combinations)</strong>. Even if your samples do not separate clearly by the experimental variable, you may still get biologically relevant results from the DE analysis. If you are expecting very small effect sizes, then it’s possible the signal is drowned out by extraneous sources of variation. In situations <strong>where you can identify those sources, it is important to account for these in your model</strong>, as it provides more power to the tool for detecting DE genes.</p> <h2 id=hierarchical-clustering-heatmap>Hierarchical Clustering Heatmap<a class=headerlink href=#hierarchical-clustering-heatmap title="Permanent link">&para;</a></h2> <p>Hierarchical clustering is another method for identifying correlation patterns in a dataset and potential sample outliers. A heatmap displays <strong>the correlation of gene expression for all pairwise combinations of samples</strong> in the dataset. The hierarchical tree along the axes indicates which samples are more similar to each other, i.e. cluster together. The color blocks at the top indicate substructure in the data, and you would expect to see your replicates cluster together as a block for each sample group. Our expectation would be that the samples cluster together similar to the groupings we’ve observed in the PCA plot.</p> <p><img src=./img/06_exploratory_analysis/heatmap_example.png width=707 style="display: block; margin: auto;"></p> <div class="admonition note"> <p class=admonition-title>Note</p> <p>In the example above, we see a clustering of wild-type (Wt) and knock-down (KD) cell line samples and we <strong>would be quite concerned</strong> that the 'Wt_3' and 'KD_3' samples are not clustering with the other replicates. Furthermore, since the majority of genes are not differentially expressed, we observe that the samples generally have high correlations with each other (values higher than 0.80). In this case, samples with correlations below 0.80 may indicate an outlier in your data and/or sample contamination. N.B It is important to stress that these is no universal cut-off for what is a good/bad correlation/distance score, it depends on the particular dataset.</p> </div> <h2 id=mov10-quality-assessment-and-exploratory-analysis-using-deseq2>Mov10 quality assessment and exploratory analysis using DESeq2<a class=headerlink href=#mov10-quality-assessment-and-exploratory-analysis-using-deseq2 title="Permanent link">&para;</a></h2> <p>Now that we have a good understanding of the QC steps normally employed for RNA-seq, let’s implement them for the Mov10 dataset we are going to be working with.</p> <h3 id=transform-normalized-counts-for-the-mov10-dataset>Transform normalized counts for the MOV10 dataset<a class=headerlink href=#transform-normalized-counts-for-the-mov10-dataset title="Permanent link">&para;</a></h3> <p><strong>To improve the distances/clustering for the PCA and hierarchical clustering visualization methods</strong>, we need to moderate the variance across the mean by applying the rlog transformation to the normalized counts.</p> <div class="admonition warning"> <p class=admonition-title>Note on transformed normalized counts</p> <p>The rlog transformation of the normalized counts is only necessary for these visualization methods during this quality assessment. <strong>We will not be using these transformed counts for determining differential expression</strong>.</p> </div> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=c1>### Transform counts for data visualization</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a><span class=n>rld</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>rlog</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>blind</span><span class=o>=</span><span class=kc>TRUE</span><span class=p>)</span>
</code></pre></div> <p>The <code>blind=TRUE</code> argument is to make sure that the <code>rlog()</code> function does not take our sample groups into account - i.e. does the transformation in an unbiased manner. When performing quality assessment, it is important to include this option. The <a href=http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#blind-dispersion-estimation>DESeq2 vignette</a> has more details about this.</p> <p>The <code>rlog()</code> function returns a <code>DESeqTransform</code> object, another type of DESeq-specific object. The reason you don’t just get a matrix of transformed values is because all of the parameters (i.e. size factors) that went into computing the rlog transform are stored in that object. We use this object to plot the PCA and hierarchical clustering figures for quality assessment.</p> <div class="admonition note"> <p class=admonition-title>Performance of rlog vs vst</p> <p>The <code>rlog()</code> function can be a bit slow when you have e.g. &gt; 20 samples. In these situations the <code>vst()</code> function is much faster and performs a similar transformation appropriate for use with <code>plotPCA()</code>. It's typically just a few seconds with <code>vst()</code> due to optimizations and the nature of the transformation.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=c1>### Transform counts for data visualization</span>
<a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a><span class=n>vsd</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>vst</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>blind</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
</code></pre></div> </div> <h3 id=principal-component-analysis-pca-for-the-mov10-dataset>Principal component analysis (PCA) for the MOV10 dataset<a class=headerlink href=#principal-component-analysis-pca-for-the-mov10-dataset title="Permanent link">&para;</a></h3> <p>We are now ready for the QC steps, let’s start with PCA!</p> <p>DESeq2 has a built-in function for generating PCA plots using <code>ggplot2</code> under the hood. This is great because it saves us having to type out lines of code and having to fiddle with the different ggplot2 layers. In addition, it takes the <code>rlog</code> object as an input directly, hence saving us the trouble of extracting the relevant information from it.</p> <p>The function <code>plotPCA()</code> requires two arguments as input: a <code>DESeqTransform</code> object and the "intgroup" (interesting group), i.e. the name of the column in our metadata that has information about the experimental sample groups.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=c1>### Plot PCA </span>
<a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a><span class=nf>plotPCA</span><span class=p>(</span><span class=n>rld</span><span class=p>,</span><span class=w> </span><span class=n>intgroup</span><span class=o>=</span><span class=s>&quot;condition&quot;</span><span class=p>)</span>
</code></pre></div> <p><img src=./img/06_exploratory_analysis/pca_500.png width=413 style="display: block; margin: auto;"></p> <details class=question> <summary><strong>Exercise 1</strong></summary> <p>By default <code>plotPCA()</code> uses the <em>top 500 most variable genes</em>. You can change this by adding the <code>ntop=</code> argument and specifying how many of the genes you want the function to consider. For example, try 1000 genes. Did the plot change a lot?</p> </details> <details class=question> <summary><strong>Exercise 2</strong></summary> <ol> <li>What does the above plot tell you about the similarity of samples?</li> <li>Does it fit the expectation from the experimental design?</li> <li>What do you think the %variance information (in the axes titles) tell you about the data in the context of the PCA?</li> </ol> </details> <h4 id=custom-pca-plot>Custom PCA plot<a class=headerlink href=#custom-pca-plot title="Permanent link">&para;</a></h4> <p>The <code>plotPCA()</code> function will only return the values for PC1 and PC2. If you would like to explore the additional PCs in your data or if you would like to identify genes that contribute most to the PCs, you can use the <code>prcomp()</code> function. For example, to plot any of the PCs we could run the following code:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=c1># Input is a matrix of log transformed values</span>
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=n>rld_mat</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>assay</span><span class=p>(</span><span class=n>rld</span><span class=p>)</span><span class=w> </span><span class=c1># extract rlog count matrix</span>
<a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a><span class=n>pca</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>prcomp</span><span class=p>(</span><span class=nf>t</span><span class=p>(</span><span class=n>rld_mat</span><span class=p>))</span><span class=w> </span><span class=c1># perform PCA on the transposed (t) matrix of data </span>
</code></pre></div> <p>To see what the PCA object contains we can use again the <code>attributes()</code> function</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=nf>attributes</span><span class=p>(</span><span class=n>pca</span><span class=p>)</span>
</code></pre></div> <p>You can check the <code>?prcomp()</code> for more information. The most important variables are: - sdev: standard deviation explained by each PC. - rotation: contribution of each gene to each PC. - x: PC values for each sample (we use this values for our plots).</p> <p>We can create a new object that contains all our metadata information and the PC values</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=n>df</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>cbind</span><span class=p>(</span><span class=n>meta</span><span class=p>,</span><span class=w> </span><span class=n>pca</span><span class=o>$</span><span class=n>x</span><span class=p>)</span><span class=w> </span><span class=c1># Create data frame with metadata and PC3 and PC4 values for input to ggplot</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=c1># ggplot with info for all PCAs</span>
<a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>df</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>geom_point</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>PC3</span><span class=p>,</span><span class=w> </span><span class=n>y</span><span class=o>=</span><span class=n>PC4</span><span class=p>,</span><span class=w> </span><span class=n>color</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>condition</span><span class=p>))</span>
</code></pre></div> <p>If you want to add PC variation information to the plot we can fetch it using the summary() function and take the second row:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=nf>summary</span><span class=p>(</span><span class=n>pca</span><span class=p>)</span>
<a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=n>pca_var</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>summary</span><span class=p>(</span><span class=n>pca</span><span class=p>)</span><span class=o>$</span><span class=n>importance</span><span class=p>[</span><span class=m>2</span><span class=p>,]</span><span class=w> </span><span class=c1># second row is stored in the object &quot;importance&quot;</span>
<a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a><span class=n>pca_var</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>round</span><span class=p>(</span><span class=n>pca_var</span><span class=w> </span><span class=o>*</span><span class=w> </span><span class=m>100</span><span class=p>,</span><span class=w> </span><span class=n>digits</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2</span><span class=p>)</span><span class=w> </span><span class=c1># make it percentage and round to 2 digits</span>
</code></pre></div> <p>Finally, we can add it to our plot</p> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>df</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>geom_point</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>PC3</span><span class=p>,</span><span class=w> </span><span class=n>y</span><span class=o>=</span><span class=n>PC4</span><span class=p>,</span><span class=w> </span><span class=n>color</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>condition</span><span class=p>))</span><span class=w> </span><span class=o>+</span><span class=w> </span>
<a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a><span class=w>  </span><span class=nf>xlab</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=s>&quot;PC3: &quot;</span><span class=p>,</span><span class=n>pca_var</span><span class=p>[</span><span class=s>&quot;PC3&quot;</span><span class=p>],</span><span class=w> </span><span class=s>&quot;% variance&quot;</span><span class=p>))</span><span class=w> </span><span class=o>+</span><span class=w> </span>
<a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a><span class=w>  </span><span class=nf>ylab</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=s>&quot;PC4: &quot;</span><span class=p>,</span><span class=n>pca_var</span><span class=p>[</span><span class=s>&quot;PC4&quot;</span><span class=p>],</span><span class=w> </span><span class=s>&quot;% variance&quot;</span><span class=p>))</span><span class=w> </span>
</code></pre></div> <h3 id=hierarchical-clustering-for-the-mov10-dataset>Hierarchical Clustering for the MOV10 dataset<a class=headerlink href=#hierarchical-clustering-for-the-mov10-dataset title="Permanent link">&para;</a></h3> <p>There is no built-in function in DESeq2 for plotting the heatmap for displaying the pairwise correlation or distances between all the samples and the hierarchical clustering information; we will use the <code>pheatmap()</code> function from the <code>pheatmap</code> package. This function cannot use the <code>DESeqTransform</code> object as input, but requires a matrix or dataframe. So, the first thing to do is retrieve that information from the <code>rld</code> object using a function called <code>assay()</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=c1># Extract the rlog matrix from the object</span>
<a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a><span class=n>rld_mat</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>assay</span><span class=p>(</span><span class=n>rld</span><span class=p>)</span><span class=w>    </span>
</code></pre></div> <p>Next, we need to compute the distances values for all the samples. We can do this using the <code>dist</code> function:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=n>sampleDists</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>dist</span><span class=p>(</span><span class=nf>t</span><span class=p>(</span><span class=n>rld_mat</span><span class=p>))</span><span class=w> </span><span class=c1># Distances are computed by rows, so we need to transpose (t) the matrix</span>
<a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=n>sampleDistMatrix</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>as.matrix</span><span class=p>(</span><span class=n>sampleDists</span><span class=p>)</span>
</code></pre></div> <p>Let’s take a look at the column and row names of the correlation matrix.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=c1># Check the output of sampleDistMatrix, make note of the row names and column names</span>
<a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a><span class=nf>head</span><span class=p>(</span><span class=n>sampleDistMatrix</span><span class=p>)</span>
<a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a>
<a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a><span class=nf>head</span><span class=p>(</span><span class=n>meta</span><span class=p>)</span>
</code></pre></div> <p>You will notice that they match the names we have given our samples in the metadata data frame we started with. It is important that these match, so we can use the <code>annotation</code> argument below to plot a color block across the top. This block enables easy visualization of the hierarchical clustering.</p> <p>Now, let’s plot the heatmap!</p> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=c1># Load pheatmap package</span>
<a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a><span class=nf>library</span><span class=p>(</span><span class=n>pheatmap</span><span class=p>)</span>
<a id=__codelineno-12-3 name=__codelineno-12-3 href=#__codelineno-12-3></a>
<a id=__codelineno-12-4 name=__codelineno-12-4 href=#__codelineno-12-4></a><span class=nf>pheatmap</span><span class=p>(</span><span class=n>sampleDistMatrix</span><span class=p>,</span><span class=w> </span><span class=n>annotation_col</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>meta</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&quot;sample&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-12-5 name=__codelineno-12-5 href=#__codelineno-12-5></a><span class=w>           </span><span class=nf>select</span><span class=p>(</span><span class=n>condition</span><span class=p>))</span><span class=w> </span><span class=c1># we only want to use the condition column as an annotation</span>
</code></pre></div> <p>When you plot using <code>pheatmap()</code> the hierarchical clustering information is used to place similar samples together and this information is represented by the tree structure along the axes. The <code>annotation</code> argument accepts a dataframe as input, in our case it is the <code>meta</code> data frame.</p> <p><img src=./img/06_exploratory_analysis/pheatmap_salmon.png width=832 style="display: block; margin: auto;"></p> <p>Overall, we observe pretty high correlations across the board ( > 0.999) suggesting no outlying sample(s). Also, similar to the PCA plot you see the samples clustering together by sample group. Together, these plots suggest to us that the data are of good quality and we have the green light to proceed to differential expression analysis.</p> <div class="admonition question"> <p class=admonition-title><strong>Exercise 3</strong></p> <p>Instead of using distances between expression patterns, check the Pearson correlation between samples using <code>cor()</code>. Use your rlog count matrix as an input.</p> </div> <h4 id=custom-heatmap>Custom heatmap<a class=headerlink href=#custom-heatmap title="Permanent link">&para;</a></h4> <p>There are many arguments and options for the <code>pheatmap()</code> function. You could, for example, change the color scale used, remove the dendograms, avoid clustering or even scale the values per row or per column.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=nf>library</span><span class=p>(</span><span class=n>RColorBrewer</span><span class=p>)</span>
<a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a><span class=n>heat.colors</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>brewer.pal</span><span class=p>(</span><span class=m>6</span><span class=p>,</span><span class=w> </span><span class=s>&quot;Blues&quot;</span><span class=p>)</span><span class=w> </span><span class=c1># Colors from the RColorBrewer package (only 6)</span>
<a id=__codelineno-13-3 name=__codelineno-13-3 href=#__codelineno-13-3></a><span class=n>heat.colors</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>colorRampPalette</span><span class=p>(</span><span class=n>heat.colors</span><span class=p>)(</span><span class=m>100</span><span class=p>)</span><span class=w> </span><span class=c1># Interpolate 100 colors</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=nf>pheatmap</span><span class=p>(</span><span class=n>sampleDistMatrix</span><span class=p>,</span><span class=w> </span><span class=n>annotation</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>meta</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&quot;sample&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>select</span><span class=p>(</span><span class=s>&quot;condition&quot;</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a><span class=w>         </span><span class=n>color</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>heat.colors</span><span class=p>,</span><span class=w> </span><span class=n>border_color</span><span class=o>=</span><span class=kc>NA</span><span class=p>,</span><span class=w> </span><span class=n>fontsize</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>10</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-14-3 name=__codelineno-14-3 href=#__codelineno-14-3></a><span class=w>         </span><span class=n>fontsize_row</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>10</span><span class=p>,</span><span class=w> </span><span class=n>height</span><span class=o>=</span><span class=m>20</span><span class=p>)</span>
</code></pre></div> <p>You can check all the colors that RColorBrewer offers by using the following command: <code>display.brewer.all()</code></p> <div class=highlight><pre><span></span><code><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=nf>display.brewer.all</span><span class=p>()</span>
</code></pre></div> <hr> <p><em>This lesson was originally developed by members of the teaching team (Mary Piper, Meeta Mistry, Radhika Khetani) at the <a href=http://bioinformatics.sph.harvard.edu/ >Harvard Chan Bioinformatics Core (HBC)</a>.</em></p> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=05c_count_normalization.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Normalization" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Normalization </div> </div> </a> <a href=07_extra_contrast_design.html class="md-footer__link md-footer__link--next" aria-label="Next: Design formula" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Design formula </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=assets/javascripts/bundle.5a2dcb6a.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>