<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
    {
      "@context": "https://schema.org",
      "@id": "https://hds-sandbox.github.io/bulk_RNAseq_course/",
      "@type": "LearningResource",
      "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
      "description": "This workshop includes a tutorial on how to approach RNAseq data, starting from your sequencing reads (fastq files). Thus, the workshop only briefly touches upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, mainly for the purpose of outlining considerations in the downstream bioinformatic analysis. This workshop is based on the materials developed by members of the teaching team at the Harvard Chan Bioinformatics Core (HBC), a collection of modified tutorials from the DESeq2, R language vignettes and the nf-core rnaseq pipeline.",
      "keywords": [
        "bioinformatics, RNAseq, bulk, NGS"
      ],
      "learningResourceType": [
        "e-Learning"
      ],
      "license": [
        {
          "@id": "https://creativecommons.org/licenses/by/4.0/",
          "@type": "CreativeWork"
        }
      ],
      "mentions": [
        {
          "@id": "10.5281/zenodo.7660225",
          "@type": "Thing"
        },
        {
          "@id": "10.5281/zenodo.7565997",
          "@type": "Thing"
        }
      ],
      "name": "Introduction to bulk RNAseq analysis workshop"
    }
    </script><title>Hypothesis testing with DESeq2 - Bulk RNAseq data analysis</title><link rel=stylesheet href=assets/stylesheets/main.975780f9.min.css><link rel=stylesheet href=assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#deseq2-model-fitting-and-hypothesis-testing class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bulk RNAseq data analysis </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Hypothesis testing with DESeq2 </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=02_experimental_planning.html class=md-tabs__link> Experimental planning </a> </li> <li class=md-tabs__item> <a href=03_data_explanation.html class=md-tabs__link> Data Explanation </a> </li> <li class=md-tabs__item> <a href=04a_preprocessing.html class=md-tabs__link> Read Preprocessing </a> </li> <li class=md-tabs__item> <a href=05a_data_analysis_setup.html class="md-tabs__link md-tabs__link--active"> Analysing RNAseq data </a> </li> <li class=md-tabs__item> <a href=09_summarized_workflow.html class=md-tabs__link> Summarized workflow </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> Bulk RNAseq data analysis </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> Course Introduction </a> </li> <li class=md-nav__item> <a href=02_experimental_planning.html class=md-nav__link> Experimental planning </a> </li> <li class=md-nav__item> <a href=03_data_explanation.html class=md-nav__link> Data Explanation </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_4 type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4> Read Preprocessing <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Read Preprocessing" data-md-level=1> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Read Preprocessing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=04a_preprocessing.html class=md-nav__link> Preprocessing steps </a> </li> <li class=md-nav__item> <a href=04b_pipelines.html class=md-nav__link> Nextflow and nf-core pipelines </a> </li> <li class=md-nav__item> <a href=04c_preprocessing_setup.html class=md-nav__link> UCloud setup </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5 type=checkbox id=__nav_5 checked> <label class=md-nav__link for=__nav_5> Analysing RNAseq data <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Analysing RNAseq data" data-md-level=1> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Analysing RNAseq data </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=05a_data_analysis_setup.html class=md-nav__link> Ucloud setup </a> </li> <li class=md-nav__item> <a href=05b_count_matrix.html class=md-nav__link> Count matrix </a> </li> <li class=md-nav__item> <a href=05c_count_normalization.html class=md-nav__link> Normalization </a> </li> <li class=md-nav__item> <a href=06_exploratory_analysis.html class=md-nav__link> Exploratory Analysis </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5_5 type=checkbox id=__nav_5_5 checked> <label class=md-nav__link for=__nav_5_5> Differential Expression <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Differential Expression" data-md-level=2> <label class=md-nav__title for=__nav_5_5> <span class="md-nav__icon md-icon"></span> Differential Expression </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=07_extra_contrast_design.html class=md-nav__link> Design formula </a> </li> <li class=md-nav__item> <a href=07a_DEA.html class=md-nav__link> DESeq2 </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Hypothesis testing <span class="md-nav__icon md-icon"></span> </label> <a href=07b_hypothesis_testing.html class="md-nav__link md-nav__link--active"> Hypothesis testing </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#generalized-linear-model class=md-nav__link> Generalized Linear Model </a> </li> <li class=md-nav__item> <a href=#hypothesis-testing class=md-nav__link> Hypothesis testing </a> <nav class=md-nav aria-label="Hypothesis testing"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#wald-test class=md-nav__link> Wald test </a> </li> <li class=md-nav__item> <a href=#likelihood-ratio-test-lrt class=md-nav__link> Likelihood ratio test (LRT) </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#multiple-test-correction class=md-nav__link> Multiple test correction </a> <nav class=md-nav aria-label="Multiple test correction"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#what-does-the-p-value-mean class=md-nav__link> What does the p-value mean? </a> </li> <li class=md-nav__item> <a href=#correcting-the-p-value-for-multiple-testing class=md-nav__link> Correcting the p-value for multiple testing </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#exploring-results-wald-test class=md-nav__link> Exploring Results (Wald test) </a> <nav class=md-nav aria-label="Exploring Results (Wald test)"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#specifying-contrasts class=md-nav__link> Specifying contrasts </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#the-results-table class=md-nav__link> The results table </a> </li> <li class=md-nav__item> <a href=#p-values class=md-nav__link> P-values </a> <nav class=md-nav aria-label=P-values> <ul class=md-nav__list> <li class=md-nav__item> <a href=#gene-level-filtering class=md-nav__link> Gene-level filtering </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#fold-changes class=md-nav__link> Fold changes </a> </li> <li class=md-nav__item> <a href=#summarizing-results class=md-nav__link> Summarizing results </a> </li> <li class=md-nav__item> <a href=#extracting-significant-differentially-expressed-genes class=md-nav__link> Extracting significant differentially expressed genes </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=07c_DEA_visualization.html class=md-nav__link> Visualization </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5_6 type=checkbox id=__nav_5_6> <label class=md-nav__link for=__nav_5_6> Functional Analysis <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Functional Analysis" data-md-level=2> <label class=md-nav__title for=__nav_5_6> <span class="md-nav__icon md-icon"></span> Functional Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=08a_FA_genomic_annotation.html class=md-nav__link> Gene annotation </a> </li> <li class=md-nav__item> <a href=08b_FA_overrepresentation.html class=md-nav__link> Over representation analysis </a> </li> <li class=md-nav__item> <a href=08c_FA_GSEA.html class=md-nav__link> Class scoring </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=09_summarized_workflow.html class=md-nav__link> Summarized workflow </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#generalized-linear-model class=md-nav__link> Generalized Linear Model </a> </li> <li class=md-nav__item> <a href=#hypothesis-testing class=md-nav__link> Hypothesis testing </a> <nav class=md-nav aria-label="Hypothesis testing"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#wald-test class=md-nav__link> Wald test </a> </li> <li class=md-nav__item> <a href=#likelihood-ratio-test-lrt class=md-nav__link> Likelihood ratio test (LRT) </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#multiple-test-correction class=md-nav__link> Multiple test correction </a> <nav class=md-nav aria-label="Multiple test correction"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#what-does-the-p-value-mean class=md-nav__link> What does the p-value mean? </a> </li> <li class=md-nav__item> <a href=#correcting-the-p-value-for-multiple-testing class=md-nav__link> Correcting the p-value for multiple testing </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#exploring-results-wald-test class=md-nav__link> Exploring Results (Wald test) </a> <nav class=md-nav aria-label="Exploring Results (Wald test)"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#specifying-contrasts class=md-nav__link> Specifying contrasts </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#the-results-table class=md-nav__link> The results table </a> </li> <li class=md-nav__item> <a href=#p-values class=md-nav__link> P-values </a> <nav class=md-nav aria-label=P-values> <ul class=md-nav__list> <li class=md-nav__item> <a href=#gene-level-filtering class=md-nav__link> Gene-level filtering </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#fold-changes class=md-nav__link> Fold changes </a> </li> <li class=md-nav__item> <a href=#summarizing-results class=md-nav__link> Summarizing results </a> </li> <li class=md-nav__item> <a href=#extracting-significant-differentially-expressed-genes class=md-nav__link> Extracting significant differentially expressed genes </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course/edit/master/docs/07b_hypothesis_testing.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1 id=deseq2-model-fitting-and-hypothesis-testing>DESeq2: Model fitting and Hypothesis testing<a class=headerlink href=#deseq2-model-fitting-and-hypothesis-testing title="Permanent link">&para;</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>&#128368; <strong>Time Estimation:</strong> 90 minutes </p> <p>&#128172; <strong>Learning Objectives:</strong> </p> <ol> <li>Demonstrate the use of the design formula with simple and complex designs</li> <li>Construct R code to execute the differential expression analysis workflow with DESeq2</li> <li>Describe the process of model fitting</li> <li>Compare two methods for hypothesis testing (Wald test vs. LRT)</li> <li>Discuss the steps required to generate a results table for pairwise comparisons (Wald test)</li> <li>Recognize the importance of multiple test correction</li> <li>Identify different methods for multiple test correction</li> <li>Summarize the different levels of gene filtering</li> <li>Evaluate the number of differentially expressed genes produced for each comparison</li> <li>Construct R objects containing significant genes from each comparison</li> </ol> </div> <p>The final step in the DESeq2 workflow is taking the counts for each gene and fitting it to the model and testing for differential expression.</p> <p><img src=./img/07b_hypothesis_testing/de_workflow.png width=509 style="display: block; margin: auto;"></p> <h2 id=generalized-linear-model>Generalized Linear Model<a class=headerlink href=#generalized-linear-model title="Permanent link">&para;</a></h2> <p>As described <a href=05b_count_matrix.html>earlier</a>, the count data generated by RNA-seq exhibits overdispersion (variance > mean) and the statistical distribution used to model the counts needs to account for this. As such, DESeq2 uses a <strong>negative binomial distribution to model the RNA-seq counts using the equation below</strong>:</p> <p><img src=./img/07b_hypothesis_testing/NB_model_formula.png width=783 style="display: block; margin: auto;"></p> <p>The two parameters required are the <strong>size factor, and the dispersion estimate</strong>. Next, a generalized linear model (GLM) of the NB family is used to fit the data. Modeling is a mathematically formalized way to approximate how the data behaves given a set of parameters.</p> <details class=note> <summary>Note on GLM</summary> <p>In statistics, the generalized linear model (GLM) is a flexible generalization of ordinary linear regression that allows for response variables that have error distribution models other than a normal distribution. The GLM generalizes linear regression by allowing the linear model to be related to the response variable via a link function and by allowing the magnitude of the variance of each measurement to be a function of its predicted value." (<a href=https://en.wikipedia.org/wiki/Generalized_linear_model>Wikipedia</a>).</p> </details> <p>After the model is fit, coefficients are estimated for each sample group along with their standard error. The coefficents are the estimates for the <strong>log2 fold changes</strong>, and will be used as input for hypothesis testing.</p> <h2 id=hypothesis-testing>Hypothesis testing<a class=headerlink href=#hypothesis-testing title="Permanent link">&para;</a></h2> <p>The first step in hypothesis testing is to set up a <strong>null hypothesis</strong> for each gene. In our case, the null hypothesis is that there is <strong>no differential expression across the two sample groups (LFC == 0)</strong>. Notice that we can do this without observing any data, because it is based on a thought experiment. Second, we use a statistical test to determine if based on the observed data, the null hypothesis can be rejected.</p> <h3 id=wald-test>Wald test<a class=headerlink href=#wald-test title="Permanent link">&para;</a></h3> <p>In DESeq2, the <strong>Wald test is the default used for hypothesis testing when comparing two groups</strong>. The Wald test is a test usually performed on parameters that have been estimated by maximum likelihood. In our case we are testing each gene model coefficient (LFC) which was derived using parameters like dispersion, which were estimated using maximum likelihood.</p> <p>DESeq2 implements the Wald test by: * Taking the LFC and dividing it by its standard error, resulting in a z-statistic * The z-statistic is compared to a standard normal distribution, and a p-value is computed reporting the probability that a z-statistic at least as extreme as the observed value would be selected at random * If the p-value is small we reject the null hypothesis and state that there is evidence against the null (i.e. the gene is differentially expressed).</p> <p>The <strong>model fit and Wald test were already run previously as part of the <code>DESeq()</code> function</strong>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=c1>## DO NOT RUN THIS CODE</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a>
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a><span class=c1>## Create DESeq2Dataset object</span>
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=n>dds</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DESeqDataSetFromTximport</span><span class=p>(</span><span class=n>txi</span><span class=p>,</span>
<a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a><span class=w>                                   </span><span class=n>colData</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>meta</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&quot;sample&quot;</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a><span class=w>                              </span><span class=n>design</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=o>~</span><span class=w> </span><span class=n>condition</span><span class=p>)</span>
<a id=__codelineno-0-7 name=__codelineno-0-7 href=#__codelineno-0-7></a><span class=c1>## Run analysis</span>
<a id=__codelineno-0-8 name=__codelineno-0-8 href=#__codelineno-0-8></a><span class=n>dds</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DESeq</span><span class=p>(</span><span class=n>dds</span><span class=p>)</span>
</code></pre></div> <h3 id=likelihood-ratio-test-lrt>Likelihood ratio test (LRT)<a class=headerlink href=#likelihood-ratio-test-lrt title="Permanent link">&para;</a></h3> <p>An alternative to pair-wise comparisons is to <strong>analyze all levels of a factor at once</strong>. By default, the Wald test is used to generate the results table, but DESeq2 also offers the LRT which is used to identify any genes that show change in expression across the different levels. This type of test can be especially useful in analyzing time course experiments.</p> <div class="admonition note"> <p class=admonition-title>LRT vs Wald test</p> <p>The <strong>Wald test</strong> evaluates whether a gene's expression is up- or down-regulated in one group compared to another, while the LRT <strong>identifies genes that are changing in expression in any combination across the different sample groups</strong>. For example:</p> <ul> <li>Gene expression is equal between Control and Overexpression, but is less in Knockdown</li> <li>Gene expression of Overexpression is lower than Control and Control is lower than Knockdown</li> <li>Gene expression of Knockdown is equal to Control, but less than Overexpression.</li> </ul> <p>Using LRT, we are evaluating the <strong>null hypothesis that the full model fits just as well as the reduced model</strong>. If we reject the null hypothesis, this suggests that there is a significant amount of variation explained by the full model (and our main factor of interest), therefore the gene is differentially expressed across the different levels. Generally, this test will result in a larger number of genes than the individual pair-wise comparisons. However, while the LRT is a test of significance for differences of any level of the factor, one should not expect it to be exactly equal to the union of sets of genes using Wald tests (although we do expect a majority overlap).</p> </div> <p>To use the LRT, we use the <code>DESeq()</code> function but this time adding two arguments:</p> <ol> <li>specifying that we want to use the LRT test</li> <li>the ‘reduced’ model</li> </ol> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=c1># The full model was specified previously with the `design = ~ condition`:</span>
<a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a><span class=c1># dds &lt;- DESeqDataSetFromTximport(txi,</span>
<a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a><span class=w>                                   </span><span class=c1># colData = meta %&gt;% column_to_rownames(&quot;sample&quot;), </span>
<a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a><span class=w>                              </span><span class=c1># design = ~ condition)</span>
<a id=__codelineno-1-5 name=__codelineno-1-5 href=#__codelineno-1-5></a>
<a id=__codelineno-1-6 name=__codelineno-1-6 href=#__codelineno-1-6></a><span class=c1># Likelihood ratio test</span>
<a id=__codelineno-1-7 name=__codelineno-1-7 href=#__codelineno-1-7></a><span class=n>dds_lrt</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DESeq</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>test</span><span class=o>=</span><span class=s>&quot;LRT&quot;</span><span class=p>,</span><span class=w> </span><span class=n>reduced</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=o>~</span><span class=w> </span><span class=m>1</span><span class=p>)</span>
</code></pre></div> <p>Since our ‘full’ model only has one factor (<code>sampletype</code>), the ‘reduced’ model (removing that factor) is just the intercept (~ 1). You will find that similar columns are reported for the LRT test. One thing to note is, <strong>even though there are fold changes present they are not directly associated with the actual hypothesis test.</strong></p> <details class=info> <summary>Time course analysis with LTR</summary> <p>The LRT test can be especially helpful when performing time course analyses. We can use the LRT to explore whether there are any significant differences in treatment effect between any of the timepoints.</p> <p>For have an experiment looking at the effect of treatment over time on mice of two different genotypes. We could use a design formula for our 'full model' that would include the major sources of variation in our data: genotype, treatment, time, and our main condition of interest, which is the difference in the effect of treatment over time (treatment:time).</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=n>full_model</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=o>~</span><span class=w> </span><span class=n>genotype</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>treatment</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>time</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>treatment</span><span class=o>:</span><span class=n>time</span>
</code></pre></div> <p>To perform the LRT test, we can determine all genes that have significant differences in expression between treatments across time points by giving the 'reduced model' without the treatment:time term:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=n>reduced_model</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=o>~</span><span class=w> </span><span class=n>genotype</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>treatment</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>time</span>
</code></pre></div> <p>Then, we could run our test by using the following code:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=n>dds_lrt</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DESeqDataSetFromMatrix</span><span class=p>(</span><span class=n>countData</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>data</span><span class=p>,</span><span class=w> </span><span class=n>colData</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>metadata</span><span class=p>,</span><span class=w> </span><span class=n>design</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=o>~</span><span class=w> </span><span class=n>genotype</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>treatment</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>time</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>treatment</span><span class=o>:</span><span class=n>time</span><span class=p>)</span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a>
<a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a><span class=n>dds_lrt_time</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DESeq</span><span class=p>(</span><span class=n>dds_lrt</span><span class=p>,</span><span class=w> </span><span class=n>test</span><span class=o>=</span><span class=s>&quot;LRT&quot;</span><span class=p>,</span><span class=w> </span><span class=n>reduced</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=o>~</span><span class=w> </span><span class=n>genotype</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>treatment</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>time</span><span class=p>)</span>
</code></pre></div> <p>This analysis will not return genes where the treatment effect does not change over time, even though the genes may be differentially expressed between groups at a particular time point, as shown in the figure below:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=n>knitr</span><span class=o>::</span><span class=nf>include_graphics</span><span class=p>(</span><span class=s>&quot;./img/07b_hypothesis_testing/lrt_time_nodiff.png&quot;</span><span class=p>)</span>
</code></pre></div> <p>The significant DE genes will represent those genes that have differences in the effect of treatment over time, an example is displayed in the figure below:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=n>knitr</span><span class=o>::</span><span class=nf>include_graphics</span><span class=p>(</span><span class=s>&quot;./img/07b_hypothesis_testing/lrt_time_yesdiff.png&quot;</span><span class=p>)</span>
</code></pre></div> </details> <div class="admonition question"> <p class=admonition-title><strong>Exercise 1</strong></p> <p>You are studying brain maturation and growth patterns in mouse cortex and have obtained RNA-seq data for a total of 31 mice. These samples were acquired at 9 developmental stages during the postnatal period of 2-40 days of growth, with at least three replicates at each stage. You also have sex information for these mice (16 males and 15 females).</p> <ol> <li>What is an appropriate hypothesis test if you are testing for expression differences across the developmental stages?</li> <li>Provide the line of code used to create the <code>dds</code> object.</li> <li>Provide the line of code used to run DESeq2.</li> <li>The results of the differential expression analysis run identifies a group of genes that spike in expression between the first and second timepoints with no change in expression thereafter. How would we go about obtaining fold changes for these genes?</li> </ol> </div> <h2 id=multiple-test-correction>Multiple test correction<a class=headerlink href=#multiple-test-correction title="Permanent link">&para;</a></h2> <p>Regardless of whether we use the Wald test or the LRT, each gene that has been tested will be associated with a p-value. It is this result which we use to determine which genes are considered significantly differentially expressed. However, <strong>we cannot use the p-value directly.</strong></p> <h3 id=what-does-the-p-value-mean>What does the p-value mean?<a class=headerlink href=#what-does-the-p-value-mean title="Permanent link">&para;</a></h3> <p>A gene with a significance cut-off of p \&lt; 0.05, means there is a 5% chance it is a false positive. For example, if we test 20,000 genes for differential expression, at p \&lt; 0.05 we would expect to find 1,000 DE genes by chance. If we found 3000 genes to be differentially expressed total, roughly one third of our genes are false positives! We would not want to sift through our "significant" genes to identify which ones are true positives.</p> <p>Since each p-value is the result of a single test (single gene). The more genes we test, the more we inflate the false positive rate. <strong>This is the multiple testing problem.</strong></p> <h3 id=correcting-the-p-value-for-multiple-testing>Correcting the p-value for multiple testing<a class=headerlink href=#correcting-the-p-value-for-multiple-testing title="Permanent link">&para;</a></h3> <p>There are a few common approaches for multiple test correction:</p> <ul> <li><strong>Bonferroni:</strong> The adjusted p-value is calculated by: p-value * m (m = total number of tests). <strong>This is a very conservative approach with a high probability of false negatives</strong>, so is generally not recommended.</li> <li><strong>FDR/Benjamini-Hochberg:</strong> Benjamini and Hochberg (1995) defined the concept of False Discovery Rate (FDR) and created an algorithm to control the expected FDR below a specified level given a list of independent p-values. <a href=https://www.statisticshowto.com/benjamini-hochberg-procedure/ >More info about BH</a>.</li> <li><strong>Q-value / Storey method:</strong> The minimum FDR that can be attained when calling that feature significant. For example, if gene X has a q-value of 0.013 it means that 1.3% of genes that show p-values at least as small as gene X are false positives.</li> </ul> <p>DESeq2 helps reduce the number of genes tested by removing those genes unlikely to be significantly DE prior to testing, such as those with low number of counts and outlier samples (<a href=07b_hypothesis_testing.html#gene-level-filtering>see below</a>). However, multiple test correction is also implemented to reduce the False Discovery Rate using an interpretation of the Benjamini-Hochberg procedure.</p> <div class="admonition info"> <p class=admonition-title>So what does FDR \&lt; 0.05 mean?</p> <p>By setting the FDR cutoff to &lt; 0.05, we're saying that the proportion of false positives we expect amongst our differentially expressed genes is 5%. For example, if you call 500 genes as differentially expressed with an FDR cutoff of 0.05, you expect 25 of them to be false positives.</p> </div> <h2 id=exploring-results-wald-test>Exploring Results (Wald test)<a class=headerlink href=#exploring-results-wald-test title="Permanent link">&para;</a></h2> <p>By default DESeq2 uses the Wald test to identify genes that are differentially expressed between two sample groups. Given the factor(s) used in the design formula, and how many factor levels are present, we can extract results for a number of different comparisons. Here, we will walk you through how to obtain results from the <code>dds</code> object and provide some explanations on how to interpret them.</p> <div class="admonition note"> <p class=admonition-title>Wald test on continuous variables</p> <p>The Wald test can also be used with <strong>continuous variables</strong>. If the variable of interest provided in the design formula is continuous-valued, then the reported <code>log2FoldChange</code> is per unit of change of that variable.</p> </div> <h3 id=specifying-contrasts>Specifying contrasts<a class=headerlink href=#specifying-contrasts title="Permanent link">&para;</a></h3> <p>In our dataset, we have three sample groups so we can make three possible pairwise comparisons:</p> <ol> <li>Control vs. Mov10 overexpression</li> <li>Control vs. Mov10 knockdown</li> <li>Mov10 knockdown vs. Mov10 overexpression</li> </ol> <p><strong>We are really only interested in #1 and #2 from above</strong>. When we initially created our <code>dds</code> object we had provided <code>~ condition</code> as our design formula, indicating that <code>sampletype</code> is our main factor of interest.</p> <p>To indicate which two sample groups we are interested in comparing, we need to specify <strong>contrasts</strong>. The contrasts are used as input to the DESeq2 <code>results()</code> function to extract the desired results.</p> <p>Contrasts can be specified in three different ways:</p> <ol> <li>Contrasts can be supplied as a <strong>character vector with exactly three elements</strong>: the name of the factor (of interest) in the design formula, the name of the two factors levels to compare. The factor level given last is the base level for the comparison. The syntax is given below:</li> </ol> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1># DO NOT RUN!</span>
<a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=n>contrast</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;condition&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;level_to_compare&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;base_level&quot;</span><span class=p>)</span>
<a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a><span class=nf>results</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>contrast</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>contrast</span><span class=p>)</span>
</code></pre></div> <ol> <li>Contrasts can be given as a <strong>list of 2 character vectors</strong>: the names of the fold changes for the level of ineterest, and the names of the fold changes for the base level. These names should match identically to the elements of <code>resultsNames(object)</code>. <em>This method can be useful for combining interaction terms and main effects.</em></li> </ol> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=c1># DO NOT RUN!</span>
<a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a><span class=nf>resultsNames</span><span class=p>(</span><span class=n>dds</span><span class=p>)</span><span class=w> </span><span class=c1># to see what names to use</span>
<a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a><span class=n>contrast</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>list</span><span class=p>(</span><span class=nf>resultsNames</span><span class=p>(</span><span class=n>dds</span><span class=p>)[</span><span class=m>1</span><span class=p>],</span><span class=w> </span><span class=nf>resultsNames</span><span class=p>(</span><span class=n>dds</span><span class=p>)[</span><span class=m>2</span><span class=p>])</span>
<a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a><span class=nf>results</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>contrast</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>contrast</span><span class=p>)</span>
</code></pre></div> <ol> <li>One of the results from <code>resultsNames(dds)</code> and the <code>name</code> argument. This one is the simplest but it can also be very restricted:</li> </ol> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=c1># DO NOT RUN!</span>
<a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a><span class=nf>resultsNames</span><span class=p>(</span><span class=n>dds</span><span class=p>)</span><span class=w> </span><span class=c1># to see what names to use</span>
<a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a><span class=nf>results</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>name</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>resultsNames</span><span class=p>(</span><span class=n>dds</span><span class=p>)[</span><span class=m>2</span><span class=p>])</span>
</code></pre></div> <p>Alternatively, if you <strong>only had two factor levels you could do nothing</strong> and not worry about specifying contrasts (i.e. <code>results(dds)</code>). In this case, DESeq2 will choose what your base factor level based on alphabetical order of the levels.</p> <p>To start, we want to evaluate <strong>expression changes between the MOV10 overexpression samples and the control samples</strong>. As such we will use the first method for specifying contrasts and create a character vector:</p> <div class="admonition question"> <p class=admonition-title><strong>Exercise 1</strong></p> <p>Define contrasts for MOV10 overexpression using one of the two methods above.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=c1>## your code here</span>
<a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=n>contrast_oe</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span>
</code></pre></div> </div> <div class="admonition warning"> <p class=admonition-title>Does it matter what I choose to be my base level?</p> <p>Yes, it does matter. <strong>Deciding what level is the base level will determine how to interpret the fold change that is reported.</strong> So for example, if we observe a log2 fold change of -2 this would mean the gene expression is lower in factor level of interest relative to the base level. Thus, if leaving it up to DESeq2 to decide on the contrasts be sure to check that the alphabetical order coincides with the fold change direction you are anticipating.</p> </div> <h2 id=the-results-table>The results table<a class=headerlink href=#the-results-table title="Permanent link">&para;</a></h2> <p>Now that we have our contrast created, we can use it as input to the <code>results()</code> function. Let’s take a quick look at the help manual for the function:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=o>?</span><span class=n>results</span>
</code></pre></div> <p>You will see we have the option to provide a wide array of arguments and tweak things from the defaults as needed. As we go through the lesson we will keep coming back to the help documentation to discuss some arguments that are good to know about.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=c1>## Extract results for MOV10 overexpression vs control</span>
<a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a><span class=n>res_tableOE</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>results</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>contrast</span><span class=o>=</span><span class=n>contrast_oe</span><span class=p>,</span><span class=w> </span><span class=n>alpha</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.05</span><span class=p>)</span>
</code></pre></div> <div class="admonition warning"> <p class=admonition-title>Warning</p> <p>For our analysis, in addition to the <code>contrast</code> argument we will also provide a value of 0.05 for the <code>alpha</code> argument. We will describe this in more detail when we talk about gene-level filtering.</p> </div> <p>The results table that is returned to us is <strong>a <code>DESeqResults</code> object</strong>, which is a simple subclass of DataFrame. In many ways it can be treated like a dataframe (i.e when accessing/subsetting data), however it is important to recognize that there are differences for downstream steps like visualization.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=c1># Check what type of object is returned</span>
<a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a><span class=nf>class</span><span class=p>(</span><span class=n>res_tableOE</span><span class=p>)</span>
</code></pre></div> <p>Now let’s take a look at <strong>what information is stored</strong> in the results:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=c1># What is stored in results?</span>
<a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a><span class=n>res_tableOE</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-14-3 name=__codelineno-14-3 href=#__codelineno-14-3></a><span class=w>  </span><span class=nf>data.frame</span><span class=p>()</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-14-4 name=__codelineno-14-4 href=#__codelineno-14-4></a><span class=w>  </span><span class=nf>head</span><span class=p>()</span>
</code></pre></div> <p>We have six columns of information reported for each gene (row). We can use the <code>mcols()</code> function to extract information on what the values stored in each column represent:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=c1># Get information on each column in results</span>
<a id=__codelineno-15-2 name=__codelineno-15-2 href=#__codelineno-15-2></a><span class=nf>mcols</span><span class=p>(</span><span class=n>res_tableOE</span><span class=p>,</span><span class=w> </span><span class=n>use.names</span><span class=o>=</span><span class=bp>T</span><span class=p>)</span>
</code></pre></div> <ul> <li><code>baseMean</code>: mean of normalized counts for all samples</li> <li><code>log2FoldChange</code>: log2 fold change</li> <li><code>lfcSE</code>: standard error</li> <li><code>stat</code>: Wald statistic</li> <li><code>pvalue</code>: Wald test p-value</li> <li><code>padj</code>: BH adjusted p-values</li> </ul> <h2 id=p-values>P-values<a class=headerlink href=#p-values title="Permanent link">&para;</a></h2> <p>The p-value is a probability value used to determine whether there is evidence to reject the null hypothesis. <strong>A smaller p-value means that there is stronger evidence in favor of the alternative hypothesis</strong>.</p> <p><strong>The <code>padj</code> column</strong> in the results table represents the p-value adjusted for multiple testing, and is the most important column of the results. Typically, a threshold such as <code>padj</code> \&lt; 0.05 is a good starting point for identifying significant genes. The default method for <strong>multiple test correction</strong> in DESeq2 is the FDR. Other methods can be used with the <code>pAdjustMethod</code> argument in the <code>results()</code> function.</p> <h3 id=gene-level-filtering>Gene-level filtering<a class=headerlink href=#gene-level-filtering title="Permanent link">&para;</a></h3> <p>Let’s take a closer look at our results table. As we scroll through it, you will notice that for <strong>selected genes there are NA values in the <code>pvalue</code> and <code>padj</code> columns</strong>. What does this mean?</p> <p><img src=./img/07b_hypothesis_testing/gene_filtering.png width=1472 style="display: block; margin: auto;"></p> <p>The missing values represent genes that have undergone filtering as part of the <code>DESeq()</code> function. Prior to differential expression analysis it is <strong>beneficial to omit genes that have little or no chance of being detected as differentially expressed.</strong> This will increase the power to detect differentially expressed genes. DESeq2 <strong>does not physically remove</strong> any genes from the original counts matrix, and so all genes will be present in your results table. The genes omitted by DESeq2 meet one of the <strong>three filtering criteria outlined below</strong>:</p> <p><strong>1. Genes with zero counts in all samples</strong></p> <p>If within a row, all samples have zero counts there is no expression information and therefore these genes are not tested. In our case, there are no genes that fulfill this criteria, since we have already filtered out these genes ourselves when we created our <code>dds</code> object.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-16-1 name=__codelineno-16-1 href=#__codelineno-16-1></a><span class=c1># Show genes with zero expression</span>
<a id=__codelineno-16-2 name=__codelineno-16-2 href=#__codelineno-16-2></a><span class=n>res_tableOE</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-16-3 name=__codelineno-16-3 href=#__codelineno-16-3></a><span class=w>  </span><span class=nf>as_tibble</span><span class=p>(</span><span class=n>rownames</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;gene&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-16-4 name=__codelineno-16-4 href=#__codelineno-16-4></a><span class=w>  </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>baseMean</span><span class=o>==</span><span class=m>0</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-16-5 name=__codelineno-16-5 href=#__codelineno-16-5></a><span class=w>  </span><span class=nf>head</span><span class=p>()</span>
</code></pre></div> <div class="admonition info"> <p class=admonition-title>Info</p> <p>If there would be any genes meeting this criteria, the baseMean column for these genes will be zero, and the log2 fold change estimates, p-value and adjusted p-value will all be set to NA.</p> </div> <p><strong>2. Genes with an extreme count outlier</strong></p> <p>The <code>DESeq()</code> function calculates, for every gene and for every sample, a diagnostic test for outliers called Cook’s distance. Cook’s distance is a measure of how much a single sample is influencing the fitted coefficients for a gene, and a large value of Cook’s distance is intended to indicate an outlier count. Genes which contain a Cook’s distance above a threshold are flagged, however at least 3 replicates are required for flagging, as it is difficult to judge which sample might be an outlier with only 2 replicates. We can turn off this filtering by using the <code>cooksCutoff</code> argument in the <code>results()</code> function.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-17-1 name=__codelineno-17-1 href=#__codelineno-17-1></a><span class=c1># Show genes that have an extreme outlier</span>
<a id=__codelineno-17-2 name=__codelineno-17-2 href=#__codelineno-17-2></a><span class=n>res_tableOE</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-17-3 name=__codelineno-17-3 href=#__codelineno-17-3></a><span class=w>  </span><span class=nf>as_tibble</span><span class=p>(</span><span class=n>rownames</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;gene&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-17-4 name=__codelineno-17-4 href=#__codelineno-17-4></a><span class=w>  </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=nf>is.na</span><span class=p>(</span><span class=n>pvalue</span><span class=p>)</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=nf>is.na</span><span class=p>(</span><span class=n>padj</span><span class=p>)</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=n>baseMean</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=m>0</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-17-5 name=__codelineno-17-5 href=#__codelineno-17-5></a><span class=w>  </span><span class=nf>head</span><span class=p>()</span>
</code></pre></div> <p>It seems that we have some genes with outliers!</p> <div class="admonition info"> <p class=admonition-title>Info</p> <p>If a gene contains a sample with an extreme count outlier then the p-value and adjusted p-value will be set to NA.</p> </div> <p><strong>3. Genes with a low mean normalized counts</strong></p> <p>DESeq2 defines a low mean threshold, that is empirically determined from your data, in which the fraction of significant genes can be increased by reducing the number of genes that are considered for multiple testing. This is based on the notion that genes with very low counts are not likely to see significant differences typically due to high dispersion.</p> <p><img src=./img/07b_hypothesis_testing/indep_filt_scatterplus.png width=2732 style="display: block; margin: auto;"></p> <p><em>Image courtesy of <a href=https://www.slideshare.net/joachimjacob/5rna-seqpart5detecting-differentialexpression>slideshare presentation</a> from Joachim Jacob, 2014.</em></p> <p>At a user-specified value (<code>alpha = 0.05</code>), DESeq2 evaluates the change in the number of significant genes as it filters out increasingly bigger portions of genes based on their mean counts, as shown in the figure above. The point at which the number of significant genes reaches a peak is the low mean threshold that is used to filter genes that undergo multiple testing. There is also an argument to turn off the filtering off by setting <code>independentFiltering = F</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-18-1 name=__codelineno-18-1 href=#__codelineno-18-1></a><span class=c1># Filter genes below the low mean threshold</span>
<a id=__codelineno-18-2 name=__codelineno-18-2 href=#__codelineno-18-2></a><span class=n>res_tableOE</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-18-3 name=__codelineno-18-3 href=#__codelineno-18-3></a><span class=w>  </span><span class=nf>as_tibble</span><span class=p>(</span><span class=n>rownames</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;gene&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-18-4 name=__codelineno-18-4 href=#__codelineno-18-4></a><span class=w>  </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=o>!</span><span class=nf>is.na</span><span class=p>(</span><span class=n>pvalue</span><span class=p>)</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=nf>is.na</span><span class=p>(</span><span class=n>padj</span><span class=p>)</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=n>baseMean</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=m>0</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-18-5 name=__codelineno-18-5 href=#__codelineno-18-5></a><span class=w>  </span><span class=nf>head</span><span class=p>()</span>
</code></pre></div> <div class="admonition info"> <p class=admonition-title>Info</p> <p>If a gene is filtered by independent filtering, then only the adjusted p-value will be set to NA.</p> </div> <div class="admonition warning"> <p class=admonition-title>Warning</p> <p>DESeq2 will perform the filtering outlined above by default; however other DE tools, such as EdgeR will not. Filtering is a necessary step, even if you are using limma-voom and/or edgeR's quasi-likelihood methods. Be sure to follow pre-filtering steps when using other tools, as outlined in their user guides found on Bioconductor as they generally perform much better.*</p> </div> <h2 id=fold-changes>Fold changes<a class=headerlink href=#fold-changes title="Permanent link">&para;</a></h2> <p>Another important column in the results table, is the <code>log2FoldChange</code>. With large significant gene lists it can be hard to extract meaningful biological relevance. To help increase stringency, one can also <strong>add a fold change threshold</strong>. Keep in mind when setting that value that we are working with log2 fold changes, so a cutoff of <code>log2FoldChange</code> \&lt; 1 would translate to an actual fold change of 2.</p> <details class=info> <summary>An alternative approach to add the fold change threshold</summary> <p>The <code>results()</code> function has an option to add a fold change threshold using the <code>lfcThrehsold</code> argument. This method is more statistically motivated, and is recommended when you want a more confident set of genes based on a certain fold-change. It actually performs a statistical test against the desired threshold, by performing a two-tailed test for log2 fold changes greater than the absolute value specified. The user can change the alternative hypothesis using <code>altHypothesis</code> and perform two one-tailed tests as well. <strong>This is a more conservative approach, so expect to retrieve a much smaller set of genes!</strong></p> <div class=highlight><pre><span></span><code><a id=__codelineno-19-1 name=__codelineno-19-1 href=#__codelineno-19-1></a><span class=n>res_tableOE_LFC1</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>results</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>contrast</span><span class=o>=</span><span class=n>contrast_oe</span><span class=p>,</span><span class=w> </span><span class=n>alpha</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.05</span><span class=p>,</span><span class=w> </span><span class=n>lfcThreshold</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=p>)</span>
</code></pre></div> </details> <p>The fold changes reported in the results table are calculated by:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-20-1 name=__codelineno-20-1 href=#__codelineno-20-1></a><span class=nf>log2</span><span class=p>(</span><span class=n>normalized_counts_group1</span><span class=w> </span><span class=o>/</span><span class=w> </span><span class=n>normalized_counts_group2</span><span class=p>)</span>
</code></pre></div> <h2 id=summarizing-results>Summarizing results<a class=headerlink href=#summarizing-results title="Permanent link">&para;</a></h2> <p>To summarize the results table, a handy function in DESeq2 is <code>summary()</code>. Confusingly it has the same name as the function used to inspect data frames. This function, when called with a DESeq results table as input, will summarize the results using a default threshold of padj \&lt; 0.1. However, since we had set the <code>alpha</code> argument to 0.05 when creating our results table threshold: FDR \&lt; 0.05 (padj/FDR is used even though the output says <code>p-value &lt; 0.05</code>). Let’s start with the OE vs control results:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-21-1 name=__codelineno-21-1 href=#__codelineno-21-1></a><span class=c1>## Summarize results</span>
<a id=__codelineno-21-2 name=__codelineno-21-2 href=#__codelineno-21-2></a><span class=nf>summary</span><span class=p>(</span><span class=n>res_tableOE</span><span class=p>,</span><span class=w> </span><span class=n>alpha</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.05</span><span class=p>)</span>
</code></pre></div> <p>In addition to the number of genes up- and down-regulated at the default threshold, <strong>the function also reports the number of genes that were tested (genes with non-zero total read count), and the number of genes not included in multiple test correction due to a low mean count</strong>.</p> <h2 id=extracting-significant-differentially-expressed-genes>Extracting significant differentially expressed genes<a class=headerlink href=#extracting-significant-differentially-expressed-genes title="Permanent link">&para;</a></h2> <p>Let’s first create variables that contain our threshold criteria. We will only be using the adjusted p-values in our criteria:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-22-1 name=__codelineno-22-1 href=#__codelineno-22-1></a><span class=c1>### Set thresholds</span>
<a id=__codelineno-22-2 name=__codelineno-22-2 href=#__codelineno-22-2></a><span class=n>padj.cutoff</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=m>0.05</span>
</code></pre></div> <p>We can easily subset the results table to only include those that are significant using the <code>dplyr::filter()</code> function, but first we will convert the results table into a tibble:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-23-1 name=__codelineno-23-1 href=#__codelineno-23-1></a><span class=c1># Create a tibble of results and add gene IDs to new object</span>
<a id=__codelineno-23-2 name=__codelineno-23-2 href=#__codelineno-23-2></a><span class=n>res_tableOE_tb</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_tableOE</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-23-3 name=__codelineno-23-3 href=#__codelineno-23-3></a><span class=w>  </span><span class=nf>as_tibble</span><span class=p>(</span><span class=n>rownames</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;gene&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-23-4 name=__codelineno-23-4 href=#__codelineno-23-4></a><span class=w>  </span><span class=nf>relocate</span><span class=p>(</span><span class=n>gene</span><span class=p>,</span><span class=w> </span><span class=n>.before</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>baseMean</span><span class=p>)</span>
<a id=__codelineno-23-5 name=__codelineno-23-5 href=#__codelineno-23-5></a>
<a id=__codelineno-23-6 name=__codelineno-23-6 href=#__codelineno-23-6></a><span class=nf>head</span><span class=p>(</span><span class=n>res_tableOE_tb</span><span class=p>)</span>
</code></pre></div> <p>Now we can subset that table to only keep the significant genes using our pre-defined thresholds:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-24-1 name=__codelineno-24-1 href=#__codelineno-24-1></a><span class=c1># Subset the tibble to keep only significant genes</span>
<a id=__codelineno-24-2 name=__codelineno-24-2 href=#__codelineno-24-2></a><span class=n>sigOE</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_tableOE_tb</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-24-3 name=__codelineno-24-3 href=#__codelineno-24-3></a><span class=w>  </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>padj</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=n>padj.cutoff</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-25-1 name=__codelineno-25-1 href=#__codelineno-25-1></a><span class=c1># Take a quick look at this tibble</span>
<a id=__codelineno-25-2 name=__codelineno-25-2 href=#__codelineno-25-2></a><span class=n>sigOE</span>
</code></pre></div> <p>Now that we have extracted the significant results, we are ready for visualization!</p> <div class="admonition question"> <p class=admonition-title><strong>Exercise 2</strong></p> <p><strong>MOV10 Differential Expression Analysis: Control versus Knockdown</strong></p> <p>Now that we have results for the overexpression results, do the same for the <strong>Control vs. Knockdown samples</strong>.</p> <ol> <li>Create a contrast vector called <code>contrast_kd</code>.</li> <li>Use contrast vector in the <code>results()</code> to extract a results table and store that to a variable called <code>res_tableKD</code>.</li> <li>Using a p-adjusted threshold of 0.05 (<code>padj.cutoff &lt; 0.05</code>), subset <code>res_tableKD</code> to report the number of genes that are up- and down-regulated in Mov10_knockdown compared to control.</li> <li>How many genes are differentially expressed in the Knockdown compared to Control? How does this compare to the overexpression significant gene list (in terms of numbers)?</li> </ol> </div> <hr> <p><em>This lesson was originally developed by members of the teaching team (Mary Piper, Meeta Mistry, Radhika Khetani) at the <a href=http://bioinformatics.sph.harvard.edu/ >Harvard Chan Bioinformatics Core (HBC)</a>.</em></p> <p><em>Some materials and hands-on activities were adapted from <a href=http://www.bioconductor.org/help/workflows/rnaseqGene/#de>RNA-seq workflow</a> on the Bioconductor website</em></p> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=07a_DEA.html class="md-footer__link md-footer__link--prev" aria-label="Previous: DESeq2" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> DESeq2 </div> </div> </a> <a href=07c_DEA_visualization.html class="md-footer__link md-footer__link--next" aria-label="Next: Visualization" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Visualization </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=assets/javascripts/bundle.5a2dcb6a.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>