<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
{
  "@context": "https://schema.org",
  "@id": "https://hds-sandbox.github.io/bulk_RNAseq_course/",
  "@type": "Course",
  "coursePrerequisites": [
    "Knowledge of R, Rstudio and Rmarkdown",
    "Basic knowledge of RNAseq technology",
    "Basic knowledge of data science and statistics such as PCA, clustering and statistical testing"
  ],
  "dct:conformsTo": "https://bioschemas.org/profiles/Course/1.0-RELEASE",
  "description": "This workshop includes a tutorial on how to approach RNAseq data, starting from your sequencing reads (fastq files). Thus, the workshop only briefly touches upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, mainly for the purpose of outlining considerations in the downstream bioinformatic analysis. This workshop is based on the materials developed by members of the teaching team at the Harvard Chan Bioinformatics Core (HBC), a collection of modified tutorials from the DESeq2, R language vignettes and the nf-core rnaseq pipeline.",
  "isBasedOn": [
    "https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html",
    "https://nf-co.re/rnaseq",
    "https://github.com/hbctraining/Intro-to-rnaseq-hpc-salmon-flipped"
  ],
  "keywords": "bioinformatics, rnaseq, bulk, R",
  "license": {
    "@id": "https://creativecommons.org/licenses/by/4.0/",
    "@type": "CreativeWork"
  },
  "mentions": [
    {
      "@id": "10.5281/zenodo.7660225",
      "@type": "Thing"
    },
    {
      "@id": "10.5281/zenodo.7565997",
      "@type": "Thing"
    }
  ],
  "name": "Introduction to bulk RNAseq analysis workshop"
}
</script><title>Functional Analysis for RNA-seq - Bulk RNAseq data analysis</title><link rel=stylesheet href=assets/stylesheets/main.975780f9.min.css><link rel=stylesheet href=assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#functional-analysis class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bulk RNAseq data analysis </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Functional Analysis for RNA-seq </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=02_experimental_planning.html class=md-tabs__link> Experimental planning </a> </li> <li class=md-tabs__item> <a href=03_data_explanation.html class=md-tabs__link> Data Explanation </a> </li> <li class=md-tabs__item> <a href=04a_preprocessing.html class=md-tabs__link> Read Preprocessing </a> </li> <li class=md-tabs__item> <a href=05a_data_analysis_setup.html class="md-tabs__link md-tabs__link--active"> Analysing RNAseq data </a> </li> <li class=md-tabs__item> <a href=09_summarized_workflow.html class=md-tabs__link> Summarized workflow </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> Bulk RNAseq data analysis </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> Course Introduction </a> </li> <li class=md-nav__item> <a href=02_experimental_planning.html class=md-nav__link> Experimental planning </a> </li> <li class=md-nav__item> <a href=03_data_explanation.html class=md-nav__link> Data Explanation </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_4 type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4> Read Preprocessing <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Read Preprocessing" data-md-level=1> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Read Preprocessing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=04a_preprocessing.html class=md-nav__link> Preprocessing steps </a> </li> <li class=md-nav__item> <a href=04b_pipelines.html class=md-nav__link> Nextflow and nf-core pipelines </a> </li> <li class=md-nav__item> <a href=04c_preprocessing_setup.html class=md-nav__link> UCloud setup </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5 type=checkbox id=__nav_5 checked> <label class=md-nav__link for=__nav_5> Analysing RNAseq data <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Analysing RNAseq data" data-md-level=1> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Analysing RNAseq data </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=05a_data_analysis_setup.html class=md-nav__link> Ucloud setup </a> </li> <li class=md-nav__item> <a href=05b_count_matrix.html class=md-nav__link> Count matrix </a> </li> <li class=md-nav__item> <a href=05c_count_normalization.html class=md-nav__link> Normalization </a> </li> <li class=md-nav__item> <a href=06_exploratory_analysis.html class=md-nav__link> Exploratory Analysis </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5_5 type=checkbox id=__nav_5_5> <label class=md-nav__link for=__nav_5_5> Differential Expression <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Differential Expression" data-md-level=2> <label class=md-nav__title for=__nav_5_5> <span class="md-nav__icon md-icon"></span> Differential Expression </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=07_extra_contrast_design.html class=md-nav__link> Design formula </a> </li> <li class=md-nav__item> <a href=07a_DEA.html class=md-nav__link> DESeq2 </a> </li> <li class=md-nav__item> <a href=07b_hypothesis_testing.html class=md-nav__link> Hypothesis testing </a> </li> <li class=md-nav__item> <a href=07c_DEA_visualization.html class=md-nav__link> Visualization </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5_6 type=checkbox id=__nav_5_6 checked> <label class=md-nav__link for=__nav_5_6> Functional Analysis <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label="Functional Analysis" data-md-level=2> <label class=md-nav__title for=__nav_5_6> <span class="md-nav__icon md-icon"></span> Functional Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=08a_FA_genomic_annotation.html class=md-nav__link> Gene annotation </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Over representation analysis <span class="md-nav__icon md-icon"></span> </label> <a href=08b_FA_overrepresentation.html class="md-nav__link md-nav__link--active"> Over representation analysis </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#over-representation-analysis class=md-nav__link> Over-representation analysis </a> <nav class=md-nav aria-label="Over-representation analysis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#hypergeometric-testing class=md-nav__link> Hypergeometric testing </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#gene-ontology-project class=md-nav__link> Gene Ontology project </a> <nav class=md-nav aria-label="Gene Ontology project"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#go-ontologies class=md-nav__link> GO Ontologies </a> </li> <li class=md-nav__item> <a href=#go-term-hierarchy class=md-nav__link> GO term hierarchy </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#clusterprofiler class=md-nav__link> clusterProfiler </a> <nav class=md-nav aria-label=clusterProfiler> <ul class=md-nav__list> <li class=md-nav__item> <a href=#running-clusterprofiler class=md-nav__link> Running clusterProfiler </a> </li> <li class=md-nav__item> <a href=#visualizing-clusterprofiler-results class=md-nav__link> Visualizing clusterProfiler results </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=08c_FA_GSEA.html class=md-nav__link> Class scoring </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=09_summarized_workflow.html class=md-nav__link> Summarized workflow </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#over-representation-analysis class=md-nav__link> Over-representation analysis </a> <nav class=md-nav aria-label="Over-representation analysis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#hypergeometric-testing class=md-nav__link> Hypergeometric testing </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#gene-ontology-project class=md-nav__link> Gene Ontology project </a> <nav class=md-nav aria-label="Gene Ontology project"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#go-ontologies class=md-nav__link> GO Ontologies </a> </li> <li class=md-nav__item> <a href=#go-term-hierarchy class=md-nav__link> GO term hierarchy </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#clusterprofiler class=md-nav__link> clusterProfiler </a> <nav class=md-nav aria-label=clusterProfiler> <ul class=md-nav__list> <li class=md-nav__item> <a href=#running-clusterprofiler class=md-nav__link> Running clusterProfiler </a> </li> <li class=md-nav__item> <a href=#visualizing-clusterprofiler-results class=md-nav__link> Visualizing clusterProfiler results </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course/edit/master/docs/08b_FA_overrepresentation.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1 id=functional-analysis>Functional analysis<a class=headerlink href=#functional-analysis title="Permanent link">&para;</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>&#128368; <strong>Time Estimation:</strong> 120 minutes </p> <p>&#128172; <strong>Learning Objectives:</strong> </p> <ol> <li>Determine how functions are attributed to genes using Gene Ontology terms</li> <li>Describe the theory of how functional enrichment tools yield statistically enriched functions or interactions</li> <li>Discuss functional analysis using over-representation analysis, functional class scoring, and pathway topology methods</li> <li>Identify popular functional analysis tools for over-representation analysis</li> </ol> </div> <p>The output of RNA-seq differential expression analysis is a list of significant differentially expressed genes (DEGs). To gain greater biological insight on the differentially expressed genes there are various analyses that can be done:</p> <ul> <li>determine whether there is enrichment of known biological functions, interactions, or pathways</li> <li>identify genes’ involvement in novel pathways or networks by grouping genes together based on similar trends</li> <li>use global changes in gene expression by visualizing all genes being significantly up- or down-regulated in the context of external interaction data</li> </ul> <p>Generally for any differential expression analysis, it is useful to interpret the resulting gene lists using freely available web- and R-based tools. While tools for functional analysis span a wide variety of techniques, they can loosely be categorized into three main types: over-representation analysis, functional class scoring, and pathway topology. See more <a href=https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/raw/master/resources/pathway_tools.pdf>here</a>.</p> <p><img src=./img/08b_FA_overrepresentation/pathway_analysis.png width=500 style="display: block; margin: auto;"></p> <p>The goal of functional analysis is to provide biological insight, so it’s necessary to analyze our results in the context of our experimental hypothesis: <strong>FMRP and MOV10 associate and regulate the translation of a subset of RNAs</strong>. Therefore, based on the authors’ hypothesis, we may expect the enrichment of processes/pathways related to <strong>translation, splicing, and the regulation of mRNAs</strong>, which we would need to validate experimentally.</p> <div class="admonition note"> <p class=admonition-title>Note</p> <p>All tools described below are great tools to validate experimental results and to make hypotheses. These tools suggest genes/pathways that may be involved with your condition of interest; however, you should NOT use these tools to make conclusions about the pathways involved in your experimental process. You will need to perform experimental validation of any suggested pathways.</p> </div> <h2 id=over-representation-analysis>Over-representation analysis<a class=headerlink href=#over-representation-analysis title="Permanent link">&para;</a></h2> <p>There are a plethora of functional enrichment tools that perform some type of "over-representation" analysis by querying databases containing information about gene function and interactions.</p> <p>These databases typically <strong>categorize genes into groups (gene sets)</strong> based on shared function, or involvement in a pathway, or presence in a specific cellular location, or other categorizations, e.g. functional pathways, etc. Essentially, known genes are binned into categories that have been consistently named (controlled vocabulary) based on how the gene has been annotated functionally. These categories are independent of any organism, however each organism has distinct categorizations available.</p> <p>To determine whether any categories are over-represented, you can determine the probability of having the observed proportion of genes associated with a specific category in your gene list based on the proportion of genes associated with the same category in the background set (gene categorizations for the appropriate organism).</p> <p><img src=./img/08b_FA_overrepresentation/go_proportions.png width=848 style="display: block; margin: auto;"></p> <p><img src=./img/08b_FA_overrepresentation/go_proportions_table3.png width=1825 style="display: block; margin: auto;"></p> <p>The statistical test that will determine whether something is actually over-represented is the <em>Hypergeometric test</em>.</p> <h3 id=hypergeometric-testing>Hypergeometric testing<a class=headerlink href=#hypergeometric-testing title="Permanent link">&para;</a></h3> <p>Using the example of the first functional category above, hypergeometric distribution is a probability distribution that describes the probability of 25 genes (k) being associated with "Functional category 1", for all genes in our gene list (n=1000), from a population of all of the genes in entire genome (N=13,000) which contains 35 genes (K) associated with "Functional category 1" [<a href=https://en.wikipedia.org/wiki/Hypergeometric_distribution>4</a>].</p> <p>The calculation of probability of k successes follows the formula:</p> <p><img src=./img/08b_FA_overrepresentation/hypergeo.png width=200 style="display: block; margin: auto;"></p> <p>This test will result in an adjusted p-value (after multiple test correction) for each category tested.</p> <h2 id=gene-ontology-project>Gene Ontology project<a class=headerlink href=#gene-ontology-project title="Permanent link">&para;</a></h2> <p>One of the most widely-used categorizations is the <strong>Gene Ontology (GO)</strong> established by the <a href=http://geneontology.org/page/introduction-go-resource>Gene Ontology project</a>.</p> <div class="admonition quote"> <p class=admonition-title>Quote</p> <p>"The Gene Ontology project is a collaborative effort to address the need for consistent descriptions of gene products across databases". </p> </div> <p>The <a href=http://geneontology.org/page/go-consortium-contributors-list>Gene Ontology Consortium</a> maintains the GO terms, and these GO terms are incorporated into gene annotations in many of the popular repositories for animal, plant, and microbial genomes.</p> <p>Tools that investigate enrichment of biological functions or interactions often use the Gene Ontology (GO) categorizations, i.e. the GO terms to determine whether any have significantly modified representation in a given list of genes. Therefore, to best use and interpret the results from these functional analysis tools, it is helpful to have a good understanding of the GO terms themselves and their organization.</p> <h3 id=go-ontologies>GO Ontologies<a class=headerlink href=#go-ontologies title="Permanent link">&para;</a></h3> <p>To describe the roles of genes and gene products, GO terms are organized into three independent controlled vocabularies (ontologies) in a species-independent manner:</p> <ul> <li><strong>Biological process:</strong> refers to the biological role involving the gene or gene product, and could include "transcription", "signal transduction", and "apoptosis". A biological process generally involves a chemical or physical change of the starting material or input.</li> <li><strong>Molecular function:</strong> represents the biochemical activity of the gene product, such activities could include "ligand", "GTPase", and "transporter".</li> <li><strong>Cellular component:</strong> refers to the location in the cell of the gene product. Cellular components could include "nucleus", "lysosome", and "plasma membrane".</li> </ul> <p>Each GO term has a term name (e.g. <strong>DNA repair</strong>) and a unique term accession number (<strong><go:0005125></strong>), and a single gene product can be associated with many GO terms, since a single gene product "may function in several processes, contain domains that carry out diverse molecular functions, and participate in multiple alternative interactions with other proteins, organelles or locations in the cell". See more <a href=https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/raw/master/resources/go_terms.pdf>here</a>.</p> <h3 id=go-term-hierarchy>GO term hierarchy<a class=headerlink href=#go-term-hierarchy title="Permanent link">&para;</a></h3> <p>Some gene products are well-researched, with vast quantities of data available regarding their biological processes and functions. However, other gene products have very little data available about their roles in the cell.</p> <p>For example, the protein, "p53", would contain a wealth of information on it’s roles in the cell, whereas another protein might only be known as a "membrane-bound protein" with no other information available.</p> <p>The GO ontologies were developed to describe and query biological knowledge with differing levels of information available. To do this, GO ontologies are loosely hierarchical, ranging from general, ‘parent’, terms to more specific, ‘child’ terms. The GO ontologies are "loosely" hierarchical since ‘child’ terms can have multiple ‘parent’ terms.</p> <p>Some genes with less information may only be associated with general ‘parent’ terms or no terms at all, while other genes with a lot of information be associated with many terms.</p> <p><img alt="Nature Reviews Cancer 7, 23-34 (January 2007)" src=img/08b_FA_overrepresentation/go_heirarchy.jpg></p> <div class="admonition tip"> <p class=admonition-title>Tip</p> </div> <p>More tips for working with GO can be found <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003343">here</a></p> <h2 id=clusterprofiler>clusterProfiler<a class=headerlink href=#clusterprofiler title="Permanent link">&para;</a></h2> <p>We will be using <a href=http://bioconductor.org/packages/release/bioc/html/clusterProfiler.html>clusterProfiler</a> to perform over-representation analysis on GO terms associated with our list of significant genes. The tool takes as input a significant gene list and a background gene list and performs statistical enrichment analysis using hypergeometric testing. The basic arguments allow the user to select the appropriate organism and GO ontology (BP, CC, MF) to test.</p> <h3 id=running-clusterprofiler>Running clusterProfiler<a class=headerlink href=#running-clusterprofiler title="Permanent link">&para;</a></h3> <p>To run clusterProfiler GO over-representation analysis, we will change our gene names into Ensembl IDs, since the tool works a bit easier with the Ensembl IDs. Then load the following libraries:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=c1># Load libraries</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a><span class=nf>library</span><span class=p>(</span><span class=n>DOSE</span><span class=p>)</span>
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a><span class=nf>library</span><span class=p>(</span><span class=n>pathview</span><span class=p>)</span>
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=nf>library</span><span class=p>(</span><span class=n>clusterProfiler</span><span class=p>)</span>
<a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a><span class=nf>library</span><span class=p>(</span><span class=n>org.Hs.eg.db</span><span class=p>)</span>
</code></pre></div> <p>To perform the over-representation analysis, we need a list of background genes and a list of significant genes. For our background dataset we will use all genes tested for differential expression (all genes in our results table). For our significant gene list we will use genes with p-adjusted values less than 0.05 (we could include a fold change threshold too if we have many DE genes).</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=c1>## Create background dataset for hypergeometric testing using all genes tested for significance in the results</span>
<a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a><span class=n>allOE_genes</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>res_ids</span><span class=p>,</span><span class=w> </span><span class=o>!</span><span class=nf>is.na</span><span class=p>(</span><span class=n>gene</span><span class=p>))</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a><span class=w>  </span><span class=nf>pull</span><span class=p>(</span><span class=n>gene</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a><span class=w>  </span><span class=nf>as.character</span><span class=p>()</span>
<a id=__codelineno-1-5 name=__codelineno-1-5 href=#__codelineno-1-5></a>
<a id=__codelineno-1-6 name=__codelineno-1-6 href=#__codelineno-1-6></a><span class=c1>## Extract significant results</span>
<a id=__codelineno-1-7 name=__codelineno-1-7 href=#__codelineno-1-7></a><span class=n>sigOE</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>res_ids</span><span class=p>,</span><span class=w> </span><span class=n>padj</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=m>0.05</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=o>!</span><span class=nf>is.na</span><span class=p>(</span><span class=n>gene</span><span class=p>))</span>
<a id=__codelineno-1-8 name=__codelineno-1-8 href=#__codelineno-1-8></a>
<a id=__codelineno-1-9 name=__codelineno-1-9 href=#__codelineno-1-9></a><span class=n>sigOE_genes</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>sigOE</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-1-10 name=__codelineno-1-10 href=#__codelineno-1-10></a><span class=w>  </span><span class=nf>pull</span><span class=p>(</span><span class=n>gene</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-1-11 name=__codelineno-1-11 href=#__codelineno-1-11></a><span class=w>  </span><span class=nf>as.character</span><span class=p>()</span>
</code></pre></div> <p>Now we can perform the GO enrichment analysis and save the results:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=c1>## Run GO enrichment analysis </span>
<a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a><span class=n>ego</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>enrichGO</span><span class=p>(</span><span class=n>gene</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sigOE_genes</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a><span class=w>                </span><span class=n>universe</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>allOE_genes</span><span class=p>,</span>
<a id=__codelineno-2-4 name=__codelineno-2-4 href=#__codelineno-2-4></a><span class=w>                </span><span class=n>keyType</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;ENSEMBL&quot;</span><span class=p>,</span>
<a id=__codelineno-2-5 name=__codelineno-2-5 href=#__codelineno-2-5></a><span class=w>                </span><span class=n>OrgDb</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>org.Hs.eg.db</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-2-6 name=__codelineno-2-6 href=#__codelineno-2-6></a><span class=w>                </span><span class=n>ont</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;BP&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-2-7 name=__codelineno-2-7 href=#__codelineno-2-7></a><span class=w>                </span><span class=n>pAdjustMethod</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;BH&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-2-8 name=__codelineno-2-8 href=#__codelineno-2-8></a><span class=w>                </span><span class=n>qvalueCutoff</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.05</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-2-9 name=__codelineno-2-9 href=#__codelineno-2-9></a><span class=w>                </span><span class=n>readable</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
</code></pre></div> <div class="admonition note"> <p class=admonition-title>Note</p> <p>The different organisms with annotation databases available to use with for the <code>OrgDb</code> argument can be found <a href=img/08b_FA_overrepresentation/orgdb_annotation_databases.png>here</a>.</p> <p>Also, the <code>keyType</code> argument may be coded as <code>keytype</code> in different versions of clusterProfiler.</p> <p>Finally, the <code>ont</code> argument can accept either "BP" (Biological Process), "MF" (Molecular Function), and "CC" (Cellular Component) subontologies, or "ALL" for all three.</p> </div> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=c1>## Output results from GO analysis to a table</span>
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=n>cluster_summary</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>data.frame</span><span class=p>(</span><span class=n>ego</span><span class=p>)</span>
<a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a><span class=n>cluster_summary</span>
<a id=__codelineno-3-4 name=__codelineno-3-4 href=#__codelineno-3-4></a>
<a id=__codelineno-3-5 name=__codelineno-3-5 href=#__codelineno-3-5></a><span class=nf>write.csv</span><span class=p>(</span><span class=n>cluster_summary</span><span class=p>,</span><span class=w> </span><span class=s>&quot;../Results/clusterProfiler_Mov10oe.csv&quot;</span><span class=p>)</span>
</code></pre></div> <p><img src=./img/08b_FA_overrepresentation/cluster_summary.png width=600 style="display: block; margin: auto;"></p> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>Instead of saving just the results summary from the <code>ego</code> object, it might also be beneficial to save the object itself. The <code>save()</code> function enables you to save it as a <code>.rda</code> file, e.g. <code>save(ego, file="results/ego.rda")</code>. The complementary function to <code>save()</code> is the function <code>load()</code>, e.g. </p> <p><code>ego &lt;- load(file="results/ego.rda")</code>.</p> <p>This is a useful set of functions to know, since it enables one to preserve analyses at specific stages and reload them when needed. More information about these functions can be found <a href=https://www.r-bloggers.com/load-save-and-rda-files/ >here</a> &amp; <a href=http://rpubs.com/euclid/387778>here</a>.</p> </div> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>You can also perform GO enrichment analysis with only the up or down regulated genes** in addition to performing it for the full list of significant genes. This can be useful to identify GO terms impacted in one direction and not the other. If very few genes are in any of these lists (&lt; 50, roughly) it may not be possible to get any significant GO terms.</p> </div> <div class="admonition question"> <p class=admonition-title><strong>Exercise 1</strong></p> <p>Create two new GO enrichment analyses one with UP and another for DOWN regulated genes for MOV10 overexpression.</p> </div> <h3 id=visualizing-clusterprofiler-results>Visualizing clusterProfiler results<a class=headerlink href=#visualizing-clusterprofiler-results title="Permanent link">&para;</a></h3> <p>clusterProfiler has a variety of options for viewing the over-represented GO terms. We will explore the dotplot, enrichment plot, and the category netplot.</p> <p>The <strong>dotplot</strong> shows the number of genes associated with the first 50 terms (size) and the p-adjusted values for these terms (color). This plot displays the top 50 GO terms by gene ratio (# genes related to GO term / total number of sig genes), not p-adjusted value.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=c1>## Dotplot </span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=nf>dotplot</span><span class=p>(</span><span class=n>ego</span><span class=p>,</span><span class=w> </span><span class=n>showCategory</span><span class=o>=</span><span class=m>50</span><span class=p>)</span>
</code></pre></div> <p><img src=./img/08b_FA_overrepresentation/mov10oe_dotplot.png width=1200 style="display: block; margin: auto;"></p> <p>The next plot is the <strong>enrichment GO plot</strong>, which shows the relationship between the top 50 most significantly enriched GO terms (padj.), by grouping similar terms together. Before creating the plot, we will need to obtain the similarity between terms using the <code>pairwise_termsim()</code> function <a href=https://rdrr.io/github/GuangchuangYu/enrichplot/man/emapplot.html>instructions for emapplot</a>. In the enrichment plot, the color represents the p-values relative to the other displayed terms (brighter red is more significant), and the size of the terms represents the number of genes that are significant from our list.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=c1># Add similarity matrix to the termsim slot of enrichment result</span>
<a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a><span class=n>ego</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>enrichplot</span><span class=o>::</span><span class=nf>pairwise_termsim</span><span class=p>(</span><span class=n>ego</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=c1># Enrichmap clusters the 50 most significant (by padj) GO terms to visualize relationships between terms</span>
<a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=nf>emapplot</span><span class=p>(</span><span class=n>ego</span><span class=p>,</span><span class=w> </span><span class=n>showCategory</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>50</span><span class=p>)</span>
</code></pre></div> <p><img src=./img/08b_FA_overrepresentation/emapplot.png width=1479 style="display: block; margin: auto;"></p> <p>Finally, the <strong>category netplot</strong> shows the relationships between the genes associated with the top five most significant GO terms and the fold changes of the significant genes associated with these terms (color). The size of the GO terms reflects the pvalues of the terms, with the more significant terms being larger. This plot is particularly useful for hypothesis generation in identifying genes that may be important to several of the most affected processes.</p> <div class="admonition warning"> <p class=admonition-title>Warning</p> <p>You may need to install the <code>ggnewscale</code> package using <code>install.packages("ggnewscale")</code> for the <code>cnetplot()</code> function to work.</p> </div> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1># To color genes by log2 fold changes, we need to extract the log2 fold changes from our results table creating a named vector</span>
<a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=n>OE_foldchanges</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>sigOE</span><span class=o>$</span><span class=n>log2FoldChange</span>
<a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a>
<a id=__codelineno-7-4 name=__codelineno-7-4 href=#__codelineno-7-4></a><span class=nf>names</span><span class=p>(</span><span class=n>OE_foldchanges</span><span class=p>)</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>sigOE</span><span class=o>$</span><span class=n>gene</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=c1># Cnetplot details the genes associated with one or more terms - by default gives the top 5 significant terms (by padj)</span>
<a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a><span class=nf>cnetplot</span><span class=p>(</span><span class=n>ego</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a><span class=w>         </span><span class=n>categorySize</span><span class=o>=</span><span class=s>&quot;pvalue&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a><span class=w>         </span><span class=n>showCategory</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>5</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-8-5 name=__codelineno-8-5 href=#__codelineno-8-5></a><span class=w>         </span><span class=n>foldChange</span><span class=o>=</span><span class=n>OE_foldchanges</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-8-6 name=__codelineno-8-6 href=#__codelineno-8-6></a><span class=w>         </span><span class=n>vertex.label.font</span><span class=o>=</span><span class=m>6</span><span class=p>)</span>
</code></pre></div> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>If some of the high fold changes are getting drowned out due to a large range, you could set a maximum fold change value</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=n>OE_foldchanges</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>ifelse</span><span class=p>(</span><span class=n>OE_foldchanges</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=m>2</span><span class=p>,</span><span class=w> </span><span class=m>2</span><span class=p>,</span><span class=w> </span><span class=n>OE_foldchanges</span><span class=p>)</span>
<a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a><span class=n>OE_foldchanges</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>ifelse</span><span class=p>(</span><span class=n>OE_foldchanges</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=m>-2</span><span class=p>,</span><span class=w> </span><span class=m>-2</span><span class=p>,</span><span class=w> </span><span class=n>OE_foldchanges</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=nf>cnetplot</span><span class=p>(</span><span class=n>ego</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=w>     </span><span class=n>categorySize</span><span class=o>=</span><span class=s>&quot;pvalue&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a><span class=w>     </span><span class=n>showCategory</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>5</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a><span class=w>     </span><span class=n>foldChange</span><span class=o>=</span><span class=n>OE_foldchanges</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-10-5 name=__codelineno-10-5 href=#__codelineno-10-5></a><span class=w>     </span><span class=n>vertex.label.font</span><span class=o>=</span><span class=m>6</span><span class=p>)</span>
</code></pre></div> </div> <p><img src=./img/08b_FA_overrepresentation/cnetplot1.png width=1800 style="display: block; margin: auto;"></p> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>If you are interested in significant processes that are <strong>not</strong> among the top five, you can subset your <code>ego</code> dataset to only display these processes:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=c1># Subsetting the ego results without overwriting original `ego` variable</span>
<a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a><span class=n>ego2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>ego</span>
<a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a>
<a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a><span class=n>ego2</span><span class=o>@</span><span class=n>result</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>ego</span><span class=o>@</span><span class=n>result</span><span class=p>[</span><span class=nf>c</span><span class=p>(</span><span class=m>1</span><span class=p>,</span><span class=m>3</span><span class=p>,</span><span class=m>4</span><span class=p>,</span><span class=m>8</span><span class=p>,</span><span class=m>9</span><span class=p>),]</span>
<a id=__codelineno-11-5 name=__codelineno-11-5 href=#__codelineno-11-5></a>
<a id=__codelineno-11-6 name=__codelineno-11-6 href=#__codelineno-11-6></a><span class=c1># Plotting terms of interest</span>
<a id=__codelineno-11-7 name=__codelineno-11-7 href=#__codelineno-11-7></a><span class=nf>cnetplot</span><span class=p>(</span><span class=n>ego2</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-11-8 name=__codelineno-11-8 href=#__codelineno-11-8></a><span class=w>     </span><span class=n>categorySize</span><span class=o>=</span><span class=s>&quot;pvalue&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-11-9 name=__codelineno-11-9 href=#__codelineno-11-9></a><span class=w>     </span><span class=n>foldChange</span><span class=o>=</span><span class=n>OE_foldchanges</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-11-10 name=__codelineno-11-10 href=#__codelineno-11-10></a><span class=w>     </span><span class=n>showCategory</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>5</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-11-11 name=__codelineno-11-11 href=#__codelineno-11-11></a><span class=w>     </span><span class=n>vertex.label.font</span><span class=o>=</span><span class=m>6</span><span class=p>)</span>
</code></pre></div> <p><img src=./img/08b_FA_overrepresentation/cnetplot-2.png width=1800 style="display: block; margin: auto;"></p> </div> <div class="admonition question"> <p class=admonition-title><strong>Exercise 2</strong></p> <p>Run a Disease Ontology (DO) overrepresentation analysis using the <code>enrichDO()</code> function. <strong>NOTE</strong> the arguments are very similar to the previous examples. </p> <ul> <li>Do you find anything interesting?</li> </ul> </div> <div class="admonition question"> <p class=admonition-title><strong>Exercise 3</strong></p> <p>Run an enrichment analysis on the results of the DEA for knockdown vs control samples. Remember to use the annotated results!</p> </div> <hr> <p><em>This lesson was originally developed by members of the teaching team (Mary Piper, Radhika Khetani) at the <a href=http://bioinformatics.sph.harvard.edu/ >Harvard Chan Bioinformatics Core (HBC)</a>.</em></p> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=08a_FA_genomic_annotation.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Gene annotation" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Gene annotation </div> </div> </a> <a href=08c_FA_GSEA.html class="md-footer__link md-footer__link--next" aria-label="Next: Class scoring" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Class scoring </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.footer", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=assets/javascripts/bundle.5a2dcb6a.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>