<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="This course is an introduction for how to approach bulk RNAseq data, starting from the sequencing reads.  It will provide an overview of the fundamentals of RNAseq analysis, including read preprocessing, data normalization, data exploration with PCAs and heatmaps, performing differential expression analysis and annotation of the differentially expressed genes.  Participants will also learn how to evaluate confounding and batch effects in the data.  The course will further touch upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, especially about how they influence downstream bioinformatic analysis.   "><meta name=author content="Jose Alejandro Romero Herrera"><link rel=prev href=08b_FA_overrepresentation.html><link rel=next href=09_summarized_workflow.html><link rel=icon href=img/favicon.png><meta name=generator content="mkdocs-1.4.2, mkdocs-material-9.0.8"><title>Functional class scoring - Bulk RNAseq data analysis</title><link rel=stylesheet href=assets/stylesheets/main.0d440cfe.min.css><link rel=stylesheet href=assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#functional-class-scoring class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bulk RNAseq data analysis </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Functional class scoring </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=02_experimental_planning.html class=md-tabs__link> Experimental planning </a> </li> <li class=md-tabs__item> <a href=03_data_explanation.html class=md-tabs__link> Data Explanation </a> </li> <li class=md-tabs__item> <a href=04a_preprocessing.html class=md-tabs__link> Read Preprocessing </a> </li> <li class=md-tabs__item> <a href=05a_data_analysis_setup.html class="md-tabs__link md-tabs__link--active"> Analysing RNAseq data </a> </li> <li class=md-tabs__item> <a href=09_summarized_workflow.html class=md-tabs__link> Summarized workflow </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> Bulk RNAseq data analysis </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> Course Introduction </a> </li> <li class=md-nav__item> <a href=02_experimental_planning.html class=md-nav__link> Experimental planning </a> </li> <li class=md-nav__item> <a href=03_data_explanation.html class=md-nav__link> Data Explanation </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> Read Preprocessing <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Read Preprocessing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=04a_preprocessing.html class=md-nav__link> Preprocessing steps </a> </li> <li class=md-nav__item> <a href=04b_pipelines.html class=md-nav__link> Nextflow and nf-core pipelines </a> </li> <li class=md-nav__item> <a href=04c_preprocessing_setup.html class=md-nav__link> UCloud setup </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5 checked> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex=0> Analysing RNAseq data <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=true> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Analysing RNAseq data </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=05a_data_analysis_setup.html class=md-nav__link> Ucloud setup </a> </li> <li class=md-nav__item> <a href=05b_count_matrix.html class=md-nav__link> Count matrix </a> </li> <li class=md-nav__item> <a href=05c_count_normalization.html class=md-nav__link> Normalization </a> </li> <li class=md-nav__item> <a href=06_exploratory_analysis.html class=md-nav__link> Exploratory Analysis </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_5> <label class=md-nav__link for=__nav_5_5 id=__nav_5_5_label tabindex=0> Differential Expression <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5_5> <span class="md-nav__icon md-icon"></span> Differential Expression </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=07_extra_contrast_design.html class=md-nav__link> Design formula </a> </li> <li class=md-nav__item> <a href=07a_DEA.html class=md-nav__link> DESeq2 </a> </li> <li class=md-nav__item> <a href=07b_hypothesis_testing.html class=md-nav__link> Hypothesis testing </a> </li> <li class=md-nav__item> <a href=07c_DEA_visualization.html class=md-nav__link> Visualization </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_6 checked> <label class=md-nav__link for=__nav_5_6 id=__nav_5_6_label tabindex=0> Functional Analysis <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_6_label aria-expanded=true> <label class=md-nav__title for=__nav_5_6> <span class="md-nav__icon md-icon"></span> Functional Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=08a_FA_genomic_annotation.html class=md-nav__link> Gene annotation </a> </li> <li class=md-nav__item> <a href=08b_FA_overrepresentation.html class=md-nav__link> Over representation analysis </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Class scoring <span class="md-nav__icon md-icon"></span> </label> <a href=08c_FA_GSEA.html class="md-nav__link md-nav__link--active"> Class scoring </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#gene-set-enrichment-analysis-using-clusterprofiler-and-pathview class=md-nav__link> Gene set enrichment analysis using clusterProfiler and Pathview </a> <nav class=md-nav aria-label="Gene set enrichment analysis using clusterProfiler and Pathview"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#preparation-for-gsea class=md-nav__link> Preparation for GSEA </a> </li> <li class=md-nav__item> <a href=#theory-of-gsea class=md-nav__link> Theory of GSEA </a> </li> <li class=md-nav__item> <a href=#performing-gsea class=md-nav__link> Performing GSEA </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#pathway-topology-tools class=md-nav__link> Pathway topology tools </a> </li> <li class=md-nav__item> <a href=#other-tools class=md-nav__link> Other Tools </a> <nav class=md-nav aria-label="Other Tools"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#co-expression-clustering class=md-nav__link> Co-expression clustering </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#resources-for-functional-analysis class=md-nav__link> Resources for functional analysis </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=09_summarized_workflow.html class=md-nav__link> Summarized workflow </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#gene-set-enrichment-analysis-using-clusterprofiler-and-pathview class=md-nav__link> Gene set enrichment analysis using clusterProfiler and Pathview </a> <nav class=md-nav aria-label="Gene set enrichment analysis using clusterProfiler and Pathview"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#preparation-for-gsea class=md-nav__link> Preparation for GSEA </a> </li> <li class=md-nav__item> <a href=#theory-of-gsea class=md-nav__link> Theory of GSEA </a> </li> <li class=md-nav__item> <a href=#performing-gsea class=md-nav__link> Performing GSEA </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#pathway-topology-tools class=md-nav__link> Pathway topology tools </a> </li> <li class=md-nav__item> <a href=#other-tools class=md-nav__link> Other Tools </a> <nav class=md-nav aria-label="Other Tools"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#co-expression-clustering class=md-nav__link> Co-expression clustering </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#resources-for-functional-analysis class=md-nav__link> Resources for functional analysis </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=functional-class-scoring>Functional class scoring<a class=headerlink href=#functional-class-scoring title="Permanent link">&para;</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>&#128368; <strong>Time Estimation:</strong> 40 minutes </p> <p>&#128172; <strong>Learning Objectives:</strong> </p> <ol> <li>Demonstrate the use of the design formula with simple and complex designs</li> <li>Construct R code to execute the differential expression analysis workflow with DESeq2</li> <li>Discuss functional class scoring, and pathway topology methods</li> <li>Construct a GSEA analysis using GO and KEGG gene sets</li> <li>Examine results of a GSEA using pathview package</li> <li>List other tools and resources for identifying genes of novel pathways or networks</li> </ol> </div> <p>Over-representation analysis is only a single type of functional analysis method that is available for teasing apart the biological processes important to your condition of interest. Other types of analyses can be equally important or informative, including functional class scoring methods.</p> <p><img src=./img/08c_FA_GSEA/pathway_analysis.png style="display: block; margin: auto;"></p> <p>Functional class scoring (FCS) tools, such as <a href=https://www.pnas.org/content/102/43/15545>GSEA</a>, most often use the gene-level statistics or log2 fold changes for all genes from the differential expression results, then look to see whether gene sets for particular biological pathways are enriched among the large positive or negative fold changes.</p> <p><img src=./img/08c_FA_GSEA/gsea_theory.png style="display: block; margin: auto;"></p> <p>The hypothesis of FCS methods is that although large changes in individual genes can have significant effects on pathways (and will be detected via ORA methods), weaker but coordinated changes in sets of functionally related genes (i.e., pathways) can also have significant effects. Thus, rather than setting an arbitrary threshold to identify ‘significant genes’, <strong>all genes are considered</strong> in the analysis. The gene-level statistics from the dataset are aggregated to generate a single pathway-level statistic and statistical significance of each pathway is reported. This type of analysis can be particularly helpful if the differential expression analysis only outputs a small list of significant DE genes.</p> <h3 id=gene-set-enrichment-analysis-using-clusterprofiler-and-pathview>Gene set enrichment analysis using clusterProfiler and Pathview<a class=headerlink href=#gene-set-enrichment-analysis-using-clusterprofiler-and-pathview title="Permanent link">&para;</a></h3> <p>Using the log2 fold changes obtained from the differential expression analysis for every gene, gene set enrichment analysis and pathway analysis can be performed using clusterProfiler and Pathview tools.</p> <p>For a gene set or pathway analysis using clusterProfiler, coordinated differential expression over gene sets is tested instead of changes of individual genes. “Gene sets are pre-defined groups of genes, which are functionally related. Commonly used gene sets include those derived from KEGG pathways, Gene Ontology terms, MSigDB, Reactome, or gene groups that share some other functional annotations, etc. Consistent perturbations over such gene sets frequently suggest mechanistic changes”.</p> <h4 id=preparation-for-gsea>Preparation for GSEA<a class=headerlink href=#preparation-for-gsea title="Permanent link">&para;</a></h4> <p>clusterProfiler offers several functions to perform GSEA using different genes sets, including but not limited to GO, KEGG, and MSigDb. We will use the KEGG gene sets, which identify genes using their Entrez IDs. Therefore, to perform the analysis, we will need to acquire the Entrez IDs. We will also need to remove the Entrez ID NA values and duplicates (due to gene ID conversion) prior to the analysis:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=c1>## Remove any NA values (reduces the data by quite a bit) and duplicates</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a>
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a><span class=n>res_entrez</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>res_ids</span><span class=p>,</span><span class=w> </span><span class=n>entrezid</span><span class=w> </span><span class=o>!=</span><span class=w> </span><span class=s>&quot;NA&quot;</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=nf>duplicated</span><span class=p>(</span><span class=n>entrezid</span><span class=p>)</span><span class=o>==</span><span class=bp>F</span><span class=p>)</span>
</code></pre></div> <p>Finally, extract and name the fold changes:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=c1>## Extract the foldchanges</span>
<a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a><span class=n>foldchanges</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_entrez</span><span class=o>$</span><span class=n>log2FoldChange</span>
<a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a>
<a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a><span class=c1>## Name each fold change with the corresponding Entrez ID</span>
<a id=__codelineno-1-5 name=__codelineno-1-5 href=#__codelineno-1-5></a><span class=nf>names</span><span class=p>(</span><span class=n>foldchanges</span><span class=p>)</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_entrez</span><span class=o>$</span><span class=n>entrezid</span>
</code></pre></div> <p>Next we need to order the fold changes in decreasing order. To do this we’ll use the <code>sort()</code> function, which takes a vector as input. This is in contrast to Tidyverse’s <code>arrange()</code>, which requires a data frame.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=c1>## Sort fold changes in decreasing order</span>
<a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a><span class=n>foldchanges</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>sort</span><span class=p>(</span><span class=n>foldchanges</span><span class=p>,</span><span class=w> </span><span class=n>decreasing</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
<a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a>
<a id=__codelineno-2-4 name=__codelineno-2-4 href=#__codelineno-2-4></a><span class=nf>head</span><span class=p>(</span><span class=n>foldchanges</span><span class=p>)</span>
</code></pre></div> <h4 id=theory-of-gsea>Theory of GSEA<a class=headerlink href=#theory-of-gsea title="Permanent link">&para;</a></h4> <p>Now we are ready to perform GSEA. The details regarding GSEA can be found in the <a href=https://www.pnas.org/content/102/43/15545>PNAS paper</a> by Subramanian et al. We will describe briefly the steps outlined in the paper below:</p> <p><img src=./img/08c_FA_GSEA/gsea_overview.png style="display: block; margin: auto;"></p> <p><em><strong>Image credit:</strong> <a href=https://www.pnas.org/content/102/43/15545>Subramanian et al. Proceedings of the National Academy of Sciences Oct 2005, 102 (43) 15545-15550; DOI: 10.1073/pnas.0506580102</a></em></p> <p>This image describes the theory of GSEA, with the ‘gene set S’ showing the metric used (in our case, ranked log2 fold changes) to determine enrichment of genes in the gene set. The left-most image is representing this metric used for the GSEA analysis. The log2 fold changes for each gene in the ‘gene set S’ is shown as a line in the middle image. The large positive log2 fold changes are at the top of the gene set image, while the largest negative log2 fold changes are at the bottom of the gene set image. In the right-most image, the gene set is turned horizontally, underneath which is an image depicting the calculations involved in determining enrichment, as described below.</p> <p><strong>Step 1:</strong> Calculation of enrichment score:</p> <p>An enrichment score for a particular gene set is calculated by walking down the list of log2 fold changes and increasing the running-sum statistic every time a gene in the gene set is encountered and decreasing it when genes are not part of the gene set. The size of the increase/decrease is determined by magnitude of the log2 fold change. Larger (positive or negative) log2 fold changes will result in larger increases or decreases. The final enrichment score is where the running-sum statistic is the largest deviation from zero.</p> <p><strong>Step 2:</strong> Estimation of significance:</p> <p>The significance of the enrichment score is determined using permutation testing, which performs rearrangements of the data points to determine the likelihood of generating an enrichment score as large as the enrichment score calculated from the observed data. Essentially, for this step, the first permutation would reorder the log2 fold changes and randomly assign them to different genes, reorder the gene ranks based on these new log2 fold changes, and recalculate the enrichment score. The second permutation would reorder the log2 fold changes again and recalculate the enrichment score again, and this would continue for the total number of permutations run. Therefore, the number of permutations run will increase the confidence in the signficance estimates.</p> <p><strong>Step 3:</strong> Adjust for multiple test correction</p> <p>After all gene sets are tested, the enrichment scores are normalized for the size of the gene set, then the p-values are corrected for multiple testing.</p> <p>The GSEA output will yield the core genes in the gene sets that most highly contribute to the enrichment score. The genes output are generally the genes at or before the running sum reaches its maximum value (eg. the most influential genes driving the differences between conditions for that gene set).</p> <h4 id=performing-gsea>Performing GSEA<a class=headerlink href=#performing-gsea title="Permanent link">&para;</a></h4> <p>To perform the GSEA using KEGG gene sets with clusterProfiler, we can use the <code>gseKEGG()</code> function:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=c1>## GSEA using gene sets from KEGG pathways</span>
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=n>gseaKEGG</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>gseKEGG</span><span class=p>(</span><span class=n>geneList</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>foldchanges</span><span class=p>,</span><span class=w> </span><span class=c1># ordered named vector of fold changes (Entrez IDs are the associated names)</span>
<a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a><span class=w>              </span><span class=n>organism</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;hsa&quot;</span><span class=p>,</span><span class=w> </span><span class=c1># supported organisms listed below</span>
<a id=__codelineno-3-4 name=__codelineno-3-4 href=#__codelineno-3-4></a><span class=w>              </span><span class=n>pvalueCutoff</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.05</span><span class=p>,</span><span class=w> </span><span class=c1># padj cutoff value</span>
<a id=__codelineno-3-5 name=__codelineno-3-5 href=#__codelineno-3-5></a><span class=w>              </span><span class=n>verbose</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-3-6 name=__codelineno-3-6 href=#__codelineno-3-6></a>
<a id=__codelineno-3-7 name=__codelineno-3-7 href=#__codelineno-3-7></a><span class=c1>## Extract the GSEA results</span>
<a id=__codelineno-3-8 name=__codelineno-3-8 href=#__codelineno-3-8></a><span class=n>gseaKEGG_results</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>gseaKEGG</span><span class=o>@</span><span class=n>result</span>
</code></pre></div> <blockquote> <p><strong>NOTE:</strong> The organisms with KEGG pathway information are listed <a href=http://www.genome.jp/kegg/catalog/org_list.html>here</a>.</p> </blockquote> <p><strong>How many pathways are enriched?</strong> View the enriched pathways:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=c1>## Write GSEA results to file</span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=nf>View</span><span class=p>(</span><span class=n>gseaKEGG_results</span><span class=p>)</span>
<a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a>
<a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a><span class=nf>write.csv</span><span class=p>(</span><span class=n>gseaKEGG_results</span><span class=p>,</span><span class=w> </span><span class=s>&quot;Results/gseaOE_kegg.csv&quot;</span><span class=p>,</span><span class=w> </span><span class=n>quote</span><span class=o>=</span><span class=bp>F</span><span class=p>)</span>
</code></pre></div> <blockquote> <p><em><strong>NOTE:</strong> We will all get different results for the GSEA because the permutations performed use random reordering. If we would like to use the same permutations every time we run a function (i.e. we would like the same results every time we run the function), then we could use the <code>set.seed(123456)</code> function prior to running. The input to <code>set.seed()</code> could be any number, but if you would want the same results, then you would need to use the same number as input.</em></p> </blockquote> <p>Explore the GSEA plot of enrichment of one of the pathways in the ranked list:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=c1>## Plot the GSEA plot for a single enriched pathway, `hsa03040`</span>
<a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a><span class=nf>gseaplot</span><span class=p>(</span><span class=n>gseaKEGG</span><span class=p>,</span><span class=w> </span><span class=n>geneSetID</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;hsa03040&#39;</span><span class=p>)</span>
</code></pre></div> <p><img src=./img/08c_FA_GSEA/gsea_kegg_hsa03040.png style="display: block; margin: auto;"></p> <p>In this plot, the lines in plot represent the genes in the gene set ‘hsa03040’, and where they occur among the log2 fold changes. The largest positive log2 fold changes are on the left-hand side of the plot, while the largest negative log2 fold changes are on the right. The top plot shows the magnitude of the log2 fold changes for each gene, while the bottom plot shows the running sum, with the enrichment score peaking at the red dotted line (which is among the negative log2 fold changes).</p> <p>Use the <a href=http://bioconductor.org/packages/release/bioc/html/pathview.html>Pathview R package</a> to integrate the KEGG pathway data from clusterProfiler into pathway images:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=c1>#detach(&quot;package:dplyr&quot;, unload=TRUE) # first unload dplyr to avoid conflicts</span>
<a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a>
<a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a><span class=c1>## Output images for a single significant KEGG pathway</span>
<a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a><span class=nf>pathview</span><span class=p>(</span><span class=n>gene.data</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>foldchanges</span><span class=p>,</span>
<a id=__codelineno-6-5 name=__codelineno-6-5 href=#__codelineno-6-5></a><span class=w>              </span><span class=n>pathway.id</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;hsa03040&quot;</span><span class=p>,</span>
<a id=__codelineno-6-6 name=__codelineno-6-6 href=#__codelineno-6-6></a><span class=w>              </span><span class=n>species</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;hsa&quot;</span><span class=p>,</span>
<a id=__codelineno-6-7 name=__codelineno-6-7 href=#__codelineno-6-7></a><span class=w>              </span><span class=n>limit</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>list</span><span class=p>(</span><span class=n>gene</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2</span><span class=p>,</span><span class=w> </span><span class=c1># value gives the max/min limit for foldchanges</span>
<a id=__codelineno-6-8 name=__codelineno-6-8 href=#__codelineno-6-8></a><span class=w>              </span><span class=n>cpd</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=p>))</span>
</code></pre></div> <blockquote> <p><strong>NOTE:</strong> If the below error message occurs: <code>Error in detach("package:dplyr", unload = T) : invalid 'name' argument</code>, that means the dplyr package is not currently loaded. Ignore the message and continue to run pathview command.</p> </blockquote> <p><img src=./img/08c_FA_GSEA/hsa03040.pathview.png style="display: block; margin: auto;"></p> <blockquote> <p><strong>NOTE:</strong> Printing out Pathview images for all significant pathways can be easily performed as follows:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1>## Output images for all significant KEGG pathways</span>
<a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=n>get_kegg_plots</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>function</span><span class=p>(</span><span class=n>x</span><span class=p>)</span><span class=w> </span><span class=p>{</span>
<a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a><span class=w>   </span><span class=nf>pathview</span><span class=p>(</span><span class=n>gene.data</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>foldchanges</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-7-4 name=__codelineno-7-4 href=#__codelineno-7-4></a><span class=w>            </span><span class=n>pathway.id</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>gseaKEGG_results</span><span class=o>$</span><span class=n>ID</span><span class=p>[</span><span class=n>x</span><span class=p>],</span><span class=w> </span>
<a id=__codelineno-7-5 name=__codelineno-7-5 href=#__codelineno-7-5></a><span class=w>            </span><span class=n>species</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;hsa&quot;</span><span class=p>,</span>
<a id=__codelineno-7-6 name=__codelineno-7-6 href=#__codelineno-7-6></a><span class=w>            </span><span class=n>limit</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>list</span><span class=p>(</span><span class=n>gene</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2</span><span class=p>,</span><span class=w> </span><span class=n>cpd</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=p>))</span>
<a id=__codelineno-7-7 name=__codelineno-7-7 href=#__codelineno-7-7></a><span class=p>}</span>
<a id=__codelineno-7-8 name=__codelineno-7-8 href=#__codelineno-7-8></a>
<a id=__codelineno-7-9 name=__codelineno-7-9 href=#__codelineno-7-9></a><span class=n>purrr</span><span class=o>::</span><span class=nf>map</span><span class=p>(</span><span class=m>1</span><span class=o>:</span><span class=nf>length</span><span class=p>(</span><span class=n>gseaKEGG_results</span><span class=o>$</span><span class=n>ID</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-7-10 name=__codelineno-7-10 href=#__codelineno-7-10></a><span class=w>           </span><span class=n>get_kegg_plots</span><span class=p>)</span>
</code></pre></div> </blockquote> <p>Instead of exploring enrichment of KEGG gene sets, we can also explore the enrichment of BP Gene Ontology terms using gene set enrichment analysis:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=c1># GSEA using gene sets associated with BP Gene Ontology terms</span>
<a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a><span class=n>gseaGO</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>gseGO</span><span class=p>(</span><span class=n>geneList</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>foldchanges</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a><span class=w>              </span><span class=n>OrgDb</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>org.Hs.eg.db</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a><span class=w>              </span><span class=n>ont</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;BP&#39;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-8-5 name=__codelineno-8-5 href=#__codelineno-8-5></a><span class=w>              </span><span class=n>minGSSize</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>20</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-8-6 name=__codelineno-8-6 href=#__codelineno-8-6></a><span class=w>              </span><span class=n>pvalueCutoff</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.05</span><span class=p>,</span>
<a id=__codelineno-8-7 name=__codelineno-8-7 href=#__codelineno-8-7></a><span class=w>              </span><span class=n>verbose</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span><span class=w> </span>
<a id=__codelineno-8-8 name=__codelineno-8-8 href=#__codelineno-8-8></a>
<a id=__codelineno-8-9 name=__codelineno-8-9 href=#__codelineno-8-9></a><span class=n>gseaGO_results</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>gseaGO</span><span class=o>@</span><span class=n>result</span>
<a id=__codelineno-8-10 name=__codelineno-8-10 href=#__codelineno-8-10></a>
<a id=__codelineno-8-11 name=__codelineno-8-11 href=#__codelineno-8-11></a><span class=nf>gseaplot</span><span class=p>(</span><span class=n>gseaGO</span><span class=p>,</span><span class=w> </span><span class=n>geneSetID</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;GO:0007423&#39;</span><span class=p>)</span>
</code></pre></div> <p>There are other gene sets available for GSEA analysis in clusterProfiler (Disease Ontology, Reactome pathways, etc.). In addition, it is possible to supply your own gene set GMT file, such as a GMT for <a href=http://software.broadinstitute.org/gsea/msigdb/index.jsp>MSigDB</a> using special clusterProfiler functions as shown below:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=c1># DO NOT RUN</span>
<a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a>
<a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a><span class=n>BiocManager</span><span class=o>::</span><span class=nf>install</span><span class=p>(</span><span class=s>&quot;GSEABase&quot;</span><span class=p>)</span>
<a id=__codelineno-9-4 name=__codelineno-9-4 href=#__codelineno-9-4></a><span class=nf>library</span><span class=p>(</span><span class=n>GSEABase</span><span class=p>)</span>
<a id=__codelineno-9-5 name=__codelineno-9-5 href=#__codelineno-9-5></a>
<a id=__codelineno-9-6 name=__codelineno-9-6 href=#__codelineno-9-6></a><span class=c1># Load in GMT file of gene sets (we downloaded from the Broad Institute for MSigDB)</span>
<a id=__codelineno-9-7 name=__codelineno-9-7 href=#__codelineno-9-7></a>
<a id=__codelineno-9-8 name=__codelineno-9-8 href=#__codelineno-9-8></a><span class=n>c2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read.gmt</span><span class=p>(</span><span class=s>&quot;/data/c2.cp.v6.0.entrez.gmt.txt&quot;</span><span class=p>)</span>
<a id=__codelineno-9-9 name=__codelineno-9-9 href=#__codelineno-9-9></a>
<a id=__codelineno-9-10 name=__codelineno-9-10 href=#__codelineno-9-10></a><span class=n>msig</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>GSEA</span><span class=p>(</span><span class=n>foldchanges</span><span class=p>,</span><span class=w> </span><span class=n>TERM2GENE</span><span class=o>=</span><span class=n>c2</span><span class=p>,</span><span class=w> </span><span class=n>verbose</span><span class=o>=</span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-9-11 name=__codelineno-9-11 href=#__codelineno-9-11></a>
<a id=__codelineno-9-12 name=__codelineno-9-12 href=#__codelineno-9-12></a><span class=n>msig_df</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>data.frame</span><span class=p>(</span><span class=n>msig</span><span class=p>)</span>
</code></pre></div> <h2 id=pathway-topology-tools>Pathway topology tools<a class=headerlink href=#pathway-topology-tools title="Permanent link">&para;</a></h2> <p><img src=./img/08c_FA_GSEA/pathway_analysis.png style="display: block; margin: auto;"></p> <p>The last main type of functional analysis technique is pathway topology analysis. Pathway topology analysis often takes into account gene interaction information along with the fold changes and adjusted p-values from differential expression analysis to identify dysregulated pathways. Depending on the tool, pathway topology tools explore how genes interact with each other (e.g. activation, inhibition, phosphorylation, ubiquitination, etc.) to determine the pathway-level statistics. Pathway topology-based methods utilize the number and type of interactions between gene product (our DE genes) and other gene products to infer gene function or pathway association.</p> <p>For instance, the <a href=http://bioconductor.org/packages/release/bioc/html/SPIA.html>SPIA (Signaling Pathway Impact Analysis)</a> tool can be used to integrate the lists of differentially expressed genes, their fold changes, and pathway topology to identify affected pathways.</p> <h2 id=other-tools>Other Tools<a class=headerlink href=#other-tools title="Permanent link">&para;</a></h2> <h3 id=co-expression-clustering>Co-expression clustering<a class=headerlink href=#co-expression-clustering title="Permanent link">&para;</a></h3> <p>Co-expression clustering is often used to identify genes of novel pathways or networks by grouping genes together based on similar trends in expression. These tools are useful in identifying genes in a pathway, when their participation in a pathway and/or the pathway itself is unknown. These tools cluster genes with similar expression patterns to create ‘modules’ of co-expressed genes which often reflect functionally similar groups of genes. These ‘modules’ can then be compared across conditions or in a time-course experiment to identify any biologically relevant pathway or network information.</p> <p>You can visualize co-expression clustering using heatmaps, which should be viewed as suggestive only; serious classification of genes needs better methods.</p> <p>The way the tools perform clustering is by taking the entire expression matrix and computing pair-wise co-expression values. A network is then generated from which we explore the topology to make inferences on gene co-regulation. The <a href=https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/ >WGCNA</a> package (in R) is one example of a more sophisticated method for co-expression clustering.</p> <h2 id=resources-for-functional-analysis>Resources for functional analysis<a class=headerlink href=#resources-for-functional-analysis title="Permanent link">&para;</a></h2> <ul> <li>g:Profiler - <a href=http://biit.cs.ut.ee/gprofiler/index.cgi>http://biit.cs.ut.ee/gprofiler/index.cgi</a></li> <li>DAVID - <a href=http://david.abcc.ncifcrf.gov/tools.jsp>http://david.abcc.ncifcrf.gov/tools.jsp</a></li> <li>clusterProfiler - <a href=http://bioconductor.org/packages/release/bioc/html/clusterProfiler.html>http://bioconductor.org/packages/release/bioc/html/clusterProfiler.html</a></li> <li>GeneMANIA - <a href=http://www.genemania.org/ >http://www.genemania.org/</a></li> <li>GenePattern - <a href=http://www.broadinstitute.org/cancer/software/genepattern/ >http://www.broadinstitute.org/cancer/software/genepattern/</a> (need to register)</li> <li>WebGestalt - <a href=http://bioinfo.vanderbilt.edu/webgestalt/ >http://bioinfo.vanderbilt.edu/webgestalt/</a> (need to register)</li> <li>AmiGO - <a href=http://amigo.geneontology.org/amigo>http://amigo.geneontology.org/amigo</a></li> <li>ReviGO (visualizing GO analysis, input is GO terms) - <a href=http://revigo.irb.hr/ >http://revigo.irb.hr/</a></li> <li>WGCNA - <a href=https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/ >https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/</a></li> <li>GSEA - <a href=http://software.broadinstitute.org/gsea/index.jsp>http://software.broadinstitute.org/gsea/index.jsp</a></li> <li>SPIA - <a href=https://www.bioconductor.org/packages/release/bioc/html/SPIA.html>https://www.bioconductor.org/packages/release/bioc/html/SPIA.html</a></li> <li>GAGE/Pathview - <a href=http://www.bioconductor.org/packages/release/bioc/html/gage.html>http://www.bioconductor.org/packages/release/bioc/html/gage.html</a></li> </ul> <hr> <p><em>This lesson was originally developed by members of the teaching team (Mary Piper, Radhika Khetani) at the <a href=http://bioinformatics.sph.harvard.edu/ >Harvard Chan Bioinformatics Core (HBC)</a>.</em></p> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=08b_FA_overrepresentation.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Over representation analysis" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Over representation analysis </div> </div> </a> <a href=09_summarized_workflow.html class="md-footer__link md-footer__link--next" aria-label="Next: Summarized workflow" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Summarized workflow </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.footer", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.db81ec45.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=assets/javascripts/bundle.a00a7c5e.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>