<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="This course is an introduction for how to approach bulk RNAseq data, starting from the sequencing reads.  It will provide an overview of the fundamentals of RNAseq analysis, including read preprocessing, data normalization, data exploration with PCAs and heatmaps, performing differential expression analysis and annotation of the differentially expressed genes.  Participants will also learn how to evaluate confounding and batch effects in the data.  The course will further touch upon laboratory protocols, library preparation, and experimental design of RNA sequencing experiments, especially about how they influence downstream bioinformatic analysis. â€‚ "><meta name=author content="Jose Alejandro Romero Herrera"><link rel=canonical href=http://hds-sandbox.github.io/09_summarized_workflow.html><link rel=prev href=08c_FA_GSEA.html><link rel=icon href=img/favicon.png><meta name=generator content="mkdocs-1.4.2, mkdocs-material-9.0.8"><title>Summarised workflow - Bulk RNAseq data analysis</title><link rel=stylesheet href=assets/stylesheets/main.0d440cfe.min.css><link rel=stylesheet href=assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=stylesheets/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#summarised-workflow class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=index.html title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bulk RNAseq data analysis </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Summarised workflow </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=02_experimental_planning.html class=md-tabs__link> Experimental planning </a> </li> <li class=md-tabs__item> <a href=03_data_explanation.html class=md-tabs__link> Data Explanation </a> </li> <li class=md-tabs__item> <a href=04a_preprocessing.html class=md-tabs__link> Read Preprocessing </a> </li> <li class=md-tabs__item> <a href=05a_data_analysis_setup.html class=md-tabs__link> Analysing RNAseq data </a> </li> <li class=md-tabs__item> <a href=09_summarized_workflow.html class="md-tabs__link md-tabs__link--active"> Summarized workflow </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation hidden> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=index.html title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component=logo> <img src=img/logo.png alt=logo> </a> Bulk RNAseq data analysis </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/bulk_RNAseq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/bulk_RNAseq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=index.html class=md-nav__link> Course Introduction </a> </li> <li class=md-nav__item> <a href=02_experimental_planning.html class=md-nav__link> Experimental planning </a> </li> <li class=md-nav__item> <a href=03_data_explanation.html class=md-nav__link> Data Explanation </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> Read Preprocessing <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Read Preprocessing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=04a_preprocessing.html class=md-nav__link> Preprocessing steps </a> </li> <li class=md-nav__item> <a href=04b_pipelines.html class=md-nav__link> Nextflow and nf-core pipelines </a> </li> <li class=md-nav__item> <a href=04c_preprocessing_setup.html class=md-nav__link> UCloud setup </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex=0> Analysing RNAseq data <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Analysing RNAseq data </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=05a_data_analysis_setup.html class=md-nav__link> Ucloud setup </a> </li> <li class=md-nav__item> <a href=05b_count_matrix.html class=md-nav__link> Count matrix </a> </li> <li class=md-nav__item> <a href=05c_count_normalization.html class=md-nav__link> Normalization </a> </li> <li class=md-nav__item> <a href=06_exploratory_analysis.html class=md-nav__link> Exploratory Analysis </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_5> <label class=md-nav__link for=__nav_5_5 id=__nav_5_5_label tabindex=0> Differential Expression <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5_5> <span class="md-nav__icon md-icon"></span> Differential Expression </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=07_extra_contrast_design.html class=md-nav__link> Design formula </a> </li> <li class=md-nav__item> <a href=07a_DEA.html class=md-nav__link> DESeq2 </a> </li> <li class=md-nav__item> <a href=07b_hypothesis_testing.html class=md-nav__link> Hypothesis testing </a> </li> <li class=md-nav__item> <a href=07c_DEA_visualization.html class=md-nav__link> Visualization </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5_6> <label class=md-nav__link for=__nav_5_6 id=__nav_5_6_label tabindex=0> Functional Analysis <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_5_6_label aria-expanded=false> <label class=md-nav__title for=__nav_5_6> <span class="md-nav__icon md-icon"></span> Functional Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=08a_FA_genomic_annotation.html class=md-nav__link> Gene annotation </a> </li> <li class=md-nav__item> <a href=08b_FA_overrepresentation.html class=md-nav__link> Over representation analysis </a> </li> <li class=md-nav__item> <a href=08c_FA_GSEA.html class=md-nav__link> Class scoring </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Summarized workflow <span class="md-nav__icon md-icon"></span> </label> <a href=09_summarized_workflow.html class="md-nav__link md-nav__link--active"> Summarized workflow </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#libraries class=md-nav__link> Libraries </a> </li> <li class=md-nav__item> <a href=#obtaining-gene-level-counts-from-your-preprocessing-and-create-deseq-object class=md-nav__link> Obtaining gene-level counts from your preprocessing and create DESeq object </a> <nav class=md-nav aria-label="Obtaining gene-level counts from your preprocessing and create DESeq object"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#if-you-have-a-traditional-raw-count-matrix class=md-nav__link> If you have a traditional raw count matrix </a> </li> <li class=md-nav__item> <a href=#if-you-have-pseudocounts class=md-nav__link> If you have pseudocounts </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#exploratory-data-analysis class=md-nav__link> Exploratory data analysis </a> <nav class=md-nav aria-label="Exploratory data analysis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#prefiltering-low-count-genes class=md-nav__link> Prefiltering low count genes </a> </li> <li class=md-nav__item> <a href=#rlog-transformation class=md-nav__link> Rlog transformation </a> </li> <li class=md-nav__item> <a href=#pca class=md-nav__link> PCA </a> </li> <li class=md-nav__item> <a href=#heatmaps class=md-nav__link> Heatmaps </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#run-deseq2 class=md-nav__link> Run DESeq2: </a> </li> <li class=md-nav__item> <a href=#check-the-fit-of-the-dispersion-estimates class=md-nav__link> Check the fit of the dispersion estimates </a> </li> <li class=md-nav__item> <a href=#create-contrasts-to-perform-wald-testing-or-the-shrunken-log2-fold-changes-between-specific-conditions class=md-nav__link> Create contrasts to perform Wald testing or the shrunken log2 fold changes between specific conditions </a> </li> <li class=md-nav__item> <a href=#output-significant-results class=md-nav__link> Output significant results: </a> </li> <li class=md-nav__item> <a href=#visualize-results-volcano-plots-heatmaps-normalized-counts-plots-of-top-genes-etc class=md-nav__link> Visualize results: volcano plots, heatmaps, normalized counts plots of top genes, etc. </a> <nav class=md-nav aria-label="Visualize results: volcano plots, heatmaps, normalized counts plots of top genes, etc."> <ul class=md-nav__list> <li class=md-nav__item> <a href=#maplot class=md-nav__link> MAplot </a> </li> <li class=md-nav__item> <a href=#volcano-plot-with-labels-top-n-genes class=md-nav__link> Volcano plot with labels (top N genes) </a> </li> <li class=md-nav__item> <a href=#heatmap-of-differentially-expressed-genes class=md-nav__link> Heatmap of differentially expressed genes </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#perform-analysis-to-extract-functional-significance-of-results-go-or-kegg-enrichment-gsea-etc class=md-nav__link> Perform analysis to extract functional significance of results: GO or KEGG enrichment, GSEA, etc. </a> <nav class=md-nav aria-label="Perform analysis to extract functional significance of results: GO or KEGG enrichment, GSEA, etc."> <ul class=md-nav__list> <li class=md-nav__item> <a href=#annotate-with-annotables class=md-nav__link> Annotate with annotables </a> </li> <li class=md-nav__item> <a href=#perform-enrichment-analysis-of-go-terms-can-be-done-as-well-with-kegg-pathways class=md-nav__link> Perform enrichment analysis of GO terms (can be done as well with KEGG pathways) </a> </li> <li class=md-nav__item> <a href=#visualize-result class=md-nav__link> Visualize result </a> </li> <li class=md-nav__item> <a href=#cnetplot class=md-nav__link> Cnetplot </a> </li> <li class=md-nav__item> <a href=#perform-gsea-analysis-of-kegg-pathways-can-be-done-as-well-with-go-terms class=md-nav__link> Perform GSEA analysis of KEGG pathways (can be done as well with GO terms) </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#make-sure-to-output-the-versions-of-all-tools-used-in-the-de-analysis class=md-nav__link> Make sure to output the versions of all tools used in the DE analysis: </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#libraries class=md-nav__link> Libraries </a> </li> <li class=md-nav__item> <a href=#obtaining-gene-level-counts-from-your-preprocessing-and-create-deseq-object class=md-nav__link> Obtaining gene-level counts from your preprocessing and create DESeq object </a> <nav class=md-nav aria-label="Obtaining gene-level counts from your preprocessing and create DESeq object"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#if-you-have-a-traditional-raw-count-matrix class=md-nav__link> If you have a traditional raw count matrix </a> </li> <li class=md-nav__item> <a href=#if-you-have-pseudocounts class=md-nav__link> If you have pseudocounts </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#exploratory-data-analysis class=md-nav__link> Exploratory data analysis </a> <nav class=md-nav aria-label="Exploratory data analysis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#prefiltering-low-count-genes class=md-nav__link> Prefiltering low count genes </a> </li> <li class=md-nav__item> <a href=#rlog-transformation class=md-nav__link> Rlog transformation </a> </li> <li class=md-nav__item> <a href=#pca class=md-nav__link> PCA </a> </li> <li class=md-nav__item> <a href=#heatmaps class=md-nav__link> Heatmaps </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#run-deseq2 class=md-nav__link> Run DESeq2: </a> </li> <li class=md-nav__item> <a href=#check-the-fit-of-the-dispersion-estimates class=md-nav__link> Check the fit of the dispersion estimates </a> </li> <li class=md-nav__item> <a href=#create-contrasts-to-perform-wald-testing-or-the-shrunken-log2-fold-changes-between-specific-conditions class=md-nav__link> Create contrasts to perform Wald testing or the shrunken log2 fold changes between specific conditions </a> </li> <li class=md-nav__item> <a href=#output-significant-results class=md-nav__link> Output significant results: </a> </li> <li class=md-nav__item> <a href=#visualize-results-volcano-plots-heatmaps-normalized-counts-plots-of-top-genes-etc class=md-nav__link> Visualize results: volcano plots, heatmaps, normalized counts plots of top genes, etc. </a> <nav class=md-nav aria-label="Visualize results: volcano plots, heatmaps, normalized counts plots of top genes, etc."> <ul class=md-nav__list> <li class=md-nav__item> <a href=#maplot class=md-nav__link> MAplot </a> </li> <li class=md-nav__item> <a href=#volcano-plot-with-labels-top-n-genes class=md-nav__link> Volcano plot with labels (top N genes) </a> </li> <li class=md-nav__item> <a href=#heatmap-of-differentially-expressed-genes class=md-nav__link> Heatmap of differentially expressed genes </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#perform-analysis-to-extract-functional-significance-of-results-go-or-kegg-enrichment-gsea-etc class=md-nav__link> Perform analysis to extract functional significance of results: GO or KEGG enrichment, GSEA, etc. </a> <nav class=md-nav aria-label="Perform analysis to extract functional significance of results: GO or KEGG enrichment, GSEA, etc."> <ul class=md-nav__list> <li class=md-nav__item> <a href=#annotate-with-annotables class=md-nav__link> Annotate with annotables </a> </li> <li class=md-nav__item> <a href=#perform-enrichment-analysis-of-go-terms-can-be-done-as-well-with-kegg-pathways class=md-nav__link> Perform enrichment analysis of GO terms (can be done as well with KEGG pathways) </a> </li> <li class=md-nav__item> <a href=#visualize-result class=md-nav__link> Visualize result </a> </li> <li class=md-nav__item> <a href=#cnetplot class=md-nav__link> Cnetplot </a> </li> <li class=md-nav__item> <a href=#perform-gsea-analysis-of-kegg-pathways-can-be-done-as-well-with-go-terms class=md-nav__link> Perform GSEA analysis of KEGG pathways (can be done as well with GO terms) </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#make-sure-to-output-the-versions-of-all-tools-used-in-the-de-analysis class=md-nav__link> Make sure to output the versions of all tools used in the DE analysis: </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=summarised-workflow>Summarised workflow<a class=headerlink href=#summarised-workflow title="Permanent link">&para;</a></h1> <p>We have detailed the various steps in a differential expression analysis workflow, providing theory with example code. To provide a more succinct reference for the code needed to run a DGE analysis, we have summarized the steps in an analysis below:</p> <h2 id=libraries>Libraries<a class=headerlink href=#libraries title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=nf>library</span><span class=p>(</span><span class=n>tidyverse</span><span class=p>)</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a><span class=nf>library</span><span class=p>(</span><span class=n>DESeq2</span><span class=p>)</span>
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a><span class=nf>library</span><span class=p>(</span><span class=n>ggrepel</span><span class=p>)</span>
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=nf>library</span><span class=p>(</span><span class=n>pheatmap</span><span class=p>)</span>
<a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a><span class=nf>library</span><span class=p>(</span><span class=n>annotables</span><span class=p>)</span>
<a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a><span class=nf>library</span><span class=p>(</span><span class=n>clusterProfiler</span><span class=p>)</span>
<a id=__codelineno-0-7 name=__codelineno-0-7 href=#__codelineno-0-7></a><span class=nf>library</span><span class=p>(</span><span class=n>DOSE</span><span class=p>)</span>
<a id=__codelineno-0-8 name=__codelineno-0-8 href=#__codelineno-0-8></a><span class=nf>library</span><span class=p>(</span><span class=n>pathview</span><span class=p>)</span>
<a id=__codelineno-0-9 name=__codelineno-0-9 href=#__codelineno-0-9></a><span class=nf>library</span><span class=p>(</span><span class=n>org.Hs.eg.db</span><span class=p>)</span>
<a id=__codelineno-0-10 name=__codelineno-0-10 href=#__codelineno-0-10></a><span class=nf>library</span><span class=p>(</span><span class=n>tximport</span><span class=p>)</span>
<a id=__codelineno-0-11 name=__codelineno-0-11 href=#__codelineno-0-11></a><span class=nf>library</span><span class=p>(</span><span class=n>RColorBrewer</span><span class=p>)</span>
</code></pre></div> <h2 id=obtaining-gene-level-counts-from-your-preprocessing-and-create-deseq-object>Obtaining gene-level counts from your preprocessing and create DESeq object<a class=headerlink href=#obtaining-gene-level-counts-from-your-preprocessing-and-create-deseq-object title="Permanent link">&para;</a></h2> <h3 id=if-you-have-a-traditional-raw-count-matrix>If you have a traditional raw count matrix<a class=headerlink href=#if-you-have-a-traditional-raw-count-matrix title="Permanent link">&para;</a></h3> <p>Load data and metadata</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=n>data</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read_table</span><span class=p>(</span><span class=s>&quot;../Data/Mov10_full_counts.txt&quot;</span><span class=p>)</span><span class=w> </span>
<a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a>
<a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a><span class=n>meta</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read_table</span><span class=p>(</span><span class=s>&quot;../Data/Mov10_full_meta.txt&quot;</span><span class=p>)</span>
</code></pre></div> <p>Check that the row names of the metadata equal the column names of the <strong>raw counts</strong> data</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=nf>all</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>data</span><span class=p>)[</span><span class=m>-1</span><span class=p>]</span><span class=w> </span><span class=o>==</span><span class=w> </span><span class=n>meta</span><span class=o>$</span><span class=n>sample</span><span class=p>)</span>
</code></pre></div> <p>Create DESeq2Dataset object</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=n>dds</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DESeqDataSetFromMatrix</span><span class=p>(</span><span class=n>countData</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>data</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&quot;GeneSymbol&quot;</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=w>                              </span><span class=n>colData</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>meta</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&quot;sample&quot;</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a><span class=w>                              </span><span class=n>design</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=o>~</span><span class=w> </span><span class=n>condition</span><span class=p>)</span>
</code></pre></div> <h3 id=if-you-have-pseudocounts>If you have pseudocounts<a class=headerlink href=#if-you-have-pseudocounts title="Permanent link">&para;</a></h3> <p>Load samplesheet with all our metadata from our pipeline</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=c1># Load data, metadata and tx2gene and create a txi object</span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=n>meta</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read_csv</span><span class=p>(</span><span class=s>&quot;../Data/samplesheet.csv&quot;</span><span class=p>)</span>
</code></pre></div> <p>Create a list of salmon results</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=n>dir</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=s>&quot;/work/sequencing_data/Preprocessing_backup/results_salmon/salmon&quot;</span>
<a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a><span class=n>tx2gene</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read_table</span><span class=p>(</span><span class=nf>file.path</span><span class=p>(</span><span class=n>dir</span><span class=p>,</span><span class=s>&quot;salmon_tx2gene.tsv&quot;</span><span class=p>),</span><span class=w> </span><span class=n>col_names</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;transcript_ID&quot;</span><span class=p>,</span><span class=s>&quot;gene_ID&quot;</span><span class=p>,</span><span class=s>&quot;gene_symbol&quot;</span><span class=p>))</span>
<a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a>
<a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a><span class=c1># Get all salmon results files</span>
<a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a><span class=n>files</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>file.path</span><span class=p>(</span><span class=n>dir</span><span class=p>,</span><span class=w> </span><span class=n>meta</span><span class=o>$</span><span class=n>sample</span><span class=p>,</span><span class=w> </span><span class=s>&quot;quant.sf&quot;</span><span class=p>)</span>
<a id=__codelineno-5-6 name=__codelineno-5-6 href=#__codelineno-5-6></a><span class=nf>names</span><span class=p>(</span><span class=n>files</span><span class=p>)</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>meta</span><span class=o>$</span><span class=n>sample</span>
</code></pre></div> <p>Create txi object</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=n>txi</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>tximport</span><span class=p>(</span><span class=n>files</span><span class=p>,</span><span class=w> </span><span class=n>type</span><span class=o>=</span><span class=s>&quot;salmon&quot;</span><span class=p>,</span><span class=w> </span><span class=n>tx2gene</span><span class=o>=</span><span class=n>tx2gene</span><span class=p>,</span><span class=w> </span><span class=n>countsFromAbundance</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;lengthScaledTPM&quot;</span><span class=p>,</span><span class=w> </span><span class=n>ignoreTxVersion</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
</code></pre></div> <p>Create dds object</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=n>dds</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DESeqDataSetFromTximport</span><span class=p>(</span><span class=n>txi</span><span class=p>,</span>
<a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=w>                                   </span><span class=n>colData</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>meta</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&quot;sample&quot;</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a><span class=w>                              </span><span class=n>design</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=o>~</span><span class=w> </span><span class=n>condition</span><span class=p>)</span>
</code></pre></div> <h2 id=exploratory-data-analysis>Exploratory data analysis<a class=headerlink href=#exploratory-data-analysis title="Permanent link">&para;</a></h2> <p>Prefiltering low count genes + PCA &amp; hierarchical clustering - identifying outliers and sources of variation in the data:</p> <h3 id=prefiltering-low-count-genes>Prefiltering low count genes<a class=headerlink href=#prefiltering-low-count-genes title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=n>keep</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>rowSums</span><span class=p>(</span><span class=nf>counts</span><span class=p>(</span><span class=n>dds</span><span class=p>))</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=m>0</span>
<a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a><span class=n>dds</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>dds</span><span class=p>[</span><span class=n>keep</span><span class=p>,]</span>
</code></pre></div> <h3 id=rlog-transformation>Rlog transformation<a class=headerlink href=#rlog-transformation title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=c1># Transform counts for data visualization</span>
<a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a><span class=n>rld</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>rlog</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a><span class=w>            </span><span class=n>blind</span><span class=o>=</span><span class=kc>TRUE</span><span class=p>)</span>
</code></pre></div> <h3 id=pca>PCA<a class=headerlink href=#pca title="Permanent link">&para;</a></h3> <p>Plot PCA</p> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=nf>plotPCA</span><span class=p>(</span><span class=n>rld</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=w>        </span><span class=n>intgroup</span><span class=o>=</span><span class=s>&quot;condition&quot;</span><span class=p>)</span>
</code></pre></div> <h3 id=heatmaps>Heatmaps<a class=headerlink href=#heatmaps title="Permanent link">&para;</a></h3> <p>Extract the rlog matrix from the object</p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=n>rld_mat</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>assay</span><span class=p>(</span><span class=n>rld</span><span class=p>)</span>
<a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a><span class=n>rld_cor</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>cor</span><span class=p>(</span><span class=n>rld_mat</span><span class=p>)</span><span class=w> </span><span class=c1># Pearson correlation betweeen samples</span>
<a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a><span class=n>rld_dist</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>as.matrix</span><span class=p>(</span><span class=nf>dist</span><span class=p>(</span><span class=nf>t</span><span class=p>(</span><span class=nf>assay</span><span class=p>(</span><span class=n>rld</span><span class=p>))))</span><span class=w> </span><span class=c1>#distances are computed by rows, so we need to transponse the matrix</span>
</code></pre></div> <p>Plot heatmap of correlations</p> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=nf>pheatmap</span><span class=p>(</span><span class=n>rld_cor</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a><span class=w>         </span><span class=n>annotation</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>meta</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&quot;sample&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>select</span><span class=p>(</span><span class=s>&quot;condition&quot;</span><span class=p>))</span>
</code></pre></div> <p>Plot heatmap of distances with a new color range</p> <div class=highlight><pre><span></span><code><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=n>heat.colors</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>brewer.pal</span><span class=p>(</span><span class=m>6</span><span class=p>,</span><span class=w> </span><span class=s>&quot;Blues&quot;</span><span class=p>)</span><span class=w> </span><span class=c1># Colors from the RColorBrewer package (only 6)</span>
<a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a><span class=n>heat.colors</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>colorRampPalette</span><span class=p>(</span><span class=n>heat.colors</span><span class=p>)(</span><span class=m>100</span><span class=p>)</span><span class=w> </span><span class=c1># Interpolate 100 colors</span>
<a id=__codelineno-13-3 name=__codelineno-13-3 href=#__codelineno-13-3></a>
<a id=__codelineno-13-4 name=__codelineno-13-4 href=#__codelineno-13-4></a><span class=nf>pheatmap</span><span class=p>(</span><span class=n>rld_dist</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-13-5 name=__codelineno-13-5 href=#__codelineno-13-5></a><span class=w>         </span><span class=n>annotation</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>meta</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&quot;sample&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>select</span><span class=p>(</span><span class=s>&quot;condition&quot;</span><span class=p>),</span>
<a id=__codelineno-13-6 name=__codelineno-13-6 href=#__codelineno-13-6></a><span class=w>         </span><span class=n>color</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>heat.colors</span><span class=p>)</span>
</code></pre></div> <h2 id=run-deseq2>Run DESeq2:<a class=headerlink href=#run-deseq2 title="Permanent link">&para;</a></h2> <p><strong>Optional step</strong> - Re-create DESeq2 dataset if the design formula has changed after QC analysis in include other sources of variation using</p> <div class=highlight><pre><span></span><code><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=c1># dds &lt;- DESeqDataSetFromMatrix(data, colData = meta, design = ~ covariate + condition)</span>
</code></pre></div> <p>Run DEseq2</p> <div class=highlight><pre><span></span><code><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=c1># Run DESeq2 differential expression analysis</span>
<a id=__codelineno-15-2 name=__codelineno-15-2 href=#__codelineno-15-2></a><span class=n>dds</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DESeq</span><span class=p>(</span><span class=n>dds</span><span class=p>)</span>
</code></pre></div> <p><strong>Optional step</strong> - Output normalized counts to save as a file to access outside RStudio using</p> <div class=highlight><pre><span></span><code><a id=__codelineno-16-1 name=__codelineno-16-1 href=#__codelineno-16-1></a><span class=n>normalized_counts</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>counts</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>normalized</span><span class=o>=</span><span class=kc>TRUE</span><span class=p>)</span>
</code></pre></div> <h2 id=check-the-fit-of-the-dispersion-estimates>Check the fit of the dispersion estimates<a class=headerlink href=#check-the-fit-of-the-dispersion-estimates title="Permanent link">&para;</a></h2> <p>Plot dispersion estimates</p> <div class=highlight><pre><span></span><code><a id=__codelineno-17-1 name=__codelineno-17-1 href=#__codelineno-17-1></a><span class=nf>plotDispEsts</span><span class=p>(</span><span class=n>dds</span><span class=p>)</span>
</code></pre></div> <h2 id=create-contrasts-to-perform-wald-testing-or-the-shrunken-log2-fold-changes-between-specific-conditions>Create contrasts to perform Wald testing or the shrunken log2 fold changes between specific conditions<a class=headerlink href=#create-contrasts-to-perform-wald-testing-or-the-shrunken-log2-fold-changes-between-specific-conditions title="Permanent link">&para;</a></h2> <p>Formal LFC calculation</p> <div class=highlight><pre><span></span><code><a id=__codelineno-18-1 name=__codelineno-18-1 href=#__codelineno-18-1></a><span class=c1># Specify contrast for comparison of interest</span>
<a id=__codelineno-18-2 name=__codelineno-18-2 href=#__codelineno-18-2></a><span class=n>contrast</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;condition&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;MOV10_overexpression&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;control&quot;</span><span class=p>)</span>
<a id=__codelineno-18-3 name=__codelineno-18-3 href=#__codelineno-18-3></a>
<a id=__codelineno-18-4 name=__codelineno-18-4 href=#__codelineno-18-4></a><span class=c1># Output results of Wald test for contrast</span>
<a id=__codelineno-18-5 name=__codelineno-18-5 href=#__codelineno-18-5></a><span class=n>res</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>results</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-18-6 name=__codelineno-18-6 href=#__codelineno-18-6></a><span class=w>               </span><span class=n>contrast</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>contrast</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-18-7 name=__codelineno-18-7 href=#__codelineno-18-7></a><span class=w>               </span><span class=n>alpha</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.05</span><span class=p>)</span>
</code></pre></div> <p>Shrinkage</p> <div class=highlight><pre><span></span><code><a id=__codelineno-19-1 name=__codelineno-19-1 href=#__codelineno-19-1></a><span class=c1># Get name of the contrast you would like to use</span>
<a id=__codelineno-19-2 name=__codelineno-19-2 href=#__codelineno-19-2></a><span class=nf>resultsNames</span><span class=p>(</span><span class=n>dds</span><span class=p>)</span>
<a id=__codelineno-19-3 name=__codelineno-19-3 href=#__codelineno-19-3></a>
<a id=__codelineno-19-4 name=__codelineno-19-4 href=#__codelineno-19-4></a><span class=c1># Shrink the log2 fold changes to be more accurate</span>
<a id=__codelineno-19-5 name=__codelineno-19-5 href=#__codelineno-19-5></a><span class=n>res</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>lfcShrink</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-19-6 name=__codelineno-19-6 href=#__codelineno-19-6></a><span class=w>                 </span><span class=n>coef</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;condition_MOV10_overexpression_vs_control&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-19-7 name=__codelineno-19-7 href=#__codelineno-19-7></a><span class=w>                 </span><span class=n>type</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;apeglm&quot;</span><span class=p>)</span>
</code></pre></div> <h2 id=output-significant-results>Output significant results:<a class=headerlink href=#output-significant-results title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code><a id=__codelineno-20-1 name=__codelineno-20-1 href=#__codelineno-20-1></a><span class=c1># Set thresholds</span>
<a id=__codelineno-20-2 name=__codelineno-20-2 href=#__codelineno-20-2></a><span class=n>padj.cutoff</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=m>0.05</span>
<a id=__codelineno-20-3 name=__codelineno-20-3 href=#__codelineno-20-3></a>
<a id=__codelineno-20-4 name=__codelineno-20-4 href=#__codelineno-20-4></a><span class=c1># Turn the results object into a tibble for use with tidyverse functions</span>
<a id=__codelineno-20-5 name=__codelineno-20-5 href=#__codelineno-20-5></a><span class=n>res_tbl</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-20-6 name=__codelineno-20-6 href=#__codelineno-20-6></a><span class=w>  </span><span class=nf>data.frame</span><span class=p>()</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-20-7 name=__codelineno-20-7 href=#__codelineno-20-7></a><span class=w>  </span><span class=nf>rownames_to_column</span><span class=p>(</span><span class=n>var</span><span class=o>=</span><span class=s>&quot;gene&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-20-8 name=__codelineno-20-8 href=#__codelineno-20-8></a><span class=w>  </span><span class=nf>as_tibble</span><span class=p>()</span>
<a id=__codelineno-20-9 name=__codelineno-20-9 href=#__codelineno-20-9></a>
<a id=__codelineno-20-10 name=__codelineno-20-10 href=#__codelineno-20-10></a><span class=c1># Subset the significant results</span>
<a id=__codelineno-20-11 name=__codelineno-20-11 href=#__codelineno-20-11></a><span class=n>sig_res</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>filter</span><span class=p>(</span><span class=n>res_tbl</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-20-12 name=__codelineno-20-12 href=#__codelineno-20-12></a><span class=w>                  </span><span class=n>padj</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=n>padj.cutoff</span><span class=p>)</span>
</code></pre></div> <h2 id=visualize-results-volcano-plots-heatmaps-normalized-counts-plots-of-top-genes-etc>Visualize results: volcano plots, heatmaps, normalized counts plots of top genes, etc.<a class=headerlink href=#visualize-results-volcano-plots-heatmaps-normalized-counts-plots-of-top-genes-etc title="Permanent link">&para;</a></h2> <p>Function to get gene_IDs based on gene names. The function will take as input a vector of gene names of interest, the tx2gene dataframe and the dds object that you analyzed.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-21-1 name=__codelineno-21-1 href=#__codelineno-21-1></a><span class=n>lookup</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>function</span><span class=p>(</span><span class=n>gene_name</span><span class=p>,</span><span class=w> </span><span class=n>tx2gene</span><span class=p>,</span><span class=w> </span><span class=n>dds</span><span class=p>){</span>
<a id=__codelineno-21-2 name=__codelineno-21-2 href=#__codelineno-21-2></a><span class=w>  </span><span class=n>hits</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>tx2gene</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>select</span><span class=p>(</span><span class=n>gene_symbol</span><span class=p>,</span><span class=w> </span><span class=n>gene_ID</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>distinct</span><span class=p>()</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-21-3 name=__codelineno-21-3 href=#__codelineno-21-3></a><span class=w>    </span><span class=nf>filter</span><span class=p>(</span><span class=n>gene_symbol</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=n>gene_name</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=n>gene_ID</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=nf>rownames</span><span class=p>(</span><span class=n>dds</span><span class=p>))</span>
<a id=__codelineno-21-4 name=__codelineno-21-4 href=#__codelineno-21-4></a><span class=w>  </span><span class=nf>return</span><span class=p>(</span><span class=n>hits</span><span class=p>)</span>
<a id=__codelineno-21-5 name=__codelineno-21-5 href=#__codelineno-21-5></a><span class=p>}</span>
<a id=__codelineno-21-6 name=__codelineno-21-6 href=#__codelineno-21-6></a>
<a id=__codelineno-21-7 name=__codelineno-21-7 href=#__codelineno-21-7></a><span class=nf>lookup</span><span class=p>(</span><span class=n>gene_name</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;MOV10&quot;</span><span class=p>,</span><span class=w> </span><span class=n>tx2gene</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>tx2gene</span><span class=p>,</span><span class=w> </span><span class=n>dds</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>dds</span><span class=p>)</span>
</code></pre></div> <p>Plot expression for single gene</p> <div class=highlight><pre><span></span><code><a id=__codelineno-22-1 name=__codelineno-22-1 href=#__codelineno-22-1></a><span class=nf>plotCounts</span><span class=p>(</span><span class=n>dds</span><span class=p>,</span><span class=w> </span><span class=n>gene</span><span class=o>=</span><span class=s>&quot;ENSG00000155363&quot;</span><span class=p>,</span><span class=w> </span><span class=n>intgroup</span><span class=o>=</span><span class=s>&quot;condition&quot;</span><span class=p>)</span>
</code></pre></div> <h3 id=maplot>MAplot<a class=headerlink href=#maplot title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><a id=__codelineno-23-1 name=__codelineno-23-1 href=#__codelineno-23-1></a><span class=nf>plotMA</span><span class=p>(</span><span class=n>res</span><span class=p>)</span>
</code></pre></div> <h3 id=volcano-plot-with-labels-top-n-genes>Volcano plot with labels (top N genes)<a class=headerlink href=#volcano-plot-with-labels-top-n-genes title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><a id=__codelineno-24-1 name=__codelineno-24-1 href=#__codelineno-24-1></a><span class=c1>## Obtain logical vector where TRUE values denote padj values &lt; 0.05 and fold change &gt; 1.5 in either direction</span>
<a id=__codelineno-24-2 name=__codelineno-24-2 href=#__codelineno-24-2></a><span class=n>res_tbl</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_tbl</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-24-3 name=__codelineno-24-3 href=#__codelineno-24-3></a><span class=nf>mutate</span><span class=p>(</span><span class=n>threshold</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>padj</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=m>0.05</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=nf>abs</span><span class=p>(</span><span class=n>log2FoldChange</span><span class=p>)</span><span class=w> </span><span class=o>&gt;=</span><span class=w> </span><span class=m>0.58</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-25-1 name=__codelineno-25-1 href=#__codelineno-25-1></a><span class=c1>## Create an empty column to indicate which genes to label</span>
<a id=__codelineno-25-2 name=__codelineno-25-2 href=#__codelineno-25-2></a><span class=n>res_tbl</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_tbl</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>mutate</span><span class=p>(</span><span class=n>genelabels</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;&quot;</span><span class=p>)</span>
<a id=__codelineno-25-3 name=__codelineno-25-3 href=#__codelineno-25-3></a>
<a id=__codelineno-25-4 name=__codelineno-25-4 href=#__codelineno-25-4></a><span class=c1>## Sort by padj values </span>
<a id=__codelineno-25-5 name=__codelineno-25-5 href=#__codelineno-25-5></a><span class=n>res_tbl</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_tbl</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>arrange</span><span class=p>(</span><span class=n>padj</span><span class=p>)</span>
<a id=__codelineno-25-6 name=__codelineno-25-6 href=#__codelineno-25-6></a>
<a id=__codelineno-25-7 name=__codelineno-25-7 href=#__codelineno-25-7></a><span class=c1>## Populate the genelabels column with contents of the gene symbols column for the first 10 rows, i.e. the top 10 most significantly expressed genes</span>
<a id=__codelineno-25-8 name=__codelineno-25-8 href=#__codelineno-25-8></a><span class=n>res_tbl</span><span class=o>$</span><span class=n>genelabels</span><span class=p>[</span><span class=m>1</span><span class=o>:</span><span class=m>10</span><span class=p>]</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>as.character</span><span class=p>(</span><span class=n>res_tbl</span><span class=o>$</span><span class=n>gene</span><span class=p>[</span><span class=m>1</span><span class=o>:</span><span class=m>10</span><span class=p>])</span>
<a id=__codelineno-25-9 name=__codelineno-25-9 href=#__codelineno-25-9></a>
<a id=__codelineno-25-10 name=__codelineno-25-10 href=#__codelineno-25-10></a><span class=nf>head</span><span class=p>(</span><span class=n>res_tbl</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-26-1 name=__codelineno-26-1 href=#__codelineno-26-1></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>res_tbl</span><span class=p>,</span><span class=w> </span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>log2FoldChange</span><span class=p>,</span><span class=w> </span><span class=n>y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=o>-</span><span class=nf>log10</span><span class=p>(</span><span class=n>padj</span><span class=p>)))</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-26-2 name=__codelineno-26-2 href=#__codelineno-26-2></a><span class=w>  </span><span class=nf>geom_point</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>colour</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>threshold</span><span class=p>))</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-26-3 name=__codelineno-26-3 href=#__codelineno-26-3></a><span class=w>  </span><span class=nf>geom_text_repel</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>label</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>genelabels</span><span class=p>))</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-26-4 name=__codelineno-26-4 href=#__codelineno-26-4></a><span class=w>  </span><span class=nf>ggtitle</span><span class=p>(</span><span class=s>&quot;Mov10 overexpression&quot;</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-26-5 name=__codelineno-26-5 href=#__codelineno-26-5></a><span class=w>  </span><span class=nf>xlab</span><span class=p>(</span><span class=s>&quot;log2 fold change&quot;</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span>
<a id=__codelineno-26-6 name=__codelineno-26-6 href=#__codelineno-26-6></a><span class=w>  </span><span class=nf>ylab</span><span class=p>(</span><span class=s>&quot;-log10 adjusted p-value&quot;</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-26-7 name=__codelineno-26-7 href=#__codelineno-26-7></a><span class=w>  </span><span class=nf>theme</span><span class=p>(</span><span class=n>legend.position</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;none&quot;</span><span class=p>,</span>
<a id=__codelineno-26-8 name=__codelineno-26-8 href=#__codelineno-26-8></a><span class=w>        </span><span class=n>plot.title</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>element_text</span><span class=p>(</span><span class=n>size</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>rel</span><span class=p>(</span><span class=m>1.5</span><span class=p>),</span><span class=w> </span><span class=n>hjust</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.5</span><span class=p>),</span>
<a id=__codelineno-26-9 name=__codelineno-26-9 href=#__codelineno-26-9></a><span class=w>        </span><span class=n>axis.title</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>element_text</span><span class=p>(</span><span class=n>size</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>rel</span><span class=p>(</span><span class=m>1.25</span><span class=p>)))</span><span class=w> </span>
</code></pre></div> <h3 id=heatmap-of-differentially-expressed-genes>Heatmap of differentially expressed genes<a class=headerlink href=#heatmap-of-differentially-expressed-genes title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><a id=__codelineno-27-1 name=__codelineno-27-1 href=#__codelineno-27-1></a><span class=c1># filter significant results from normalized counts</span>
<a id=__codelineno-27-2 name=__codelineno-27-2 href=#__codelineno-27-2></a><span class=n>norm_sig</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>normalized_counts</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>as_tibble</span><span class=p>(</span><span class=n>rownames</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;gene&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-27-3 name=__codelineno-27-3 href=#__codelineno-27-3></a><span class=w>  </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>gene</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=n>sig_res</span><span class=o>$</span><span class=n>gene</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=n>var</span><span class=o>=</span><span class=s>&quot;gene&quot;</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-28-1 name=__codelineno-28-1 href=#__codelineno-28-1></a><span class=nf>pheatmap</span><span class=p>(</span><span class=n>norm_sig</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-28-2 name=__codelineno-28-2 href=#__codelineno-28-2></a><span class=w>         </span><span class=n>cluster_rows</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>T</span><span class=p>,</span><span class=w> </span><span class=c1>#cluster by expression pattern</span>
<a id=__codelineno-28-3 name=__codelineno-28-3 href=#__codelineno-28-3></a><span class=w>         </span><span class=n>scale</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;row&quot;</span><span class=p>,</span><span class=w> </span><span class=c1># scale by gene so expression pattern is visible</span>
<a id=__codelineno-28-4 name=__codelineno-28-4 href=#__codelineno-28-4></a><span class=w>         </span><span class=n>treeheight_row</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0</span><span class=p>,</span><span class=w> </span><span class=c1># dont show the row dendogram</span>
<a id=__codelineno-28-5 name=__codelineno-28-5 href=#__codelineno-28-5></a><span class=w>         </span><span class=n>show_rownames</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>F</span><span class=p>,</span><span class=w> </span><span class=c1># remove rownames so it is more clear</span>
<a id=__codelineno-28-6 name=__codelineno-28-6 href=#__codelineno-28-6></a><span class=w>         </span><span class=n>annotation</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>meta</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>column_to_rownames</span><span class=p>(</span><span class=n>var</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;sample&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=s>&quot;condition&quot;</span><span class=p>)</span>
<a id=__codelineno-28-7 name=__codelineno-28-7 href=#__codelineno-28-7></a><span class=w>         </span><span class=p>)</span>
</code></pre></div> <h2 id=perform-analysis-to-extract-functional-significance-of-results-go-or-kegg-enrichment-gsea-etc>Perform analysis to extract functional significance of results: GO or KEGG enrichment, GSEA, etc.<a class=headerlink href=#perform-analysis-to-extract-functional-significance-of-results-go-or-kegg-enrichment-gsea-etc title="Permanent link">&para;</a></h2> <h3 id=annotate-with-annotables>Annotate with <code>annotables</code><a class=headerlink href=#annotate-with-annotables title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><a id=__codelineno-29-1 name=__codelineno-29-1 href=#__codelineno-29-1></a><span class=n>ids</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>grch37</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>ensgene</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=n>res_tbl</span><span class=o>$</span><span class=n>gene</span><span class=p>)</span><span class=w> </span>
<a id=__codelineno-29-2 name=__codelineno-29-2 href=#__codelineno-29-2></a><span class=n>res_ids</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>inner_join</span><span class=p>(</span><span class=n>res_tbl</span><span class=p>,</span><span class=w> </span><span class=n>ids</span><span class=p>,</span><span class=w> </span><span class=n>by</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;gene&quot;</span><span class=o>=</span><span class=s>&quot;ensgene&quot;</span><span class=p>))</span>
</code></pre></div> <h3 id=perform-enrichment-analysis-of-go-terms-can-be-done-as-well-with-kegg-pathways>Perform enrichment analysis of GO terms (can be done as well with KEGG pathways)<a class=headerlink href=#perform-enrichment-analysis-of-go-terms-can-be-done-as-well-with-kegg-pathways title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><a id=__codelineno-30-1 name=__codelineno-30-1 href=#__codelineno-30-1></a><span class=c1># Create background dataset for hypergeometric testing using all genes tested for significance in the results</span>
<a id=__codelineno-30-2 name=__codelineno-30-2 href=#__codelineno-30-2></a><span class=n>all_genes</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>res_ids</span><span class=p>,</span><span class=w> </span><span class=o>!</span><span class=nf>is.na</span><span class=p>(</span><span class=n>gene</span><span class=p>))</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-30-3 name=__codelineno-30-3 href=#__codelineno-30-3></a><span class=w>  </span><span class=nf>pull</span><span class=p>(</span><span class=n>gene</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-30-4 name=__codelineno-30-4 href=#__codelineno-30-4></a><span class=w>  </span><span class=nf>as.character</span><span class=p>()</span>
<a id=__codelineno-30-5 name=__codelineno-30-5 href=#__codelineno-30-5></a>
<a id=__codelineno-30-6 name=__codelineno-30-6 href=#__codelineno-30-6></a><span class=c1># Extract significant results</span>
<a id=__codelineno-30-7 name=__codelineno-30-7 href=#__codelineno-30-7></a><span class=n>sig</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>res_ids</span><span class=p>,</span><span class=w> </span><span class=n>padj</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=m>0.05</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=o>!</span><span class=nf>is.na</span><span class=p>(</span><span class=n>gene</span><span class=p>))</span>
<a id=__codelineno-30-8 name=__codelineno-30-8 href=#__codelineno-30-8></a>
<a id=__codelineno-30-9 name=__codelineno-30-9 href=#__codelineno-30-9></a><span class=n>sig_genes</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>sig</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-30-10 name=__codelineno-30-10 href=#__codelineno-30-10></a><span class=w>  </span><span class=nf>pull</span><span class=p>(</span><span class=n>gene</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span>
<a id=__codelineno-30-11 name=__codelineno-30-11 href=#__codelineno-30-11></a><span class=w>  </span><span class=nf>as.character</span><span class=p>()</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-31-1 name=__codelineno-31-1 href=#__codelineno-31-1></a><span class=c1># Perform enrichment analysis</span>
<a id=__codelineno-31-2 name=__codelineno-31-2 href=#__codelineno-31-2></a><span class=n>ego</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>enrichGO</span><span class=p>(</span><span class=n>gene</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sig_genes</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-31-3 name=__codelineno-31-3 href=#__codelineno-31-3></a><span class=w>                </span><span class=n>universe</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>all_genes</span><span class=p>,</span>
<a id=__codelineno-31-4 name=__codelineno-31-4 href=#__codelineno-31-4></a><span class=w>                </span><span class=n>keyType</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;ENSEMBL&quot;</span><span class=p>,</span>
<a id=__codelineno-31-5 name=__codelineno-31-5 href=#__codelineno-31-5></a><span class=w>                </span><span class=n>OrgDb</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>org.Hs.eg.db</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-31-6 name=__codelineno-31-6 href=#__codelineno-31-6></a><span class=w>                </span><span class=n>ont</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;BP&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-31-7 name=__codelineno-31-7 href=#__codelineno-31-7></a><span class=w>                </span><span class=n>pAdjustMethod</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;BH&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-31-8 name=__codelineno-31-8 href=#__codelineno-31-8></a><span class=w>                </span><span class=n>qvalueCutoff</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.05</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-31-9 name=__codelineno-31-9 href=#__codelineno-31-9></a><span class=w>                </span><span class=n>readable</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
<a id=__codelineno-31-10 name=__codelineno-31-10 href=#__codelineno-31-10></a><span class=n>ego</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>enrichplot</span><span class=o>::</span><span class=nf>pairwise_termsim</span><span class=p>(</span><span class=n>ego</span><span class=p>)</span>
</code></pre></div> <h3 id=visualize-result>Visualize result<a class=headerlink href=#visualize-result title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><a id=__codelineno-32-1 name=__codelineno-32-1 href=#__codelineno-32-1></a><span class=nf>dotplot</span><span class=p>(</span><span class=n>ego</span><span class=p>,</span><span class=w> </span><span class=n>showCategory</span><span class=o>=</span><span class=m>50</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-33-1 name=__codelineno-33-1 href=#__codelineno-33-1></a><span class=nf>emapplot</span><span class=p>(</span><span class=n>ego</span><span class=p>,</span><span class=w> </span><span class=n>showCategory</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>50</span><span class=p>)</span>
</code></pre></div> <h3 id=cnetplot>Cnetplot<a class=headerlink href=#cnetplot title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><a id=__codelineno-34-1 name=__codelineno-34-1 href=#__codelineno-34-1></a><span class=c1>## To color genes by log2 fold changes, we need to extract the log2 fold changes from our results table creating a named vector</span>
<a id=__codelineno-34-2 name=__codelineno-34-2 href=#__codelineno-34-2></a><span class=n>sig_foldchanges</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>sig</span><span class=o>$</span><span class=n>log2FoldChange</span>
<a id=__codelineno-34-3 name=__codelineno-34-3 href=#__codelineno-34-3></a>
<a id=__codelineno-34-4 name=__codelineno-34-4 href=#__codelineno-34-4></a><span class=nf>names</span><span class=p>(</span><span class=n>sig_foldchanges</span><span class=p>)</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>sig</span><span class=o>$</span><span class=n>gene</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-35-1 name=__codelineno-35-1 href=#__codelineno-35-1></a><span class=c1>## Cnetplot details the genes associated with one or more terms - by default gives the top 5 significant terms (by padj)</span>
<a id=__codelineno-35-2 name=__codelineno-35-2 href=#__codelineno-35-2></a><span class=nf>cnetplot</span><span class=p>(</span><span class=n>ego</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-35-3 name=__codelineno-35-3 href=#__codelineno-35-3></a><span class=w>         </span><span class=n>categorySize</span><span class=o>=</span><span class=s>&quot;pvalue&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-35-4 name=__codelineno-35-4 href=#__codelineno-35-4></a><span class=w>         </span><span class=n>showCategory</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>5</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-35-5 name=__codelineno-35-5 href=#__codelineno-35-5></a><span class=w>         </span><span class=n>foldChange</span><span class=o>=</span><span class=n>sig_foldchanges</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-35-6 name=__codelineno-35-6 href=#__codelineno-35-6></a><span class=w>         </span><span class=n>vertex.label.font</span><span class=o>=</span><span class=m>6</span><span class=p>)</span>
</code></pre></div> <h3 id=perform-gsea-analysis-of-kegg-pathways-can-be-done-as-well-with-go-terms>Perform GSEA analysis of KEGG pathways (can be done as well with GO terms)<a class=headerlink href=#perform-gsea-analysis-of-kegg-pathways-can-be-done-as-well-with-go-terms title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><a id=__codelineno-36-1 name=__codelineno-36-1 href=#__codelineno-36-1></a><span class=c1># Extract entrez IDs. IDs should not be duplicated or NA</span>
<a id=__codelineno-36-2 name=__codelineno-36-2 href=#__codelineno-36-2></a><span class=n>res_entrez</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>res_ids</span><span class=p>,</span><span class=w> </span><span class=n>entrez</span><span class=w> </span><span class=o>!=</span><span class=w> </span><span class=s>&quot;NA&quot;</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=n>entrez</span><span class=w> </span><span class=o>!=</span><span class=w> </span><span class=s>&quot;NULL&quot;</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=nf>duplicated</span><span class=p>(</span><span class=n>entrez</span><span class=p>)</span><span class=o>==</span><span class=bp>F</span><span class=p>)</span>
<a id=__codelineno-36-3 name=__codelineno-36-3 href=#__codelineno-36-3></a>
<a id=__codelineno-36-4 name=__codelineno-36-4 href=#__codelineno-36-4></a><span class=c1>## Extract the foldchanges</span>
<a id=__codelineno-36-5 name=__codelineno-36-5 href=#__codelineno-36-5></a><span class=n>foldchanges</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_entrez</span><span class=o>$</span><span class=n>log2FoldChange</span>
<a id=__codelineno-36-6 name=__codelineno-36-6 href=#__codelineno-36-6></a>
<a id=__codelineno-36-7 name=__codelineno-36-7 href=#__codelineno-36-7></a><span class=c1>## Name each fold change with the corresponding Entrez ID</span>
<a id=__codelineno-36-8 name=__codelineno-36-8 href=#__codelineno-36-8></a><span class=nf>names</span><span class=p>(</span><span class=n>foldchanges</span><span class=p>)</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>res_entrez</span><span class=o>$</span><span class=n>entrez</span>
<a id=__codelineno-36-9 name=__codelineno-36-9 href=#__codelineno-36-9></a>
<a id=__codelineno-36-10 name=__codelineno-36-10 href=#__codelineno-36-10></a><span class=c1>## Sort fold changes in decreasing order</span>
<a id=__codelineno-36-11 name=__codelineno-36-11 href=#__codelineno-36-11></a><span class=n>foldchanges</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>sort</span><span class=p>(</span><span class=n>foldchanges</span><span class=p>,</span><span class=w> </span><span class=n>decreasing</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-37-1 name=__codelineno-37-1 href=#__codelineno-37-1></a><span class=c1># Run GSEA of KEGG</span>
<a id=__codelineno-37-2 name=__codelineno-37-2 href=#__codelineno-37-2></a><span class=n>gseaKEGG</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>gseKEGG</span><span class=p>(</span><span class=n>geneList</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>foldchanges</span><span class=p>,</span><span class=w> </span><span class=c1># ordered named vector of fold changes (Entrez IDs are the associated names)</span>
<a id=__codelineno-37-3 name=__codelineno-37-3 href=#__codelineno-37-3></a><span class=w>              </span><span class=n>organism</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;hsa&quot;</span><span class=p>,</span><span class=w> </span><span class=c1># supported organisms listed below</span>
<a id=__codelineno-37-4 name=__codelineno-37-4 href=#__codelineno-37-4></a><span class=w>              </span><span class=n>pvalueCutoff</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.05</span><span class=p>,</span><span class=w> </span><span class=c1># padj cutoff value</span>
<a id=__codelineno-37-5 name=__codelineno-37-5 href=#__codelineno-37-5></a><span class=w>              </span><span class=n>verbose</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-37-6 name=__codelineno-37-6 href=#__codelineno-37-6></a>
<a id=__codelineno-37-7 name=__codelineno-37-7 href=#__codelineno-37-7></a><span class=n>gseaKEGG_results</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>gseaKEGG</span><span class=o>@</span><span class=n>result</span>
<a id=__codelineno-37-8 name=__codelineno-37-8 href=#__codelineno-37-8></a><span class=nf>head</span><span class=p>(</span><span class=n>gseaKEGG_results</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-38-1 name=__codelineno-38-1 href=#__codelineno-38-1></a><span class=c1>## Plot the GSEA plot for a single enriched pathway:</span>
<a id=__codelineno-38-2 name=__codelineno-38-2 href=#__codelineno-38-2></a><span class=nf>gseaplot</span><span class=p>(</span><span class=n>gseaKEGG</span><span class=p>,</span><span class=w> </span><span class=n>geneSetID</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>gseaKEGG_results</span><span class=o>$</span><span class=n>ID</span><span class=p>[</span><span class=m>1</span><span class=p>])</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-39-1 name=__codelineno-39-1 href=#__codelineno-39-1></a><span class=c1>## Output images for a single significant KEGG pathway</span>
<a id=__codelineno-39-2 name=__codelineno-39-2 href=#__codelineno-39-2></a><span class=nf>pathview</span><span class=p>(</span><span class=n>gene.data</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>foldchanges</span><span class=p>,</span>
<a id=__codelineno-39-3 name=__codelineno-39-3 href=#__codelineno-39-3></a><span class=w>              </span><span class=n>pathway.id</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>gseaKEGG_results</span><span class=o>$</span><span class=n>ID</span><span class=p>[</span><span class=m>1</span><span class=p>],</span>
<a id=__codelineno-39-4 name=__codelineno-39-4 href=#__codelineno-39-4></a><span class=w>              </span><span class=n>species</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;hsa&quot;</span><span class=p>,</span>
<a id=__codelineno-39-5 name=__codelineno-39-5 href=#__codelineno-39-5></a><span class=w>              </span><span class=n>limit</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>list</span><span class=p>(</span><span class=n>gene</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2</span><span class=p>,</span><span class=w> </span><span class=c1># value gives the max/min limit for foldchanges</span>
<a id=__codelineno-39-6 name=__codelineno-39-6 href=#__codelineno-39-6></a><span class=w>              </span><span class=n>cpd</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=p>))</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-40-1 name=__codelineno-40-1 href=#__codelineno-40-1></a><span class=n>knitr</span><span class=o>::</span><span class=nf>include_graphics</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=s>&quot;./&quot;</span><span class=p>,</span><span class=n>gseaKEGG_results</span><span class=o>$</span><span class=n>ID</span><span class=p>[</span><span class=m>1</span><span class=p>],</span><span class=s>&quot;.png&quot;</span><span class=p>))</span>
</code></pre></div> <h2 id=make-sure-to-output-the-versions-of-all-tools-used-in-the-de-analysis>Make sure to output the versions of all tools used in the DE analysis:<a class=headerlink href=#make-sure-to-output-the-versions-of-all-tools-used-in-the-de-analysis title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code><a id=__codelineno-41-1 name=__codelineno-41-1 href=#__codelineno-41-1></a><span class=nf>sessionInfo</span><span class=p>()</span>
</code></pre></div> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "assets/javascripts/workers/search.db81ec45.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=assets/javascripts/bundle.a00a7c5e.min.js></script> <script src=javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>